// hfy-tools@2.1.6/dist/component/web/Select/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomSelect = exports.CustomOptionGroup = exports.CustomOption = void 0;
// 自定义选项组件
class CustomOption extends HTMLElement {
    static observedAttributes = ['value', 'disabled'];
    constructor() {
        super();
    }
    connectedCallback() {
        this.render();
    }
    attributeChangedCallback() {
        this.render();
    }
    get value() {
        return this.getAttribute('value') || this.textContent;
    }
    get disabled() {
        return this.hasAttribute('disabled');
    }
    render() {
        // 选项的渲染由父组件custom-select处理
    }
}
exports.CustomOption = CustomOption;
// 自定义选项分组组件
class CustomOptionGroup extends HTMLElement {
    static observedAttributes = ['label'];
    constructor() {
        super();
    }
    connectedCallback() {
        this.render();
    }
    attributeChangedCallback() {
        this.render();
    }
    get label() {
        return this.getAttribute('label') || '';
    }
    render() {
        // 分组渲染由父组件custom-select处理
    }
}
exports.CustomOptionGroup = CustomOptionGroup;
// 自定义选择框组件
class CustomSelect extends HTMLElement {
    static observedAttributes = [
        'value',
        'placeholder',
        'disabled',
        'size',
        'multiple',
        'searchable',
    ];
    _value;
    _options;
    _isOpen;
    _filteredOptions;
    constructor() {
        super();
        this.attachShadow({ mode: 'open' });
        this._value = [];
        this._options = [];
        this._isOpen = false;
        this._filteredOptions = [];
        // 点击外部关闭下拉框
        this._onClickOutside = this._onClickOutside.bind(this);
    }
    connectedCallback() {
        this._initOptions();
        this._attachEventListeners();
        // 设置初始tabindex以支持键盘导航
        if (!this.hasAttribute('tabindex')) {
            this.setAttribute('tabindex', '0');
        }
        this.render();
    }
    disconnectedCallback() {
        this._removeEventListeners();
    }
    attributeChangedCallback(name, oldValue, newValue) {
        if (oldValue !== newValue) {
            if (name === 'value') {
                this._parseValue(newValue);
            }
            this.render();
        }
    }
    _parseValue(value) {
        if (!value) {
            this._value = [];
            return;
        }
        try {
            // 尝试解析JSON数组
            this._value = JSON.parse(value);
            if (!Array.isArray(this._value)) {
                this._value = [this._value];
            }
        }
        catch (e) {
            // 如果不是JSON，则当作单个值处理
            this._value = [value];
        }
    }
    _initOptions() {
        this._options = [];
        const options = this.querySelectorAll('custom-option');
        options.forEach((option) => {
            this._options.push({
                value: option.value,
                text: option.textContent,
                disabled: option.disabled,
            });
        });
        this._filteredOptions = [...this._options];
    }
    _attachEventListeners() {
        this.shadowRoot.addEventListener('click', this._onTriggerClick.bind(this));
        this.addEventListener('keydown', this._onKeyDown.bind(this));
        // 点击外部关闭下拉框
        document.addEventListener('click', this._onClickOutside);
    }
    _removeEventListeners() {
        document.removeEventListener('click', this._onClickOutside);
    }
    _onClickOutside(event) {
        if (!this.contains(event.target) && this._isOpen) {
            this.close();
        }
    }
    _onTriggerClick(event) {
        if (this.disabled)
            return;
        if (this._isOpen) {
            this.close();
        }
        else {
            this.open();
        }
    }
    _onKeyDown(event) {
        if (this.disabled)
            return;
        switch (event.key) {
            case ' ':
            case 'Enter':
                event.preventDefault();
                if (this._isOpen) {
                    this.close();
                }
                else {
                    this.open();
                }
                break;
            case 'Escape':
                if (this._isOpen) {
                    this.close();
                }
                break;
            case 'ArrowDown':
                event.preventDefault();
                if (!this._isOpen) {
                    this.open();
                }
                break;
        }
    }
    _onOptionClick(value, text) {
        if (this.multiple) {
            this._toggleMultipleValue(value, text);
        }
        else {
            this._setSingleValue(value, text);
        }
        // 无论单选还是多选，选择后都关闭下拉框
        this.close();
        this._updateSelectedDisplay();
        // 触发change事件
        this.dispatchEvent(new CustomEvent('change', {
            detail: {
                value: this.multiple ? this._value : this._value[0],
                values: this._value,
            },
            bubbles: true,
        }));
    }
    _setSingleValue(value, text) {
        this._value = [value];
        this.setAttribute('value', value);
    }
    _toggleMultipleValue(value, text) {
        const index = this._value.indexOf(value);
        if (index > -1) {
            this._value.splice(index, 1);
        }
        else {
            this._value.push(value);
        }
        // 更新属性值
        this.setAttribute('value', JSON.stringify(this._value));
    }
    _updateSelectedDisplay() {
        const displayEl = this.shadowRoot.querySelector('.select-display');
        if (!displayEl)
            return;
        if (this.multiple && this._value.length > 0) {
            const selectedOptions = this._options.filter((opt) => this._value.includes(opt.value));
            if (selectedOptions.length > 0) {
                displayEl.textContent = selectedOptions
                    .map((opt) => opt.text)
                    .join(', ');
            }
            else {
                displayEl.textContent = this.placeholder;
            }
        }
        else if (!this.multiple && this._value.length > 0) {
            const selectedOption = this._options.find((opt) => opt.value === this._value[0]);
            displayEl.textContent = selectedOption
                ? selectedOption.text
                : this._value[0];
        }
        else {
            displayEl.textContent = this.placeholder;
        }
    }
    _filterOptions(searchText) {
        if (!searchText) {
            this._filteredOptions = [...this._options];
        }
        else {
            this._filteredOptions = this._options.filter((option) => option.text.toLowerCase().includes(searchText.toLowerCase()));
        }
        this.renderOptions();
    }
    open() {
        if (this.disabled)
            return;
        this._isOpen = true;
        this.render();
        // 触发open事件
        this.dispatchEvent(new CustomEvent('open', {
            bubbles: true,
        }));
    }
    close() {
        this._isOpen = false;
        this.render();
        // 触发close事件
        this.dispatchEvent(new CustomEvent('close', {
            bubbles: true,
        }));
    }
    get value() {
        return this.multiple ? this._value : this._value[0] || null;
    }
    set value(newValue) {
        if (this.multiple) {
            if (Array.isArray(newValue)) {
                this._value = [...newValue];
            }
            else {
                this._value = newValue ? [newValue] : [];
            }
            this.setAttribute('value', JSON.stringify(this._value));
        }
        else {
            this._value = newValue ? [newValue] : [];
            this.setAttribute('value', newValue || '');
        }
        this._updateSelectedDisplay();
    }
    get placeholder() {
        return this.getAttribute('placeholder') || '请选择';
    }
    get disabled() {
        return this.hasAttribute('disabled');
    }
    get size() {
        return this.getAttribute('size') || 'medium';
    }
    get multiple() {
        return this.hasAttribute('multiple');
    }
    get searchable() {
        return this.hasAttribute('searchable');
    }
    render() {
        const disabled = this.disabled;
        const size = this.size;
        const multiple = this.multiple;
        const searchable = this.searchable;
        const isOpen = this._isOpen;
        const selectClass = [
            'select',
            `select-${size}`,
            disabled ? 'select-disabled' : '',
            isOpen ? 'select-open' : '',
        ]
            .filter(Boolean)
            .join(' ');
        this.shadowRoot.innerHTML = `
                    <style>
                        :host {
                            display: inline-block;
                            width: 100%;
                            position: relative;
                        }

                        .select {
                            position: relative;
                            width: 100%;
                            background: white;
                            border: 1px solid #d1d5db;
                            border-radius: 6px;
                            cursor: pointer;
                            transition: all 0.2s ease;
                            user-select: none;
                        }

                        .select:hover:not(.select-disabled) {
                            border-color: #9ca3af;
                        }

                        .select:focus-within:not(.select-disabled) {
                            border-color: #3b82f6;
                            box-shadow: 0 0 0 3px rgba(59, 130, 246, 0.1);
                        }

                        .select-disabled {
                            background-color: #f9fafb;
                            color: #9ca3af;
                            cursor: not-allowed;
                        }

                        .select-small {
                            font-size: 14px;
                        }

                        .select-medium {
                            font-size: 16px;
                        }

                        .select-large {
                            font-size: 18px;
                        }

                        .select-trigger {
                            display: flex;
                            align-items: center;
                            justify-content: space-between;
                            padding: 8px 12px;
                            min-height: 40px;
                        }

                        .select-small .select-trigger {
                            padding: 6px 10px;
                            min-height: 32px;
                        }

                        .select-large .select-trigger {
                            padding: 12px 16px;
                            min-height: 48px;
                        }

                        .select-display {
                            flex: 1;
                            overflow: hidden;
                            text-overflow: ellipsis;
                            white-space: nowrap;
                            color: #6b7280;
                        }

                        .select-value {
                            color: #111827;
                        }

                        .select-arrow {
                            transition: transform 0.2s ease;
                            margin-left: 8px;
                            color: #6b7280;
                        }

                        .select-open .select-arrow {
                            transform: rotate(180deg);
                        }

                        .select-dropdown {
                            position: absolute;
                            top: 100%;
                            left: 0;
                            right: 0;
                            background: white;
                            border: 1px solid #e5e7eb;
                            border-radius: 6px;
                            box-shadow: 0 10px 15px -3px rgba(0, 0, 0, 0.1);
                            z-index: 50;
                            margin-top: 4px;
                            max-height: 300px;
                            overflow-y: auto;
                            display: ${isOpen ? 'block' : 'none'};
                        }

                        .select-search {
                            padding: 8px 12px;
                            border-bottom: 1px solid #e5e7eb;
                            position: sticky;
                            top: 0;
                            background: white;
                            z-index: 1;
                        }

                        .select-search-input {
                            width: 100%;
                            padding: 8px;
                            border: 1px solid #d1d5db;
                            border-radius: 4px;
                            font-size: 14px;
                        }

                        .select-search-input:focus {
                            outline: none;
                            border-color: #3b82f6;
                        }

                        .select-options {
                            padding: 4px 0;
                            max-height: 250px;
                            overflow-y: auto;
                        }

                        .select-group {
                            padding: 8px 0;
                        }

                        .select-group-label {
                            padding: 8px 12px 4px;
                            font-size: 12px;
                            color: #6b7280;
                            font-weight: 600;
                            text-transform: uppercase;
                            letter-spacing: 0.5px;
                            background-color: #f9fafb;
                            border-bottom: 1px solid #e5e7eb;
                            margin-bottom: 4px;
                        }

                        .select-option {
                            display: flex;
                            align-items: center;
                            padding: 8px 12px;
                            cursor: pointer;
                            transition: background-color 0.1s ease;
                        }

                        .select-option:hover:not(.select-option-disabled) {
                            background-color: #f3f4f6;
                        }

                        .select-option-selected {
                            background-color: #eff6ff;
                            color: #1d4ed8;
                        }

                        .select-option-disabled {
                            color: #9ca3af;
                            cursor: not-allowed;
                        }

                        .select-option-checkbox {
                            margin-right: 8px;
                            width: 16px;
                            height: 16px;
                            border: 1px solid #d1d5db;
                            border-radius: 3px;
                            display: flex;
                            align-items: center;
                            justify-content: center;
                            background: white;
                        }

                        .select-option-selected .select-option-checkbox {
                            background-color: #3b82f6;
                            border-color: #3b82f6;
                            color: white;
                        }

                        .select-option-text {
                            flex: 1;
                        }

                        .select-empty {
                            padding: 16px;
                            text-align: center;
                            color: #6b7280;
                            font-size: 14px;
                        }
                    </style>

                    <div class="${selectClass}">
                        <div class="select-trigger">
                            <div class="select-display">
                                <span class="select-value">${this.placeholder}</span>
                            </div>
                            <div class="select-arrow">
                                <svg width="16" height="16" viewBox="0 0 24 24" fill="none" stroke="currentColor" stroke-width="2">
                                    <polyline points="6 9 12 15 18 9"></polyline>
                                </svg>
                            </div>
                        </div>

                        <div class="select-dropdown">
                            ${searchable
            ? `
                                <div class="select-search">
                                    <input type="text" class="select-search-input" placeholder="搜索..." />
                                </div>
                            `
            : ''}

                            <div class="select-options">
                                ${this._renderOptions()}
                            </div>
                        </div>
                    </div>
                `;
        this._updateSelectedDisplay();
        // 绑定事件
        if (isOpen) {
            this._bindOptionEvents();
            // 绑定搜索事件
            if (searchable) {
                const searchInput = this.shadowRoot.querySelector('.select-search-input');
                searchInput.addEventListener('input', (e) => {
                    this._filterOptions(e.target.value);
                });
                // 聚焦搜索框
                searchInput.focus();
            }
        }
    }
    renderOptions() {
        const optionsContainer = this.shadowRoot.querySelector('.select-options');
        if (optionsContainer) {
            optionsContainer.innerHTML = this._renderOptions();
            this._bindOptionEvents();
        }
    }
    _renderOptions() {
        // 检查是否有分组
        const groups = this.querySelectorAll('custom-option-group');
        if (groups.length > 0) {
            return this._renderGroupedOptions(groups);
        }
        // 没有分组，渲染普通选项
        if (this._filteredOptions.length === 0) {
            return `<div class="select-empty">${this.searchable ? '未找到匹配项' : '暂无选项'}</div>`;
        }
        return this._filteredOptions
            .map((option) => this._renderOption(option))
            .join('');
    }
    _renderGroupedOptions(groups) {
        let html = '';
        // 首先渲染没有分组的独立选项
        const standaloneOptions = Array.from(this.children).filter((child) => child.tagName === 'CUSTOM-OPTION');
        if (standaloneOptions.length > 0) {
            html += standaloneOptions
                .map((option) => this._renderOption({
                value: option.value,
                text: option.textContent,
                disabled: option.disabled,
            }))
                .join('');
        }
        // 然后渲染分组
        groups.forEach((group) => {
            const label = group.getAttribute('label') || '';
            const options = group.querySelectorAll('custom-option');
            if (options.length > 0) {
                html += `
                            <div class="select-group">
                                <div class="select-group-label">${label}</div>
                                ${Array.from(options)
                    .map((option) => {
                    return this._renderOption({
                        value: option.value,
                        text: option.textContent,
                        disabled: option.disabled,
                    });
                })
                    .join('')}
                            </div>
                        `;
            }
        });
        if (!html) {
            return `<div class="select-empty">${this.searchable ? '未找到匹配项' : '暂无选项'}</div>`;
        }
        return html;
    }
    _renderOption(option) {
        const isSelected = this._value.includes(option.value);
        const optionClass = [
            'select-option',
            isSelected ? 'select-option-selected' : '',
            option.disabled ? 'select-option-disabled' : '',
        ]
            .filter(Boolean)
            .join(' ');
        return `
                    <div class="${optionClass}" data-value="${option.value}">
                        ${this.multiple
            ? `
                            <div class="select-option-checkbox">
                                ${isSelected ? '✓' : ''}
                            </div>
                        `
            : ''}
                        <div class="select-option-text">${option.text}</div>
                    </div>
                `;
    }
    _bindOptionEvents() {
        const options = this.shadowRoot.querySelectorAll('.select-option:not(.select-option-disabled)');
        options.forEach((option) => {
            option.addEventListener('click', () => {
                const value = option.dataset.value;
                const text = option.querySelector('.select-option-text').textContent;
                this._onOptionClick(value, text);
            });
        });
    }
}
exports.CustomSelect = CustomSelect;
