// hfy-tools/dist/utils/document.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.clearCookie = exports.getCookie = exports.setCookie = exports.getCurrentPlatform = exports.strPage = exports.getQueryString = exports.loadScript = exports.isMobile = exports.getUrlParams = void 0;
/**
 * url参数转对象
 * @param query
 * @returns
 */
const getUrlParams = (query) => Array.from(new URLSearchParams(query)).reduce((p, [k, v]) => Object.assign({}, p, { [k]: p[k] ? (Array.isArray(p[k]) ? p[k] : [p[k]]).concat(v) : v }), {});
exports.getUrlParams = getUrlParams;
/**
 * 移动端
 */
exports.isMobile = /Android|webOS|iPhone|iPad|iPod|BlackBerry|IEMobile|Opera Mini/i.test(navigator?.userAgent);
/**
 * 加载一个远程脚本
 * @param {String} src 一个远程脚本
 * @param {Function} callback 回调
 */
const loadScript = (src, callback = (_) => { }) => {
    const docTag = window.top.document;
    if (docTag.querySelector(`[src='${src}']`) || docTag.getElementById(src)) {
        callback(true);
        return;
    }
    const $script = document.createElement('script');
    $script.src = src;
    $script.id = src;
    $script.async = true;
    $script.defer = true;
    docTag.children[0].appendChild($script);
    if ('onload' in $script) {
        $script.onload = (_) => {
            callback(true);
        };
        $script.onerror = (_) => {
            callback(new Error(`Failed to load ${src}`));
        };
        return;
    }
    // ie
    if ($script['onreadystatechange']) {
        Object.bind($script, 'onreadystatechange', function () {
            if (['loaded', 'complete'].includes(this.readyState)) {
                callback(true);
            }
            else {
                setTimeout(function () {
                    arguments.callee();
                }, 500);
            }
        });
    }
};
exports.loadScript = loadScript;
/**
 * 获取URL GET参数
 * @param name
 * @returns
 */
const getQueryString = (name) => {
    var reg = new RegExp('(^|&)' + name + '=([^&]*)(&|$)');
    var r = window.location.search.substr(1).match(reg);
    if (r != null)
        return unescape(r[2]);
    return '';
};
exports.getQueryString = getQueryString;
/**
 * 获取当前页面名称
 * @returns
 */
const strPage = () => {
    var strUrl = window.location.href;
    var arrUrl = strUrl.split('/');
    var strPage = arrUrl[arrUrl.length - 1];
    if (strPage.indexOf('?') > -1) {
        var pageName = strPage.split('?');
        strPage = pageName[0];
    }
    return strPage;
};
exports.strPage = strPage;
/**
 * 操作系统
 * @returns
 */
const getCurrentPlatform = () => {
    return {
        Android: function () {
            //@ts-ignore
            return navigator?.userAgent?.match?.(/Android/i) ? true : false;
        },
        BlackBerry: function () {
            //@ts-ignore
            return navigator?.userAgent?.match?.(/BlackBerry/i) ? true : false;
        },
        iOS: function () {
            //@ts-ignore
            return navigator?.userAgent?.match?.(/iPhone|iPad|iPod/i) ? true : false;
        },
        Windows: function () {
            //@ts-ignore
            return navigator?.userAgent?.match?.(/IEMobile/i) ? true : false;
        },
        any: function () {
            return this.Android() || this.BlackBerry() || this.iOS() || this.Windows();
        }
    };
};
exports.getCurrentPlatform = getCurrentPlatform;
/**
 * 设置cookie
 * @param name
 * @param value
 */
const setCookie = (name, value) => {
    if (value) {
        var Days = 365;
        var exp = new Date();
        exp.setTime(exp.getTime() + Days * 24 * 60 * 60 * 1000);
        document.cookie = name + '=' + escape(value) + ';expires=' + exp.toDateString();
    }
};
exports.setCookie = setCookie;
/**
 * 获取cookie
 * @param name
 * @returns
 */
const getCookie = (name) => {
    if (document.cookie.length > 0) {
        var begin = document.cookie.indexOf(name + '=');
        if (begin !== -1) {
            // cookie值的初始位置
            begin += name.length + 1;
            // 结束位置
            var end = document.cookie.indexOf(';', begin);
            if (end === -1) {
                // 没有;则end为字符串结束位置
                end = document.cookie.length;
            }
            return unescape(document.cookie.substring(begin, end));
        }
    }
    // cookie不存在返回
    return '';
};
exports.getCookie = getCookie;
/**
 * 清除所有cookie
 */
const clearCookie = () => {
    var keys = document.cookie.match(/[^ =;]+(?=\=)/g);
    if (keys) {
        for (var i = keys.length; i--;) {
            document.cookie = keys[i] + '=0;expires=' + new Date(0).toUTCString();
        }
    }
};
exports.clearCookie = clearCookie;
