// hfy-tools/dist/utils/file.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FILE_TYPES = void 0;
exports.overwrite = overwrite;
exports.chooseFile = chooseFile;
exports.downloadFile = downloadFile;
exports.getFileType = getFileType;
exports.urlToBlob = urlToBlob;
async function overwrite(content, options) {
    if (!options.compatible && window.hasOwnProperty('showOpenFilePicker')) {
        let { accept = '*', multiple, description } = options;
        const acceptMap = {};
        for (let a of accept.split(',')) {
            acceptMap[a] = [];
        }
        //@ts-ignore
        const fileHandleList = await window.showOpenFilePicker?.({
            multiple,
            excludeAcceptAllOption: false,
            types: [
                {
                    description,
                    accept: acceptMap,
                },
            ],
        });
        for (const fileHandle of fileHandleList) {
            // 创建一个可写流
            const writable = await fileHandle.createWritable();
            // 写入数据
            await writable.write(content);
            // 关闭流
            await writable.close();
        }
    }
    throw new Error('浏览器不支持!');
}
async function chooseFile(options) {
    let { accept = '*', compatible = true, cancel = 300, multiple, webkitdirectory, description, } = options;
    if (options.forceInput !== true) {
        accept.constructor === Array && (accept = accept.join(','));
        // 实验性功能
        if (!compatible && window.hasOwnProperty('showOpenFilePicker')) {
            const files = [];
            const acceptMap = {};
            for (let a of accept.split(',')) {
                acceptMap[a] = [];
            }
            //@ts-ignore
            const fileHandleList = await window.showOpenFilePicker?.({
                multiple,
                excludeAcceptAllOption: false,
                types: [
                    {
                        description,
                        accept: acceptMap,
                    },
                ],
            });
            for (const f of fileHandleList) {
                files.push(await f.getFile());
            }
            return files;
        }
    }
    const inpEle = document.createElement('input');
    inpEle.id = `__file_${Math.trunc(Math.random() * 100000)}`;
    inpEle.type = 'file';
    inpEle.style.display = 'none';
    // 文件类型限制
    inpEle.accept = accept;
    // 多选限制
    multiple && (inpEle.multiple = multiple);
    // 选择目录限制
    if (webkitdirectory) {
        console.warn('该特性是非标准的，请尽量不要在生产环境中使用它！\n' +
            'This feature is non-standard, so try not to use it in a production environment!');
        inpEle.webkitdirectory = webkitdirectory;
    }
    inpEle.click();
    return await new Promise((resolve, reject) => {
        let _isSelected = false;
        const changeEvent = () => {
            const files = inpEle.files;
            if (files) {
                _isSelected = true;
                resolve(Array.from(files));
            }
        };
        const focusEvent = (event) => {
            if (event.target?.constructor === Window) {
                setTimeout(() => {
                    !_isSelected && reject('未选定文件\nUnselected file');
                }, cancel);
            }
        };
        inpEle.addEventListener('change', changeEvent, { once: true });
        cancel && window.addEventListener('focus', focusEvent, { once: true });
    });
}
function downloadFile(blob, downloadName) {
    const a = document.createElement('a');
    a.href = URL.createObjectURL(blob);
    // window.open(URL.createObjectURL(res.data));
    a.download = downloadName;
    a.click();
    URL.revokeObjectURL(a.href);
    a.remove();
}
// 枚举类型定义文件类型
exports.FILE_TYPES = Object.freeze({
    AUDIO: 'audio',
    VIDEO: 'video',
    OFFICE: 'office',
    IMAGE: 'image',
    MODEL: 'model',
    CAD: 'cad', // 图纸文件
    ARCHIVE: 'archive',
    UNKNOWN: 'unknown',
});
// 判断文件类型函数
function getFileType(fileName) {
    const audioExtensions = ['.mp3', '.wav', '.flac', '.aac', '.ogg'];
    const videoExtensions = ['.mp4', '.avi', '.mov', '.flv', '.mkv', '.webm'];
    const officeExtensions = [
        '.doc',
        '.docx',
        '.xls',
        '.xlsx',
        '.ppt',
        '.pptx',
        '.pdf',
    ];
    const imageExtensions = ['.jpg', '.jpeg', '.png', '.gif', '.bmp', '.webp'];
    const modelExtensions = ['.obj', '.fbx', '.glb', '.stl', '.ply']; // 常见模型文件格式
    const cadExtensions = ['.dwg', '.dxf', '.dwt', '.stp', '.step']; // 图纸文件格式
    const archiveExtensions = ['.zip', '.tar', '.rar', '.gz', '.7z'];
    // 获取文件扩展名（忽略大小写）
    const fileExtension = fileName
        .slice(((fileName.lastIndexOf('.') - 1) >>> 0) + 2)
        .toLowerCase();
    // 根据扩展名判断文件类型
    if (audioExtensions.includes(`.${fileExtension}`)) {
        return exports.FILE_TYPES.AUDIO;
    }
    else if (videoExtensions.includes(`.${fileExtension}`)) {
        return exports.FILE_TYPES.VIDEO;
    }
    else if (officeExtensions.includes(`.${fileExtension}`)) {
        return exports.FILE_TYPES.OFFICE;
    }
    else if (imageExtensions.includes(`.${fileExtension}`)) {
        return exports.FILE_TYPES.IMAGE;
    }
    else if (modelExtensions.includes(`.${fileExtension}`)) {
        return exports.FILE_TYPES.MODEL;
    }
    else if (cadExtensions.includes(`.${fileExtension}`)) {
        return exports.FILE_TYPES.CAD;
    }
    else if (archiveExtensions.includes(`.${fileExtension}`)) {
        return exports.FILE_TYPES.ARCHIVE;
    }
    else {
        return exports.FILE_TYPES.UNKNOWN;
    }
}
async function urlToBlob(url) {
    try {
        // 使用 fetch 获取文件内容
        const response = await fetch(url);
        // 检查响应是否成功
        if (!response.ok) {
            throw new Error('网络错误，无法获取文件');
        }
        // 将响应体转换为 Blob
        const blob = await response.blob();
        // 返回 Blob 对象
        return blob;
    }
    catch (error) {
        console.error('获取 URL 失败:', error);
    }
}
