// hfy-tools/dist/utils/is.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isClient = exports.isServer = void 0;
exports.is = is;
exports.isDef = isDef;
exports.isUnDef = isUnDef;
exports.isObject = isObject;
exports.isEmpty = isEmpty;
exports.isDate = isDate;
exports.isNull = isNull;
exports.isNullAndUnDef = isNullAndUnDef;
exports.isNullOrUnDef = isNullOrUnDef;
exports.isNumber = isNumber;
exports.isPromise = isPromise;
exports.isString = isString;
exports.isJsonObjectString = isJsonObjectString;
exports.isFunction = isFunction;
exports.isBoolean = isBoolean;
exports.isRegExp = isRegExp;
exports.isArray = isArray;
exports.isWindow = isWindow;
exports.isElement = isElement;
exports.isMap = isMap;
exports.isUrl = isUrl;
const toString = Object.prototype.toString;
function is(val, type) {
    return toString.call(val) === `[object ${type}]`;
}
function isDef(val) {
    return typeof val !== "undefined";
}
function isUnDef(val) {
    return !isDef(val);
}
function isObject(val) {
    return val !== null && is(val, "Object");
}
function isEmpty(val) {
    if (isArray(val) || isString(val)) {
        return val.length === 0;
    }
    if (val instanceof Map || val instanceof Set) {
        return val.size === 0;
    }
    if (isObject(val)) {
        return Object.keys(val).length === 0;
    }
    return false;
}
function isDate(val) {
    return is(val, "Date");
}
function isNull(val) {
    return val === null;
}
function isNullAndUnDef(val) {
    return isUnDef(val) && isNull(val);
}
function isNullOrUnDef(val) {
    return isUnDef(val) || isNull(val);
}
function isNumber(val) {
    return is(val, "Number");
}
function isPromise(val) {
    return is(val, "Promise") && isFunction(val.then) && isFunction(val.catch);
}
function isString(val) {
    return is(val, "String");
}
function isJsonObjectString(val) {
    if (!val) {
        return false;
    }
    return val.startsWith("{") && val.endsWith("}");
}
function isFunction(val) {
    return typeof val === "function";
}
function isBoolean(val) {
    return is(val, "Boolean");
}
function isRegExp(val) {
    return is(val, "RegExp");
}
function isArray(val) {
    return val && Array.isArray(val);
}
function isWindow(val) {
    return typeof window !== "undefined" && is(val, "Window");
}
function isElement(val) {
    return isObject(val) && !!val.tagName;
}
function isMap(val) {
    return is(val, "Map");
}
exports.isServer = typeof window === "undefined";
exports.isClient = !exports.isServer;
function isUrl(path) {
    const reg = /(((^https?:(?:\/\/)?)(?:[-;:&=\+\$,\w]+@)?[A-Za-z0-9.-]+(?::\d+)?|(?:www.|[-;:&=\+\$,\w]+@)[A-Za-z0-9.-]+)((?:\/[\+~%\/.\w-_]*)?\??(?:[-\+=&;%@.\w_]*)#?(?:[\w]*))?)$/;
    return reg.test(path);
}
