// hfy-tools/dist/utils/office.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.renderExcel = renderExcel;
exports.renderDoc = renderDoc;
exports.renderXMind = renderXMind;
const polaris_excel_viewer_1 = __importDefault(require("polaris-excel-viewer"));
const polaris_xmind_viewer_1 = require("polaris-xmind-viewer");
const polaris_docx_preview_1 = require("polaris-docx-preview");
function renderExcel(dom, url, opts) {
    if (typeof url !== 'string') {
        url = URL.createObjectURL(url);
    }
    return new polaris_excel_viewer_1.default(dom, url, {
        readonly: true,
        disabled: true,
        ...(opts || {}),
    });
}
function renderDoc(blob, dom) {
    return (0, polaris_docx_preview_1.renderAsync)(blob, dom);
}
function renderXMind(blob, dom, opts) {
    return new polaris_xmind_viewer_1.XMindEmbedViewer({
        el: dom,
        file: blob,
        ...(opts || {}),
    });
}
