// hfy-tools/dist/utils/tree.js
"use strict";
/**
 * 根据 id 查找节点
 * @param {Array} treeData 树数据
 * @param {string} id 要查找的节点 id
 * @param {string} key 要查找的节点 key
 * @returns {Object|null} 返回节点对象或 null
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.findNodeById = findNodeById;
exports.findParentPathById = findParentPathById;
exports.findParentObjectById = findParentObjectById;
exports.mapTree = mapTree;
exports.filterTree = filterTree;
exports.containsValueTree = containsValueTree;
exports.arrayToTree = arrayToTree;
exports.strongArrayToTree = strongArrayToTree;
function findNodeById(treeData, val, key = 'id', childName = 'children') {
    for (let node of treeData) {
        // @ts-ignore
        if (Object.is(node[key], val)) {
            return node;
        }
        if (node[childName]) {
            const foundNode = findNodeById(node[childName], val, key);
            if (foundNode) {
                return foundNode;
            }
        }
    }
    return null;
}
/**
 * 根据当前节点 id 返回所有父节点 id 路径
 * @param {Array} treeData 树数据
 * @param {string} targetId 目标节点的 id
 * @param {string} idField 节点id 字段
 * @param {string} parentField 父节点 字段
 * @returns {Array} 返回从目标节点到根节点的 id 路径
 */
function findParentPathById(treeData, targetId, idField = 'id', parentField = 'pid') {
    for (let node of treeData) {
        if (Object.is(node[idField], targetId)) {
            const path = [node[idField]]; // 初始化路径数组，包含当前节点的 id
            let parentId = node[parentField];
            // 逐级回溯父节点
            while (parentId !== null) {
                const parentNode = findNodeById(treeData, parentId, idField); // 查找父节点
                if (parentNode) {
                    path.unshift(parentNode[idField]); // 将父节点 id 加到路径前面
                    parentId = parentNode[parentField]; // 更新父节点 id
                }
                else {
                    break; // 如果找不到父节点，结束回溯
                }
            }
            return path; // 返回从目标节点到根节点的路径
        }
        // 如果有子节点，递归查找
        if (node.children) {
            const result = findParentPathById(node.children, targetId, idField, parentField);
            if (result) {
                return [node[idField], ...result]; // 拼接当前节点和子节点路径
            }
        }
    }
    // 如果没有找到目标节点，返回 null
    return null;
}
/**
 * 根据当前节点 id 返回所有父节点 对象
 * @param {Array} treeData 树数据
 * @param {string} targetId 目标节点的 id
 * @param {string} idField 节点id 字段
 * @param {string} parentField 父节点 字段
 * @returns {Array} 返回从目标节点到根节点的 id 路径
 */
function findParentObjectById(treeData, targetId, idField = 'id', parentField = 'pid') {
    for (let node of treeData) {
        if (Object.is(node[idField], targetId)) {
            const parents = [node]; // 初始化路径数组，包含当前节点的 id
            let parentId = node[parentField];
            // 逐级回溯父节点
            while (parentId !== null) {
                const parentNode = findNodeById(treeData, parentId, idField); // 查找父节点
                if (parentNode) {
                    parents.unshift(parentNode); // 将父节点 id 加到路径前面
                    parentId = parentNode[parentField]; // 更新父节点 id
                }
                else {
                    break; // 如果找不到父节点，结束回溯
                }
            }
            return parents; // 返回从目标节点到根节点的路径
        }
        // 如果有子节点，递归查找
        if (node.children) {
            const result = findParentObjectById(node.children, targetId, idField, parentField);
            return [node, ...result]; // 拼接当前节点和子节点路径
        }
    }
    // 如果没有找到目标节点
    return [];
}
function mapTree(treeData, callback) {
    return treeData?.map((o) => {
        const node = callback(o);
        if (node) {
            return { ...node, children: mapTree(o.children, callback) };
        }
        return node;
    });
}
function filterTree(treeData, callback) {
    return treeData.filter((o) => {
        if (o.children) {
            o.children = filterTree(o.children, callback);
        }
        return callback(o);
    });
}
function containsValueTree(tree, field, value = '') {
    return tree.some((o) => {
        if (Object.is(o[field], value)) {
            return true;
        }
        if (o.children) {
            return containsValueTree(o.children, field, value);
        }
        return false;
    });
}
// // 数组转树
function arrayToTree(array, keyField = 'id', parentField = 'pid', rootValue = 0) {
    return array
        .filter((o) => Object.is(o[parentField], rootValue))
        .map((o) => {
        return {
            ...o,
            children: arrayToTree(array, keyField, parentField, o[keyField]),
        };
    });
}
function strongArrayToTree(items, options = {}) {
    const { idKey = 'id', parentKey = 'parentId', childrenKey = 'children', rootValue = null, } = options;
    const nodeMap = new Map();
    const tree = [];
    items.forEach((item) => {
        nodeMap.set(item[idKey], {
            ...item,
            [childrenKey]: [],
        });
    });
    items.forEach((item) => {
        const node = nodeMap.get(item[idKey]);
        const parentId = item[parentKey];
        if (parentId === rootValue || !nodeMap.has(parentId)) {
            tree.push(node);
        }
        else {
            const parent = nodeMap.get(parentId);
            if (parent) {
                parent[childrenKey].push(node);
            }
        }
    });
    return tree;
}
