// hfy-tools/src/utils/document.ts
declare namespace navigator{

}

/**
 * url参数转对象
 * @param query
 * @returns
 */
export const getUrlParams = (query): Record<string, string> =>
  Array.from(new URLSearchParams(query)).reduce(
    (p, [k, v]) =>
      Object.assign({}, p, { [k]: p[k] ? (Array.isArray(p[k]) ? p[k] : [p[k]]).concat(v) : v }),
    {}
  )

/**
 * 移动端
 */
export const isMobile = /Android|webOS|iPhone|iPad|iPod|BlackBerry|IEMobile|Opera Mini/i.test(
  navigator?.userAgent
)
/**
 * 加载一个远程脚本
 * @param {String} src 一个远程脚本
 * @param {Function} callback 回调
 */
export const loadScript = (src, callback = (_) => {}): void => {
  const docTag = window.top.document
  if (docTag.querySelector(`[src='${src}']`) || docTag.getElementById(src)) {
    callback(true)
    return
  }

  const $script: HTMLScriptElement & { onreadystatechange?: Function } =
    document.createElement('script')

  $script.src = src
  $script.id = src
  $script.async = true
  $script.defer = true
  docTag.children[0].appendChild($script)
  if ('onload' in $script) {
    $script.onload = (_) => {
      callback(true)
    }

    $script.onerror = (_) => {
      callback(new Error(`Failed to load ${src}`))
    }
    return
  }

  // ie
  if ($script['onreadystatechange']) {
    Object.bind($script, 'onreadystatechange', function () {
      if (['loaded', 'complete'].includes(this.readyState)) {
        callback(true)
      } else {
        setTimeout(function () {
          arguments.callee()
        }, 500)
      }
    })
  }
}

/**
 * 获取URL GET参数
 * @param name
 * @returns
 */
export const getQueryString = (name): string => {
  var reg = new RegExp('(^|&)' + name + '=([^&]*)(&|$)')
  var r = window.location.search.substr(1).match(reg)
  if (r != null) return unescape(r[2])
  return ''
}

/**
 * 获取当前页面名称
 * @returns
 */
export const strPage = (): String => {
  var strUrl = window.location.href
  var arrUrl = strUrl.split('/')
  var strPage = arrUrl[arrUrl.length - 1]
  if (strPage.indexOf('?') > -1) {
    var pageName = strPage.split('?')
    strPage = pageName[0]
  }
  return strPage
}

/**
 * 操作系统
 * @returns
 */
export const getCurrentPlatform = (): Record<string, Function> => {
  return {
    Android: function () {
      //@ts-ignore
      return navigator?.userAgent?.match?.(/Android/i) ? true : false
    },
    BlackBerry: function () {
      //@ts-ignore
      return navigator?.userAgent?.match?.(/BlackBerry/i) ? true : false
    },
    iOS: function () {
      //@ts-ignore
      return navigator?.userAgent?.match?.(/iPhone|iPad|iPod/i) ? true : false
    },
    Windows: function () {
      //@ts-ignore
      return navigator?.userAgent?.match?.(/IEMobile/i) ? true : false
    },
    any: function () {
      return this.Android() || this.BlackBerry() || this.iOS() || this.Windows()
    }
  }
}

/**
 * 设置cookie
 * @param name
 * @param value
 */
export const setCookie = (name, value): void => {
  if (value) {
    var Days = 365
    var exp = new Date()
    exp.setTime(exp.getTime() + Days * 24 * 60 * 60 * 1000)
    document.cookie = name + '=' + escape(value) + ';expires=' + exp.toDateString()
  }
}

/**
 * 获取cookie
 * @param name
 * @returns
 */
export const getCookie = (name): string => {
  if (document.cookie.length > 0) {
    var begin = document.cookie.indexOf(name + '=')
    if (begin !== -1) {
      // cookie值的初始位置
      begin += name.length + 1
      // 结束位置
      var end = document.cookie.indexOf(';', begin)
      if (end === -1) {
        // 没有;则end为字符串结束位置
        end = document.cookie.length
      }
      return unescape(document.cookie.substring(begin, end))
    }
  }
  // cookie不存在返回
  return ''
}

/**
 * 清除所有cookie
 */
export const clearCookie = (): void => {
  var keys = document.cookie.match(/[^ =;]+(?=\=)/g)
  if (keys) {
    for (var i = keys.length; i--; ) {
      document.cookie = keys[i] + '=0;expires=' + new Date(0).toUTCString()
    }
  }
}
