// hfy-tools/src/utils/fetch.ts
/**
 * fetch get
 * @param url
 * @returns
 */
export async function get<T>(
  url: string,
  data?: Record<string, never>,
  headers?: Record<string, never>,
): Promise<T> {
  if (data) {
    if (url.indexOf('?') === -1) {
      url += '?'
    } else {
      url += '&'
    }
    url += new URLSearchParams(data).toString()
  }
  return await (
    await fetch(url, {
      headers,
      method: 'get',
    })
  ).json()
}

/**
 * fetch post
 * @param url
 * @returns
 */
export async function post<
  T,
  D extends Record<string, never> = Record<string, never>,
>(url: string, data?: D, headers?: Record<string, string>): Promise<T> {
  return await (
    await fetch(url, {
      method: 'post',
      body: data ? JSON.stringify(data) : undefined,
      headers: {
        'Content-Type': 'application/json',
        ...(headers || {}),
      },
    })
  ).json()
}

/**
 * fetch blob
 * @param url
 * @returns
 */
export async function postForm<
  T,
  D extends Record<string, never> = Record<string, never>,
>(url: string, data: D, headers?: Record<string, never>): Promise<T> {
  let body: FormData = new FormData()
  Object.keys(data).forEach((o) => {
    body.append(o, data[o])
  })

  return await (
    await fetch(url, {
      method: 'post',
      body,
      headers: {
        ...headers,
        'Content-Type': 'multipart/form-data',
      },
    })
  ).json()
}

export async function del<T>(
  url: string,
  headers?: Record<string, never>,
): Promise<T> {
  return await (
    await fetch(url, {
      method: 'delete',
      headers,
    })
  ).json()
}

export async function put<
  T,
  P extends Record<string, never> = Record<string, never>,
>(url: string, data?: P, headers?: Record<string, never>) {
  return await (
    await fetch(url, {
      method: 'put',
      body: data ? JSON.stringify(data) : undefined,
      headers: {
        'Content-Type': 'application/json',
        ...(headers || {}),
      },
    })
  ).json()
}
