// hfy-tools/src/utils/math.ts
import Decimal from 'decimal.js'

type ILonLat = number | string | any
/**
 * 保留几位小数不是四舍五入
 * @param value 数值
 * @param decimal  =保留几位小数
 * @returns 处理后的值
 */
export const formatDecimal = (value: ILonLat, decimal: number = 7) => {
  value = value.toString()
  const index = value.indexOf('.')
  if (index !== -1) {
    value = value.substring(0, decimal + index + 1)
  } else {
    value = value.substring(0)
  }
  return +parseFloat(value).toFixed(decimal)
}

// 加法 (0.1 + 0.2 = 0.3)
export function add(a, b) {
  return new Decimal(a).add(new Decimal(b)).toNumber()
}

// 减法 (1.0 - 0.7 = 0.3)
export function subtract(a, b) {
  return new Decimal(a).sub(new Decimal(b)).toNumber()
}

// 乘法 (0.6 * 3 = 1.8)
export function multiply(a, b) {
  return new Decimal(a).mul(new Decimal(b)).toNumber()
}

// 除法 (1.2 / 3 = 0.4)
export function divide(a, b) {
  return new Decimal(a).div(new Decimal(b)).toNumber()
}

// 四舍五入
export function round(num, precision = 2) {
  return new Decimal(num).toDecimalPlaces(precision).toNumber()
}

// 比较大小
export function compare(a, b) {
  return new Decimal(a).comparedTo(new Decimal(b))
}
