// hfy-tools/dist/component/web/Switch/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomSwitch = void 0;
class CustomSwitch extends HTMLElement {
    static observedAttributes = [
        'checked',
        'disabled',
        'size',
        'color',
        'label-on',
        'label-off',
    ];
    constructor() {
        super();
        this.attachShadow({ mode: 'open' });
        this.render();
    }
    connectedCallback() {
        this.addEventListener('click', this.toggle);
        this.addEventListener('keydown', this.handleKeydown);
        // 设置初始tabindex以支持键盘导航
        if (!this.hasAttribute('tabindex')) {
            this.setAttribute('tabindex', '0');
        }
    }
    disconnectedCallback() {
        this.removeEventListener('click', this.toggle);
        this.removeEventListener('keydown', this.handleKeydown);
    }
    attributeChangedCallback(name, oldValue, newValue) {
        if (oldValue !== newValue) {
            this.render();
        }
    }
    handleKeydown(event) {
        // 支持空格键和回车键切换开关
        if (event.key === ' ' || event.key === 'Enter') {
            event.preventDefault();
            this.toggle();
        }
    }
    toggle() {
        if (this.disabled)
            return;
        this.checked = !this.checked;
        // 触发自定义事件
        this.dispatchEvent(new CustomEvent('change', {
            detail: { checked: this.checked },
            bubbles: true,
        }));
        this.dispatchEvent(new CustomEvent('toggle', {
            detail: { checked: this.checked },
            bubbles: true,
        }));
    }
    get checked() {
        return this.hasAttribute('checked');
    }
    set checked(value) {
        if (value) {
            this.setAttribute('checked', '');
        }
        else {
            this.removeAttribute('checked');
        }
    }
    get disabled() {
        return this.hasAttribute('disabled');
    }
    get size() {
        return this.getAttribute('size') || 'medium';
    }
    get color() {
        return this.getAttribute('color') || '#4cd964';
    }
    get labelOn() {
        return this.getAttribute('label-on');
    }
    get labelOff() {
        return this.getAttribute('label-off');
    }
    getColorValue(color) {
        const colorMap = {
            blue: '#3498db',
            green: '#2ecc71',
            orange: '#e67e22',
            purple: '#9b59b6',
        };
        return colorMap[color] || color;
    }
    render() {
        const checked = this.checked;
        const disabled = this.disabled;
        const size = this.size;
        const color = this.getColorValue(this.color);
        const labelOn = this.labelOn;
        const labelOff = this.labelOff;
        const switchClass = [
            'switch',
            `switch-${size}`,
            checked ? 'switch-checked' : '',
            disabled ? 'switch-disabled' : '',
        ]
            .filter(Boolean)
            .join(' ');
        this.shadowRoot.innerHTML = `
                    <style>
                        :host {
                            display: inline-block;
                            cursor: pointer;
                            user-select: none;
                        }

                        :host([disabled]) {
                            cursor: not-allowed;
                            opacity: 0.6;
                        }

                        .switch {
                            position: relative;
                            display: inline-flex;
                            align-items: center;
                            transition: all 0.3s ease;
                        }

                        .switch-track {
                            position: relative;
                            background-color: #e0e0e0;
                            border-radius: 34px;
                            transition: all 0.3s ease;
                        }

                        .switch-small .switch-track {
                            width: 40px;
                            height: 20px;
                        }

                        .switch-medium .switch-track {
                            width: 52px;
                            height: 26px;
                        }

                        .switch-large .switch-track {
                            width: 64px;
                            height: 32px;
                        }

                        .switch-thumb {
                            position: absolute;
                            top: 2px;
                            background-color: white;
                            border-radius: 50%;
                            box-shadow: 0 2px 4px rgba(0, 0, 0, 0.2);
                            transition: all 0.3s ease;
                        }

                        .switch-small .switch-thumb {
                            width: 16px;
                            height: 16px;
                            left: 2px;
                        }

                        .switch-medium .switch-thumb {
                            width: 22px;
                            height: 22px;
                            left: 2px;
                        }

                        .switch-large .switch-thumb {
                            width: 28px;
                            height: 28px;
                            left: 2px;
                        }

                        .switch-checked .switch-track {
                            background-color: ${color};
                        }

                        .switch-small.switch-checked .switch-thumb {
                            transform: translateX(20px);
                        }

                        .switch-medium.switch-checked .switch-thumb {
                            transform: translateX(26px);
                        }

                        .switch-large.switch-checked .switch-thumb {
                            transform: translateX(32px);
                        }

                        .switch-label {
                            margin-left: 8px;
                            font-size: 14px;
                            font-weight: 500;
                            color: #666;
                            transition: color 0.3s ease;
                        }

                        .switch-checked .switch-label {
                            color: #333;
                        }

                        /* 聚焦样式 */
                        :host(:focus) {
                            outline: none;
                        }

                        :host(:focus) .switch-track {
                            box-shadow: 0 0 0 3px rgba(52, 152, 219, 0.3);
                        }

                        /* 动画效果 */
                        .switch-track::after {
                            content: '';
                            position: absolute;
                            top: 0;
                            left: 0;
                            width: 100%;
                            height: 100%;
                            border-radius: 34px;
                            background-color: inherit;
                            transition: opacity 0.3s ease;
                            opacity: 0;
                        }

                        :host(:active) .switch-track::after {
                            opacity: 0.2;
                        }
                    </style>

                    <div class="${switchClass}">
                        <div class="switch-track">
                            <div class="switch-thumb"></div>
                        </div>
                        ${labelOn || labelOff
            ? `
                            <div class="switch-label">
                                ${checked ? labelOn : labelOff}
                            </div>
                        `
            : ''}
                    </div>
                `;
    }
}
exports.CustomSwitch = CustomSwitch;
