// hfy-tools/dist/component/web/Button/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomButton = void 0;
class CustomButton extends HTMLElement {
    static observedAttributes = [
        'variant',
        'size',
        'disabled',
        'loading',
        'outline',
        'icon',
    ];
    constructor() {
        super();
        this.attachShadow({ mode: 'open' });
        this.render();
    }
    connectedCallback() {
        this.addEventListener('click', this.handleClick);
    }
    disconnectedCallback() {
        this.removeEventListener('click', this.handleClick);
    }
    attributeChangedCallback(name, oldValue, newValue) {
        if (oldValue !== newValue) {
            this.render();
        }
    }
    handleClick(event) {
        if (this.disabled || this.loading) {
            event.preventDefault();
            event.stopPropagation();
            return;
        }
        // 触发自定义事件
        this.dispatchEvent(new CustomEvent('button-click', {
            detail: {
                variant: this.variant,
                text: this.textContent,
            },
            bubbles: true,
        }));
    }
    get variant() {
        return this.getAttribute('variant') || 'primary';
    }
    get size() {
        return this.getAttribute('size') || 'medium';
    }
    get disabled() {
        return this.hasAttribute('disabled');
    }
    get loading() {
        return this.hasAttribute('loading');
    }
    get outline() {
        return this.hasAttribute('outline');
    }
    get icon() {
        return this.getAttribute('icon');
    }
    render() {
        const variant = this.variant;
        const size = this.size;
        const disabled = this.disabled;
        const loading = this.loading;
        const outline = this.outline;
        const icon = this.icon;
        const buttonClass = [
            'btn',
            `btn-${variant}`,
            `btn-${size}`,
            disabled ? 'btn-disabled' : '',
            loading ? 'btn-loading' : '',
            outline ? 'btn-outline' : '',
        ]
            .filter(Boolean)
            .join(' ');
        this.shadowRoot.innerHTML = `
                    <style>
                        :host {
                            display: inline-block;
                        }

                        .btn {
                            display: inline-flex;
                            align-items: center;
                            justify-content: center;
                            gap: 8px;
                            border: none;
                            border-radius: 6px;
                            font-family: inherit;
                            font-weight: 500;
                            cursor: pointer;
                            transition: all 0.2s ease;
                            position: relative;
                            overflow: hidden;
                            user-select: none;
                        }

                        .btn:focus {
                            outline: 2px solid rgba(52, 152, 219, 0.5);
                            outline-offset: 2px;
                        }

                        /* 尺寸 */
                        .btn-small {
                            padding: 6px 12px;
                            font-size: 14px;
                        }

                        .btn-medium {
                            padding: 10px 18px;
                            font-size: 16px;
                        }

                        .btn-large {
                            padding: 14px 24px;
                            font-size: 18px;
                        }

                        /* 变体 - 实心 */
                        .btn-primary {
                            background-color: #3498db;
                            color: white;
                        }

                        .btn-secondary {
                            background-color: #95a5a6;
                            color: white;
                        }

                        .btn-success {
                            background-color: #2ecc71;
                            color: white;
                        }

                        .btn-warning {
                            background-color: #f39c12;
                            color: white;
                        }

                        .btn-danger {
                            background-color: #e74c3c;
                            color: white;
                        }

                        /* 变体 - 轮廓 */
                        .btn-outline {
                            background-color: transparent;
                        }

                        .btn-outline.btn-primary {
                            color: #3498db;
                            border: 2px solid #3498db;
                        }

                        .btn-outline.btn-secondary {
                            color: #95a5a6;
                            border: 2px solid #95a5a6;
                        }

                        .btn-outline.btn-success {
                            color: #2ecc71;
                            border: 2px solid #2ecc71;
                        }

                        .btn-outline.btn-warning {
                            color: #f39c12;
                            border: 2px solid #f39c12;
                        }

                        .btn-outline.btn-danger {
                            color: #e74c3c;
                            border: 2px solid #e74c3c;
                        }

                        /* 悬停效果 */
                        .btn:not(.btn-disabled):not(.btn-loading):hover {
                            transform: translateY(-2px);
                            box-shadow: 0 4px 8px rgba(0, 0, 0, 0.1);
                        }

                        .btn-primary:not(.btn-disabled):not(.btn-loading):hover {
                            background-color: #2980b9;
                        }

                        .btn-secondary:not(.btn-disabled):not(.btn-loading):hover {
                            background-color: #7f8c8d;
                        }

                        .btn-success:not(.btn-disabled):not(.btn-loading):hover {
                            background-color: #27ae60;
                        }

                        .btn-warning:not(.btn-disabled):not(.btn-loading):hover {
                            background-color: #e67e22;
                        }

                        .btn-danger:not(.btn-disabled):not(.btn-loading):hover {
                            background-color: #c0392b;
                        }

                        /* 轮廓按钮悬停 */
                        .btn-outline.btn-primary:not(.btn-disabled):not(.btn-loading):hover {
                            background-color: #3498db;
                            color: white;
                        }

                        .btn-outline.btn-secondary:not(.btn-disabled):not(.btn-loading):hover {
                            background-color: #95a5a6;
                            color: white;
                        }

                        .btn-outline.btn-success:not(.btn-disabled):not(.btn-loading):hover {
                            background-color: #2ecc71;
                            color: white;
                        }

                        .btn-outline.btn-warning:not(.btn-disabled):not(.btn-loading):hover {
                            background-color: #f39c12;
                            color: white;
                        }

                        .btn-outline.btn-danger:not(.btn-disabled):not(.btn-loading):hover {
                            background-color: #e74c3c;
                            color: white;
                        }

                        /* 禁用状态 */
                        .btn-disabled {
                            opacity: 0.6;
                            cursor: not-allowed;
                            transform: none !important;
                            box-shadow: none !important;
                        }

                        /* 加载状态 */
                        .btn-loading {
                            pointer-events: none;
                            color: transparent;
                        }

                        .btn-loading::after {
                            content: "";
                            position: absolute;
                            width: 16px;
                            height: 16px;
                            top: 0;
                            left: 0;
                            right: 0;
                            bottom: 0;
                            margin: auto;
                            border: 2px solid transparent;
                            border-top-color: currentColor;
                            border-radius: 50%;
                            animation: button-loading-spinner 1s linear infinite;
                        }

                        @keyframes button-loading-spinner {
                            from {
                                transform: rotate(0turn);
                            }

                            to {
                                transform: rotate(1turn);
                            }
                        }
                    </style>

                    <button class="${buttonClass}" ${disabled ? 'disabled' : ''}>
                        ${icon ? `<span class="btn-icon">${icon}</span>` : ''}
                        <span class="btn-text"><slot></slot></span>
                    </button>
                `;
    }
}
exports.CustomButton = CustomButton;
