// hfy-tools/src/directive/empty.js
function replaceDOM(EmptyDOM, targetDOM, text, image) {
	// 没有数据的情况
	const { parentNode } = targetDOM;
	const box = document.createElement('div');

	const targetBox = new EmptyDOM({
		el: box,
		props: {
			desc: {
				default: () => {
					return text;
				},
			},
			image: {
				default: () => {
					return image;
				},
			},
		},
	});
	parentNode.replaceChild(targetBox.$el, targetDOM);
}

// 使用方式 v-empty={hasData:（数据列表||布尔）,text:'为空时的提示',image:图片}
export default {
	install(Vue, options) {
		const { Empty: emptyDOM } = options;
		Vue.directive('empty', {
			inserted(el, bind) {
				const { hasData } = bind.value;
				const text = bind.value.text || '暂无更多数据';
				// let image = bind.value.image || Empty.props.image.default;
				const image = '';
				const EmptyDOM = Vue.extend(emptyDOM);
				if (typeof hasData === 'undefined' || hasData === null) {
					// 非空
					replaceDOM(EmptyDOM, el, text, image);
					return;
				}

				if (typeof hasData === 'object' && Array.isArray(hasData) && hasData.length <= 0) {
					// 数组
					replaceDOM(EmptyDOM, el, text, image);
					return;
				}

				if (typeof hasData === 'boolean' && !hasData) {
					// 布尔 false 替换
					replaceDOM(EmptyDOM, el, text, image);
				}
			},
		});
	},
};
