// polaris-docx-preview@0.3.2/dist/docx-preview.mjs
import JSZip from 'jszip';

/******************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */
/* global Reflect, Promise, SuppressedError, Symbol */

var extendStatics = function(d, b) {
    extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
    return extendStatics(d, b);
};

function __extends(d, b) {
    if (typeof b !== "function" && b !== null)
        throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
    extendStatics(d, b);
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
}

var __assign = function() {
    __assign = Object.assign || function __assign(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};

function __awaiter(thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
}

function __generator(thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
}

function __spreadArray(to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
}

typeof SuppressedError === "function" ? SuppressedError : function (error, suppressed, message) {
    var e = new Error(message);
    return e.name = "SuppressedError", e.error = error, e.suppressed = suppressed, e;
};

var RelationshipTypes;
(function (RelationshipTypes) {
    RelationshipTypes["OfficeDocument"] = "http://schemas.openxmlformats.org/officeDocument/2006/relationships/officeDocument";
    RelationshipTypes["FontTable"] = "http://schemas.openxmlformats.org/officeDocument/2006/relationships/fontTable";
    RelationshipTypes["Image"] = "http://schemas.openxmlformats.org/officeDocument/2006/relationships/image";
    RelationshipTypes["Numbering"] = "http://schemas.openxmlformats.org/officeDocument/2006/relationships/numbering";
    RelationshipTypes["Styles"] = "http://schemas.openxmlformats.org/officeDocument/2006/relationships/styles";
    RelationshipTypes["StylesWithEffects"] = "http://schemas.microsoft.com/office/2007/relationships/stylesWithEffects";
    RelationshipTypes["Theme"] = "http://schemas.openxmlformats.org/officeDocument/2006/relationships/theme";
    RelationshipTypes["Settings"] = "http://schemas.openxmlformats.org/officeDocument/2006/relationships/settings";
    RelationshipTypes["WebSettings"] = "http://schemas.openxmlformats.org/officeDocument/2006/relationships/webSettings";
    RelationshipTypes["Hyperlink"] = "http://schemas.openxmlformats.org/officeDocument/2006/relationships/hyperlink";
    RelationshipTypes["Footnotes"] = "http://schemas.openxmlformats.org/officeDocument/2006/relationships/footnotes";
    RelationshipTypes["Endnotes"] = "http://schemas.openxmlformats.org/officeDocument/2006/relationships/endnotes";
    RelationshipTypes["Footer"] = "http://schemas.openxmlformats.org/officeDocument/2006/relationships/footer";
    RelationshipTypes["Header"] = "http://schemas.openxmlformats.org/officeDocument/2006/relationships/header";
    RelationshipTypes["ExtendedProperties"] = "http://schemas.openxmlformats.org/officeDocument/2006/relationships/extended-properties";
    RelationshipTypes["CoreProperties"] = "http://schemas.openxmlformats.org/package/2006/relationships/metadata/core-properties";
    RelationshipTypes["CustomProperties"] = "http://schemas.openxmlformats.org/package/2006/relationships/metadata/custom-properties";
    RelationshipTypes["Comments"] = "http://schemas.openxmlformats.org/officeDocument/2006/relationships/comments";
    RelationshipTypes["CommentsExtended"] = "http://schemas.microsoft.com/office/2011/relationships/commentsExtended";
})(RelationshipTypes || (RelationshipTypes = {}));
function parseRelationships(root, xml) {
    return xml.elements(root).map(function (e) { return ({
        id: xml.attr(e, "Id"),
        type: xml.attr(e, "Type"),
        target: xml.attr(e, "Target"),
        targetMode: xml.attr(e, "TargetMode")
    }); });
}

var ns$1 = {
    wordml: "http://schemas.openxmlformats.org/wordprocessingml/2006/main",
    drawingml: "http://schemas.openxmlformats.org/drawingml/2006/main",
    picture: "http://schemas.openxmlformats.org/drawingml/2006/picture",
    compatibility: "http://schemas.openxmlformats.org/markup-compatibility/2006",
    math: "http://schemas.openxmlformats.org/officeDocument/2006/math"
};
var LengthUsage = {
    Dxa: { mul: 0.05, unit: "pt" },
    Emu: { mul: 1 / 12700, unit: "pt" },
    FontSize: { mul: 0.5, unit: "pt" },
    Border: { mul: 0.125, unit: "pt" },
    Point: { mul: 1, unit: "pt" },
    Percent: { mul: 0.02, unit: "%" },
    LineHeight: { mul: 1 / 240, unit: "" },
    VmlEmu: { mul: 1 / 12700, unit: "" },
};
function convertLength(val, usage) {
    if (usage === void 0) { usage = LengthUsage.Dxa; }
    if (val == null || /.+(p[xt]|[%])$/.test(val)) {
        return val;
    }
    return "".concat((parseInt(val) * usage.mul).toFixed(2)).concat(usage.unit);
}
function convertBoolean(v, defaultValue) {
    if (defaultValue === void 0) { defaultValue = false; }
    switch (v) {
        case "1": return true;
        case "0": return false;
        case "on": return true;
        case "off": return false;
        case "true": return true;
        case "false": return false;
        default: return defaultValue;
    }
}
function parseCommonProperty(elem, props, xml) {
    if (elem.namespaceURI != ns$1.wordml)
        return false;
    switch (elem.localName) {
        case "color":
            props.color = xml.attr(elem, "val");
            break;
        case "sz":
            props.fontSize = xml.lengthAttr(elem, "val", LengthUsage.FontSize);
            break;
        default:
            return false;
    }
    return true;
}

function parseXmlString(xmlString, trimXmlDeclaration) {
    if (trimXmlDeclaration === void 0) { trimXmlDeclaration = false; }
    if (trimXmlDeclaration)
        xmlString = xmlString.replace(/<[?].*[?]>/, "");
    xmlString = removeUTF8BOM(xmlString);
    var result = new DOMParser().parseFromString(xmlString, "application/xml");
    var errorText = hasXmlParserError(result);
    if (errorText)
        throw new Error(errorText);
    return result;
}
function hasXmlParserError(doc) {
    var _a;
    return (_a = doc.getElementsByTagName("parsererror")[0]) === null || _a === void 0 ? void 0 : _a.textContent;
}
function removeUTF8BOM(data) {
    return data.charCodeAt(0) === 0xFEFF ? data.substring(1) : data;
}
function serializeXmlString(elem) {
    return new XMLSerializer().serializeToString(elem);
}
var XmlParser = (function () {
    function XmlParser() {
    }
    XmlParser.prototype.elements = function (elem, localName) {
        if (localName === void 0) { localName = null; }
        var result = [];
        for (var i = 0, l = elem.childNodes.length; i < l; i++) {
            var c = elem.childNodes.item(i);
            if (c.nodeType == 1 && (localName == null || c.localName == localName))
                result.push(c);
        }
        return result;
    };
    XmlParser.prototype.element = function (elem, localName) {
        for (var i = 0, l = elem.childNodes.length; i < l; i++) {
            var c = elem.childNodes.item(i);
            if (c.nodeType == 1 && c.localName == localName)
                return c;
        }
        return null;
    };
    XmlParser.prototype.elementAttr = function (elem, localName, attrLocalName) {
        var el = this.element(elem, localName);
        return el ? this.attr(el, attrLocalName) : undefined;
    };
    XmlParser.prototype.attrs = function (elem) {
        return Array.from(elem.attributes);
    };
    XmlParser.prototype.attr = function (elem, localName) {
        for (var i = 0, l = elem.attributes.length; i < l; i++) {
            var a = elem.attributes.item(i);
            if (a.localName == localName)
                return a.value;
        }
        return null;
    };
    XmlParser.prototype.intAttr = function (node, attrName, defaultValue) {
        if (defaultValue === void 0) { defaultValue = null; }
        var val = this.attr(node, attrName);
        return val ? parseInt(val) : defaultValue;
    };
    XmlParser.prototype.hexAttr = function (node, attrName, defaultValue) {
        if (defaultValue === void 0) { defaultValue = null; }
        var val = this.attr(node, attrName);
        return val ? parseInt(val, 16) : defaultValue;
    };
    XmlParser.prototype.floatAttr = function (node, attrName, defaultValue) {
        if (defaultValue === void 0) { defaultValue = null; }
        var val = this.attr(node, attrName);
        return val ? parseFloat(val) : defaultValue;
    };
    XmlParser.prototype.boolAttr = function (node, attrName, defaultValue) {
        if (defaultValue === void 0) { defaultValue = null; }
        return convertBoolean(this.attr(node, attrName), defaultValue);
    };
    XmlParser.prototype.lengthAttr = function (node, attrName, usage) {
        if (usage === void 0) { usage = LengthUsage.Dxa; }
        return convertLength(this.attr(node, attrName), usage);
    };
    return XmlParser;
}());
var globalXmlParser = new XmlParser();

var Part = (function () {
    function Part(_package, path) {
        this._package = _package;
        this.path = path;
    }
    Part.prototype.load = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _a, xmlText, xmlDoc;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _a = this;
                        return [4, this._package.loadRelationships(this.path)];
                    case 1:
                        _a.rels = _b.sent();
                        return [4, this._package.load(this.path)];
                    case 2:
                        xmlText = _b.sent();
                        xmlDoc = this._package.parseXmlDocument(xmlText);
                        if (this._package.options.keepOrigin) {
                            this._xmlDocument = xmlDoc;
                        }
                        this.parseXml(xmlDoc.firstElementChild);
                        return [2];
                }
            });
        });
    };
    Part.prototype.save = function () {
        this._package.update(this.path, serializeXmlString(this._xmlDocument));
    };
    Part.prototype.parseXml = function (root) {
    };
    return Part;
}());

var embedFontTypeMap = {
    embedRegular: 'regular',
    embedBold: 'bold',
    embedItalic: 'italic',
    embedBoldItalic: 'boldItalic',
};
function parseFonts(root, xml) {
    return xml.elements(root).map(function (el) { return parseFont(el, xml); });
}
function parseFont(elem, xml) {
    var result = {
        name: xml.attr(elem, "name"),
        embedFontRefs: []
    };
    for (var _i = 0, _a = xml.elements(elem); _i < _a.length; _i++) {
        var el = _a[_i];
        switch (el.localName) {
            case "family":
                result.family = xml.attr(el, "val");
                break;
            case "altName":
                result.altName = xml.attr(el, "val");
                break;
            case "embedRegular":
            case "embedBold":
            case "embedItalic":
            case "embedBoldItalic":
                result.embedFontRefs.push(parseEmbedFontRef(el, xml));
                break;
        }
    }
    return result;
}
function parseEmbedFontRef(elem, xml) {
    return {
        id: xml.attr(elem, "id"),
        key: xml.attr(elem, "fontKey"),
        type: embedFontTypeMap[elem.localName]
    };
}

var FontTablePart = (function (_super) {
    __extends(FontTablePart, _super);
    function FontTablePart() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    FontTablePart.prototype.parseXml = function (root) {
        this.fonts = parseFonts(root, this._package.xmlParser);
    };
    return FontTablePart;
}(Part));

function escapeClassName(className) {
    return className === null || className === void 0 ? void 0 : className.replace(/[ .]+/g, '-').replace(/[&]+/g, 'and').toLowerCase();
}
function splitPath(path) {
    var si = path.lastIndexOf('/') + 1;
    var folder = si == 0 ? "" : path.substring(0, si);
    var fileName = si == 0 ? path : path.substring(si);
    return [folder, fileName];
}
function resolvePath(path, base) {
    try {
        var prefix = "http://docx/";
        var url = new URL(path, prefix + base).toString();
        return url.substring(prefix.length);
    }
    catch (_a) {
        return "".concat(base).concat(path);
    }
}
function keyBy(array, by) {
    return array.reduce(function (a, x) {
        a[by(x)] = x;
        return a;
    }, {});
}
function blobToBase64(blob) {
    return new Promise(function (resolve, reject) {
        var reader = new FileReader();
        reader.onloadend = function () { return resolve(reader.result); };
        reader.onerror = function () { return reject(); };
        reader.readAsDataURL(blob);
    });
}
function isObject(item) {
    return item && typeof item === 'object' && !Array.isArray(item);
}
function isString(item) {
    return typeof item === 'string' || item instanceof String;
}
function mergeDeep(target) {
    var _a;
    var sources = [];
    for (var _i = 1; _i < arguments.length; _i++) {
        sources[_i - 1] = arguments[_i];
    }
    if (!sources.length)
        return target;
    var source = sources.shift();
    if (isObject(target) && isObject(source)) {
        for (var key in source) {
            if (isObject(source[key])) {
                var val = (_a = target[key]) !== null && _a !== void 0 ? _a : (target[key] = {});
                mergeDeep(val, source[key]);
            }
            else {
                target[key] = source[key];
            }
        }
    }
    return mergeDeep.apply(void 0, __spreadArray([target], sources, false));
}
function asArray(val) {
    return Array.isArray(val) ? val : [val];
}

var OpenXmlPackage = (function () {
    function OpenXmlPackage(_zip, options) {
        this._zip = _zip;
        this.options = options;
        this.xmlParser = new XmlParser();
    }
    OpenXmlPackage.prototype.get = function (path) {
        var _a;
        var p = normalizePath(path);
        return (_a = this._zip.files[p]) !== null && _a !== void 0 ? _a : this._zip.files[p.replace(/\//g, '\\')];
    };
    OpenXmlPackage.prototype.update = function (path, content) {
        this._zip.file(path, content);
    };
    OpenXmlPackage.load = function (input, options) {
        return __awaiter(this, void 0, void 0, function () {
            var zip;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, JSZip.loadAsync(input)];
                    case 1:
                        zip = _a.sent();
                        return [2, new OpenXmlPackage(zip, options)];
                }
            });
        });
    };
    OpenXmlPackage.prototype.save = function (type) {
        if (type === void 0) { type = "blob"; }
        return this._zip.generateAsync({ type: type });
    };
    OpenXmlPackage.prototype.load = function (path, type) {
        var _a, _b;
        if (type === void 0) { type = "string"; }
        return (_b = (_a = this.get(path)) === null || _a === void 0 ? void 0 : _a.async(type)) !== null && _b !== void 0 ? _b : Promise.resolve(null);
    };
    OpenXmlPackage.prototype.loadRelationships = function () {
        return __awaiter(this, arguments, void 0, function (path) {
            var relsPath, _a, f, fn, txt;
            if (path === void 0) { path = null; }
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        relsPath = "_rels/.rels";
                        if (path != null) {
                            _a = splitPath(path), f = _a[0], fn = _a[1];
                            relsPath = "".concat(f, "_rels/").concat(fn, ".rels");
                        }
                        return [4, this.load(relsPath)];
                    case 1:
                        txt = _b.sent();
                        return [2, txt ? parseRelationships(this.parseXmlDocument(txt).firstElementChild, this.xmlParser) : null];
                }
            });
        });
    };
    OpenXmlPackage.prototype.parseXmlDocument = function (txt) {
        return parseXmlString(txt, this.options.trimXmlDeclaration);
    };
    return OpenXmlPackage;
}());
function normalizePath(path) {
    return path.startsWith('/') ? path.substr(1) : path;
}

var DocumentPart = (function (_super) {
    __extends(DocumentPart, _super);
    function DocumentPart(pkg, path, parser) {
        var _this = _super.call(this, pkg, path) || this;
        _this._documentParser = parser;
        return _this;
    }
    DocumentPart.prototype.parseXml = function (root) {
        this.body = this._documentParser.parseDocumentFile(root);
    };
    return DocumentPart;
}(Part));

function parseBorder(elem, xml) {
    return {
        type: xml.attr(elem, "val"),
        color: xml.attr(elem, "color"),
        size: xml.lengthAttr(elem, "sz", LengthUsage.Border),
        offset: xml.lengthAttr(elem, "space", LengthUsage.Point),
        frame: xml.boolAttr(elem, 'frame'),
        shadow: xml.boolAttr(elem, 'shadow')
    };
}
function parseBorders(elem, xml) {
    var result = {};
    for (var _i = 0, _a = xml.elements(elem); _i < _a.length; _i++) {
        var e = _a[_i];
        switch (e.localName) {
            case "left":
                result.left = parseBorder(e, xml);
                break;
            case "top":
                result.top = parseBorder(e, xml);
                break;
            case "right":
                result.right = parseBorder(e, xml);
                break;
            case "bottom":
                result.bottom = parseBorder(e, xml);
                break;
        }
    }
    return result;
}

var SectionType;
(function (SectionType) {
    SectionType["Continuous"] = "continuous";
    SectionType["NextPage"] = "nextPage";
    SectionType["NextColumn"] = "nextColumn";
    SectionType["EvenPage"] = "evenPage";
    SectionType["OddPage"] = "oddPage";
})(SectionType || (SectionType = {}));
function parseSectionProperties(elem, xml) {
    var _a, _b;
    if (xml === void 0) { xml = globalXmlParser; }
    var section = {};
    for (var _i = 0, _c = xml.elements(elem); _i < _c.length; _i++) {
        var e = _c[_i];
        switch (e.localName) {
            case "pgSz":
                section.pageSize = {
                    width: xml.lengthAttr(e, "w"),
                    height: xml.lengthAttr(e, "h"),
                    orientation: xml.attr(e, "orient")
                };
                break;
            case "type":
                section.type = xml.attr(e, "val");
                break;
            case "pgMar":
                section.pageMargins = {
                    left: xml.lengthAttr(e, "left"),
                    right: xml.lengthAttr(e, "right"),
                    top: xml.lengthAttr(e, "top"),
                    bottom: xml.lengthAttr(e, "bottom"),
                    header: xml.lengthAttr(e, "header"),
                    footer: xml.lengthAttr(e, "footer"),
                    gutter: xml.lengthAttr(e, "gutter"),
                };
                break;
            case "cols":
                section.columns = parseColumns(e, xml);
                break;
            case "headerReference":
                ((_a = section.headerRefs) !== null && _a !== void 0 ? _a : (section.headerRefs = [])).push(parseFooterHeaderReference(e, xml));
                break;
            case "footerReference":
                ((_b = section.footerRefs) !== null && _b !== void 0 ? _b : (section.footerRefs = [])).push(parseFooterHeaderReference(e, xml));
                break;
            case "titlePg":
                section.titlePage = xml.boolAttr(e, "val", true);
                break;
            case "pgBorders":
                section.pageBorders = parseBorders(e, xml);
                break;
            case "pgNumType":
                section.pageNumber = parsePageNumber(e, xml);
                break;
        }
    }
    return section;
}
function parseColumns(elem, xml) {
    return {
        numberOfColumns: xml.intAttr(elem, "num"),
        space: xml.lengthAttr(elem, "space"),
        separator: xml.boolAttr(elem, "sep"),
        equalWidth: xml.boolAttr(elem, "equalWidth", true),
        columns: xml.elements(elem, "col")
            .map(function (e) { return ({
            width: xml.lengthAttr(e, "w"),
            space: xml.lengthAttr(e, "space")
        }); })
    };
}
function parsePageNumber(elem, xml) {
    return {
        chapSep: xml.attr(elem, "chapSep"),
        chapStyle: xml.attr(elem, "chapStyle"),
        format: xml.attr(elem, "fmt"),
        start: xml.intAttr(elem, "start")
    };
}
function parseFooterHeaderReference(elem, xml) {
    return {
        id: xml.attr(elem, "id"),
        type: xml.attr(elem, "type"),
    };
}

function parseLineSpacing(elem, xml) {
    return {
        before: xml.lengthAttr(elem, "before"),
        after: xml.lengthAttr(elem, "after"),
        line: xml.intAttr(elem, "line"),
        lineRule: xml.attr(elem, "lineRule")
    };
}

function parseRunProperties(elem, xml) {
    var result = {};
    for (var _i = 0, _a = xml.elements(elem); _i < _a.length; _i++) {
        var el = _a[_i];
        parseRunProperty(el, result, xml);
    }
    return result;
}
function parseRunProperty(elem, props, xml) {
    if (parseCommonProperty(elem, props, xml))
        return true;
    return false;
}

function parseParagraphProperties(elem, xml) {
    var result = {};
    for (var _i = 0, _a = xml.elements(elem); _i < _a.length; _i++) {
        var el = _a[_i];
        parseParagraphProperty(el, result, xml);
    }
    return result;
}
function parseParagraphProperty(elem, props, xml) {
    if (elem.namespaceURI != ns$1.wordml)
        return false;
    if (parseCommonProperty(elem, props, xml))
        return true;
    switch (elem.localName) {
        case "tabs":
            props.tabs = parseTabs(elem, xml);
            break;
        case "sectPr":
            props.sectionProps = parseSectionProperties(elem, xml);
            break;
        case "numPr":
            props.numbering = parseNumbering$1(elem, xml);
            break;
        case "spacing":
            props.lineSpacing = parseLineSpacing(elem, xml);
            return false;
        case "textAlignment":
            props.textAlignment = xml.attr(elem, "val");
            return false;
        case "keepLines":
            props.keepLines = xml.boolAttr(elem, "val", true);
            break;
        case "keepNext":
            props.keepNext = xml.boolAttr(elem, "val", true);
            break;
        case "pageBreakBefore":
            props.pageBreakBefore = xml.boolAttr(elem, "val", true);
            break;
        case "outlineLvl":
            props.outlineLevel = xml.intAttr(elem, "val");
            break;
        case "pStyle":
            props.styleName = xml.attr(elem, "val");
            break;
        case "rPr":
            props.runProps = parseRunProperties(elem, xml);
            break;
        default:
            return false;
    }
    return true;
}
function parseTabs(elem, xml) {
    return xml.elements(elem, "tab")
        .map(function (e) { return ({
        position: xml.lengthAttr(e, "pos"),
        leader: xml.attr(e, "leader"),
        style: xml.attr(e, "val")
    }); });
}
function parseNumbering$1(elem, xml) {
    var result = {};
    for (var _i = 0, _a = xml.elements(elem); _i < _a.length; _i++) {
        var e = _a[_i];
        switch (e.localName) {
            case "numId":
                result.id = xml.attr(e, "val");
                break;
            case "ilvl":
                result.level = xml.intAttr(e, "val");
                break;
        }
    }
    return result;
}

function parseNumberingPart(elem, xml) {
    var result = {
        numberings: [],
        abstractNumberings: [],
        bulletPictures: []
    };
    for (var _i = 0, _a = xml.elements(elem); _i < _a.length; _i++) {
        var e = _a[_i];
        switch (e.localName) {
            case "num":
                result.numberings.push(parseNumbering(e, xml));
                break;
            case "abstractNum":
                result.abstractNumberings.push(parseAbstractNumbering(e, xml));
                break;
            case "numPicBullet":
                result.bulletPictures.push(parseNumberingBulletPicture(e, xml));
                break;
        }
    }
    return result;
}
function parseNumbering(elem, xml) {
    var result = {
        id: xml.attr(elem, 'numId'),
        overrides: []
    };
    for (var _i = 0, _a = xml.elements(elem); _i < _a.length; _i++) {
        var e = _a[_i];
        switch (e.localName) {
            case "abstractNumId":
                result.abstractId = xml.attr(e, "val");
                break;
            case "lvlOverride":
                result.overrides.push(parseNumberingLevelOverrride(e, xml));
                break;
        }
    }
    return result;
}
function parseAbstractNumbering(elem, xml) {
    var result = {
        id: xml.attr(elem, 'abstractNumId'),
        levels: []
    };
    for (var _i = 0, _a = xml.elements(elem); _i < _a.length; _i++) {
        var e = _a[_i];
        switch (e.localName) {
            case "name":
                result.name = xml.attr(e, "val");
                break;
            case "multiLevelType":
                result.multiLevelType = xml.attr(e, "val");
                break;
            case "numStyleLink":
                result.numberingStyleLink = xml.attr(e, "val");
                break;
            case "styleLink":
                result.styleLink = xml.attr(e, "val");
                break;
            case "lvl":
                result.levels.push(parseNumberingLevel(e, xml));
                break;
        }
    }
    return result;
}
function parseNumberingLevel(elem, xml) {
    var result = {
        level: xml.intAttr(elem, 'ilvl')
    };
    for (var _i = 0, _a = xml.elements(elem); _i < _a.length; _i++) {
        var e = _a[_i];
        switch (e.localName) {
            case "start":
                result.start = xml.attr(e, "val");
                break;
            case "lvlRestart":
                result.restart = xml.intAttr(e, "val");
                break;
            case "numFmt":
                result.format = xml.attr(e, "val");
                break;
            case "lvlText":
                result.text = xml.attr(e, "val");
                break;
            case "lvlJc":
                result.justification = xml.attr(e, "val");
                break;
            case "lvlPicBulletId":
                result.bulletPictureId = xml.attr(e, "val");
                break;
            case "pStyle":
                result.paragraphStyle = xml.attr(e, "val");
                break;
            case "pPr":
                result.paragraphProps = parseParagraphProperties(e, xml);
                break;
            case "rPr":
                result.runProps = parseRunProperties(e, xml);
                break;
        }
    }
    return result;
}
function parseNumberingLevelOverrride(elem, xml) {
    var result = {
        level: xml.intAttr(elem, 'ilvl')
    };
    for (var _i = 0, _a = xml.elements(elem); _i < _a.length; _i++) {
        var e = _a[_i];
        switch (e.localName) {
            case "startOverride":
                result.start = xml.intAttr(e, "val");
                break;
            case "lvl":
                result.numberingLevel = parseNumberingLevel(e, xml);
                break;
        }
    }
    return result;
}
function parseNumberingBulletPicture(elem, xml) {
    var pict = xml.element(elem, "pict");
    var shape = pict && xml.element(pict, "shape");
    var imagedata = shape && xml.element(shape, "imagedata");
    return imagedata ? {
        id: xml.attr(elem, "numPicBulletId"),
        referenceId: xml.attr(imagedata, "id"),
        style: xml.attr(shape, "style")
    } : null;
}

var NumberingPart = (function (_super) {
    __extends(NumberingPart, _super);
    function NumberingPart(pkg, path, parser) {
        var _this = _super.call(this, pkg, path) || this;
        _this._documentParser = parser;
        return _this;
    }
    NumberingPart.prototype.parseXml = function (root) {
        Object.assign(this, parseNumberingPart(root, this._package.xmlParser));
        this.domNumberings = this._documentParser.parseNumberingFile(root);
    };
    return NumberingPart;
}(Part));

var StylesPart = (function (_super) {
    __extends(StylesPart, _super);
    function StylesPart(pkg, path, parser) {
        var _this = _super.call(this, pkg, path) || this;
        _this._documentParser = parser;
        return _this;
    }
    StylesPart.prototype.parseXml = function (root) {
        this.styles = this._documentParser.parseStylesFile(root);
    };
    return StylesPart;
}(Part));

var DomType;
(function (DomType) {
    DomType["Document"] = "document";
    DomType["Paragraph"] = "paragraph";
    DomType["Run"] = "run";
    DomType["Break"] = "break";
    DomType["NoBreakHyphen"] = "noBreakHyphen";
    DomType["Table"] = "table";
    DomType["Row"] = "row";
    DomType["Cell"] = "cell";
    DomType["Hyperlink"] = "hyperlink";
    DomType["Drawing"] = "drawing";
    DomType["Image"] = "image";
    DomType["Text"] = "text";
    DomType["Tab"] = "tab";
    DomType["Symbol"] = "symbol";
    DomType["BookmarkStart"] = "bookmarkStart";
    DomType["BookmarkEnd"] = "bookmarkEnd";
    DomType["Footer"] = "footer";
    DomType["Header"] = "header";
    DomType["FootnoteReference"] = "footnoteReference";
    DomType["EndnoteReference"] = "endnoteReference";
    DomType["Footnote"] = "footnote";
    DomType["Endnote"] = "endnote";
    DomType["SimpleField"] = "simpleField";
    DomType["ComplexField"] = "complexField";
    DomType["Instruction"] = "instruction";
    DomType["VmlPicture"] = "vmlPicture";
    DomType["MmlMath"] = "mmlMath";
    DomType["MmlMathParagraph"] = "mmlMathParagraph";
    DomType["MmlFraction"] = "mmlFraction";
    DomType["MmlFunction"] = "mmlFunction";
    DomType["MmlFunctionName"] = "mmlFunctionName";
    DomType["MmlNumerator"] = "mmlNumerator";
    DomType["MmlDenominator"] = "mmlDenominator";
    DomType["MmlRadical"] = "mmlRadical";
    DomType["MmlBase"] = "mmlBase";
    DomType["MmlDegree"] = "mmlDegree";
    DomType["MmlSuperscript"] = "mmlSuperscript";
    DomType["MmlSubscript"] = "mmlSubscript";
    DomType["MmlPreSubSuper"] = "mmlPreSubSuper";
    DomType["MmlSubArgument"] = "mmlSubArgument";
    DomType["MmlSuperArgument"] = "mmlSuperArgument";
    DomType["MmlNary"] = "mmlNary";
    DomType["MmlDelimiter"] = "mmlDelimiter";
    DomType["MmlRun"] = "mmlRun";
    DomType["MmlEquationArray"] = "mmlEquationArray";
    DomType["MmlLimit"] = "mmlLimit";
    DomType["MmlLimitLower"] = "mmlLimitLower";
    DomType["MmlMatrix"] = "mmlMatrix";
    DomType["MmlMatrixRow"] = "mmlMatrixRow";
    DomType["MmlBox"] = "mmlBox";
    DomType["MmlBar"] = "mmlBar";
    DomType["MmlGroupChar"] = "mmlGroupChar";
    DomType["VmlElement"] = "vmlElement";
    DomType["Inserted"] = "inserted";
    DomType["Deleted"] = "deleted";
    DomType["DeletedText"] = "deletedText";
    DomType["Comment"] = "comment";
    DomType["CommentReference"] = "commentReference";
    DomType["CommentRangeStart"] = "commentRangeStart";
    DomType["CommentRangeEnd"] = "commentRangeEnd";
})(DomType || (DomType = {}));
var OpenXmlElementBase = (function () {
    function OpenXmlElementBase() {
        this.children = [];
        this.cssStyle = {};
    }
    return OpenXmlElementBase;
}());

var WmlHeader = (function (_super) {
    __extends(WmlHeader, _super);
    function WmlHeader() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.type = DomType.Header;
        return _this;
    }
    return WmlHeader;
}(OpenXmlElementBase));
var WmlFooter = (function (_super) {
    __extends(WmlFooter, _super);
    function WmlFooter() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.type = DomType.Footer;
        return _this;
    }
    return WmlFooter;
}(OpenXmlElementBase));

var BaseHeaderFooterPart = (function (_super) {
    __extends(BaseHeaderFooterPart, _super);
    function BaseHeaderFooterPart(pkg, path, parser) {
        var _this = _super.call(this, pkg, path) || this;
        _this._documentParser = parser;
        return _this;
    }
    BaseHeaderFooterPart.prototype.parseXml = function (root) {
        this.rootElement = this.createRootElement();
        this.rootElement.children = this._documentParser.parseBodyElements(root);
    };
    return BaseHeaderFooterPart;
}(Part));
var HeaderPart = (function (_super) {
    __extends(HeaderPart, _super);
    function HeaderPart() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    HeaderPart.prototype.createRootElement = function () {
        return new WmlHeader();
    };
    return HeaderPart;
}(BaseHeaderFooterPart));
var FooterPart = (function (_super) {
    __extends(FooterPart, _super);
    function FooterPart() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    FooterPart.prototype.createRootElement = function () {
        return new WmlFooter();
    };
    return FooterPart;
}(BaseHeaderFooterPart));

function parseExtendedProps(root, xmlParser) {
    var result = {};
    for (var _i = 0, _a = xmlParser.elements(root); _i < _a.length; _i++) {
        var el = _a[_i];
        switch (el.localName) {
            case "Template":
                result.template = el.textContent;
                break;
            case "Pages":
                result.pages = safeParseToInt(el.textContent);
                break;
            case "Words":
                result.words = safeParseToInt(el.textContent);
                break;
            case "Characters":
                result.characters = safeParseToInt(el.textContent);
                break;
            case "Application":
                result.application = el.textContent;
                break;
            case "Lines":
                result.lines = safeParseToInt(el.textContent);
                break;
            case "Paragraphs":
                result.paragraphs = safeParseToInt(el.textContent);
                break;
            case "Company":
                result.company = el.textContent;
                break;
            case "AppVersion":
                result.appVersion = el.textContent;
                break;
        }
    }
    return result;
}
function safeParseToInt(value) {
    if (typeof value === 'undefined')
        return;
    return parseInt(value);
}

var ExtendedPropsPart = (function (_super) {
    __extends(ExtendedPropsPart, _super);
    function ExtendedPropsPart() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ExtendedPropsPart.prototype.parseXml = function (root) {
        this.props = parseExtendedProps(root, this._package.xmlParser);
    };
    return ExtendedPropsPart;
}(Part));

function parseCoreProps(root, xmlParser) {
    var result = {};
    for (var _i = 0, _a = xmlParser.elements(root); _i < _a.length; _i++) {
        var el = _a[_i];
        switch (el.localName) {
            case "title":
                result.title = el.textContent;
                break;
            case "description":
                result.description = el.textContent;
                break;
            case "subject":
                result.subject = el.textContent;
                break;
            case "creator":
                result.creator = el.textContent;
                break;
            case "keywords":
                result.keywords = el.textContent;
                break;
            case "language":
                result.language = el.textContent;
                break;
            case "lastModifiedBy":
                result.lastModifiedBy = el.textContent;
                break;
            case "revision":
                el.textContent && (result.revision = parseInt(el.textContent));
                break;
        }
    }
    return result;
}

var CorePropsPart = (function (_super) {
    __extends(CorePropsPart, _super);
    function CorePropsPart() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    CorePropsPart.prototype.parseXml = function (root) {
        this.props = parseCoreProps(root, this._package.xmlParser);
    };
    return CorePropsPart;
}(Part));

var DmlTheme = (function () {
    function DmlTheme() {
    }
    return DmlTheme;
}());
function parseTheme(elem, xml) {
    var result = new DmlTheme();
    var themeElements = xml.element(elem, "themeElements");
    for (var _i = 0, _a = xml.elements(themeElements); _i < _a.length; _i++) {
        var el = _a[_i];
        switch (el.localName) {
            case "clrScheme":
                result.colorScheme = parseColorScheme(el, xml);
                break;
            case "fontScheme":
                result.fontScheme = parseFontScheme(el, xml);
                break;
        }
    }
    return result;
}
function parseColorScheme(elem, xml) {
    var result = {
        name: xml.attr(elem, "name"),
        colors: {}
    };
    for (var _i = 0, _a = xml.elements(elem); _i < _a.length; _i++) {
        var el = _a[_i];
        var srgbClr = xml.element(el, "srgbClr");
        var sysClr = xml.element(el, "sysClr");
        if (srgbClr) {
            result.colors[el.localName] = xml.attr(srgbClr, "val");
        }
        else if (sysClr) {
            result.colors[el.localName] = xml.attr(sysClr, "lastClr");
        }
    }
    return result;
}
function parseFontScheme(elem, xml) {
    var result = {
        name: xml.attr(elem, "name"),
    };
    for (var _i = 0, _a = xml.elements(elem); _i < _a.length; _i++) {
        var el = _a[_i];
        switch (el.localName) {
            case "majorFont":
                result.majorFont = parseFontInfo(el, xml);
                break;
            case "minorFont":
                result.minorFont = parseFontInfo(el, xml);
                break;
        }
    }
    return result;
}
function parseFontInfo(elem, xml) {
    return {
        latinTypeface: xml.elementAttr(elem, "latin", "typeface"),
        eaTypeface: xml.elementAttr(elem, "ea", "typeface"),
        csTypeface: xml.elementAttr(elem, "cs", "typeface"),
    };
}

var ThemePart = (function (_super) {
    __extends(ThemePart, _super);
    function ThemePart(pkg, path) {
        return _super.call(this, pkg, path) || this;
    }
    ThemePart.prototype.parseXml = function (root) {
        this.theme = parseTheme(root, this._package.xmlParser);
    };
    return ThemePart;
}(Part));

var WmlBaseNote = (function () {
    function WmlBaseNote() {
    }
    return WmlBaseNote;
}());
var WmlFootnote = (function (_super) {
    __extends(WmlFootnote, _super);
    function WmlFootnote() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.type = DomType.Footnote;
        return _this;
    }
    return WmlFootnote;
}(WmlBaseNote));
var WmlEndnote = (function (_super) {
    __extends(WmlEndnote, _super);
    function WmlEndnote() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.type = DomType.Endnote;
        return _this;
    }
    return WmlEndnote;
}(WmlBaseNote));

var BaseNotePart = (function (_super) {
    __extends(BaseNotePart, _super);
    function BaseNotePart(pkg, path, parser) {
        var _this = _super.call(this, pkg, path) || this;
        _this._documentParser = parser;
        return _this;
    }
    return BaseNotePart;
}(Part));
var FootnotesPart = (function (_super) {
    __extends(FootnotesPart, _super);
    function FootnotesPart(pkg, path, parser) {
        return _super.call(this, pkg, path, parser) || this;
    }
    FootnotesPart.prototype.parseXml = function (root) {
        this.notes = this._documentParser.parseNotes(root, "footnote", WmlFootnote);
    };
    return FootnotesPart;
}(BaseNotePart));
var EndnotesPart = (function (_super) {
    __extends(EndnotesPart, _super);
    function EndnotesPart(pkg, path, parser) {
        return _super.call(this, pkg, path, parser) || this;
    }
    EndnotesPart.prototype.parseXml = function (root) {
        this.notes = this._documentParser.parseNotes(root, "endnote", WmlEndnote);
    };
    return EndnotesPart;
}(BaseNotePart));

function parseSettings(elem, xml) {
    var result = {};
    for (var _i = 0, _a = xml.elements(elem); _i < _a.length; _i++) {
        var el = _a[_i];
        switch (el.localName) {
            case "defaultTabStop":
                result.defaultTabStop = xml.lengthAttr(el, "val");
                break;
            case "footnotePr":
                result.footnoteProps = parseNoteProperties(el, xml);
                break;
            case "endnotePr":
                result.endnoteProps = parseNoteProperties(el, xml);
                break;
            case "autoHyphenation":
                result.autoHyphenation = xml.boolAttr(el, "val");
                break;
        }
    }
    return result;
}
function parseNoteProperties(elem, xml) {
    var result = {
        defaultNoteIds: []
    };
    for (var _i = 0, _a = xml.elements(elem); _i < _a.length; _i++) {
        var el = _a[_i];
        switch (el.localName) {
            case "numFmt":
                result.nummeringFormat = xml.attr(el, "val");
                break;
            case "footnote":
            case "endnote":
                result.defaultNoteIds.push(xml.attr(el, "id"));
                break;
        }
    }
    return result;
}

var SettingsPart = (function (_super) {
    __extends(SettingsPart, _super);
    function SettingsPart(pkg, path) {
        return _super.call(this, pkg, path) || this;
    }
    SettingsPart.prototype.parseXml = function (root) {
        this.settings = parseSettings(root, this._package.xmlParser);
    };
    return SettingsPart;
}(Part));

function parseCustomProps(root, xml) {
    return xml.elements(root, "property").map(function (e) {
        var firstChild = e.firstChild;
        return {
            formatId: xml.attr(e, "fmtid"),
            name: xml.attr(e, "name"),
            type: firstChild.nodeName,
            value: firstChild.textContent
        };
    });
}

var CustomPropsPart = (function (_super) {
    __extends(CustomPropsPart, _super);
    function CustomPropsPart() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    CustomPropsPart.prototype.parseXml = function (root) {
        this.props = parseCustomProps(root, this._package.xmlParser);
    };
    return CustomPropsPart;
}(Part));

var CommentsPart = (function (_super) {
    __extends(CommentsPart, _super);
    function CommentsPart(pkg, path, parser) {
        var _this = _super.call(this, pkg, path) || this;
        _this._documentParser = parser;
        return _this;
    }
    CommentsPart.prototype.parseXml = function (root) {
        this.comments = this._documentParser.parseComments(root);
        this.commentMap = keyBy(this.comments, function (x) { return x.id; });
    };
    return CommentsPart;
}(Part));

var CommentsExtendedPart = (function (_super) {
    __extends(CommentsExtendedPart, _super);
    function CommentsExtendedPart(pkg, path) {
        var _this = _super.call(this, pkg, path) || this;
        _this.comments = [];
        return _this;
    }
    CommentsExtendedPart.prototype.parseXml = function (root) {
        var xml = this._package.xmlParser;
        for (var _i = 0, _a = xml.elements(root, "commentEx"); _i < _a.length; _i++) {
            var el = _a[_i];
            this.comments.push({
                paraId: xml.attr(el, 'paraId'),
                paraIdParent: xml.attr(el, 'paraIdParent'),
                done: xml.boolAttr(el, 'done')
            });
        }
        this.commentMap = keyBy(this.comments, function (x) { return x.paraId; });
    };
    return CommentsExtendedPart;
}(Part));

var topLevelRels = [
    { type: RelationshipTypes.OfficeDocument, target: "word/document.xml" },
    { type: RelationshipTypes.ExtendedProperties, target: "docProps/app.xml" },
    { type: RelationshipTypes.CoreProperties, target: "docProps/core.xml" },
    { type: RelationshipTypes.CustomProperties, target: "docProps/custom.xml" },
];
var WordDocument = (function () {
    function WordDocument() {
        this.parts = [];
        this.partsMap = {};
    }
    WordDocument.load = function (blob, parser, options) {
        return __awaiter(this, void 0, void 0, function () {
            var d, _a, _b;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        d = new WordDocument();
                        d._options = options;
                        d._parser = parser;
                        _a = d;
                        return [4, OpenXmlPackage.load(blob, options)];
                    case 1:
                        _a._package = _c.sent();
                        _b = d;
                        return [4, d._package.loadRelationships()];
                    case 2:
                        _b.rels = _c.sent();
                        return [4, Promise.all(topLevelRels.map(function (rel) {
                                var _a;
                                var r = (_a = d.rels.find(function (x) { return x.type === rel.type; })) !== null && _a !== void 0 ? _a : rel;
                                return d.loadRelationshipPart(r.target, r.type);
                            }))];
                    case 3:
                        _c.sent();
                        return [2, d];
                }
            });
        });
    };
    WordDocument.prototype.save = function (type) {
        if (type === void 0) { type = "blob"; }
        return this._package.save(type);
    };
    WordDocument.prototype.loadRelationshipPart = function (path, type) {
        return __awaiter(this, void 0, void 0, function () {
            var part, folder_1;
            var _this = this;
            var _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (this.partsMap[path])
                            return [2, this.partsMap[path]];
                        if (!this._package.get(path))
                            return [2, null];
                        part = null;
                        switch (type) {
                            case RelationshipTypes.OfficeDocument:
                                this.documentPart = part = new DocumentPart(this._package, path, this._parser);
                                break;
                            case RelationshipTypes.FontTable:
                                this.fontTablePart = part = new FontTablePart(this._package, path);
                                break;
                            case RelationshipTypes.Numbering:
                                this.numberingPart = part = new NumberingPart(this._package, path, this._parser);
                                break;
                            case RelationshipTypes.Styles:
                                this.stylesPart = part = new StylesPart(this._package, path, this._parser);
                                break;
                            case RelationshipTypes.Theme:
                                this.themePart = part = new ThemePart(this._package, path);
                                break;
                            case RelationshipTypes.Footnotes:
                                this.footnotesPart = part = new FootnotesPart(this._package, path, this._parser);
                                break;
                            case RelationshipTypes.Endnotes:
                                this.endnotesPart = part = new EndnotesPart(this._package, path, this._parser);
                                break;
                            case RelationshipTypes.Footer:
                                part = new FooterPart(this._package, path, this._parser);
                                break;
                            case RelationshipTypes.Header:
                                part = new HeaderPart(this._package, path, this._parser);
                                break;
                            case RelationshipTypes.CoreProperties:
                                this.corePropsPart = part = new CorePropsPart(this._package, path);
                                break;
                            case RelationshipTypes.ExtendedProperties:
                                this.extendedPropsPart = part = new ExtendedPropsPart(this._package, path);
                                break;
                            case RelationshipTypes.CustomProperties:
                                part = new CustomPropsPart(this._package, path);
                                break;
                            case RelationshipTypes.Settings:
                                this.settingsPart = part = new SettingsPart(this._package, path);
                                break;
                            case RelationshipTypes.Comments:
                                this.commentsPart = part = new CommentsPart(this._package, path, this._parser);
                                break;
                            case RelationshipTypes.CommentsExtended:
                                this.commentsExtendedPart = part = new CommentsExtendedPart(this._package, path);
                                break;
                        }
                        if (part == null)
                            return [2, Promise.resolve(null)];
                        this.partsMap[path] = part;
                        this.parts.push(part);
                        return [4, part.load()];
                    case 1:
                        _b.sent();
                        if (!(((_a = part.rels) === null || _a === void 0 ? void 0 : _a.length) > 0)) return [3, 3];
                        folder_1 = splitPath(part.path)[0];
                        return [4, Promise.all(part.rels.map(function (rel) { return _this.loadRelationshipPart(resolvePath(rel.target, folder_1), rel.type); }))];
                    case 2:
                        _b.sent();
                        _b.label = 3;
                    case 3: return [2, part];
                }
            });
        });
    };
    WordDocument.prototype.loadDocumentImage = function (id, part) {
        return __awaiter(this, void 0, void 0, function () {
            var x;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.loadResource(part !== null && part !== void 0 ? part : this.documentPart, id, "blob")];
                    case 1:
                        x = _a.sent();
                        return [2, this.blobToURL(x)];
                }
            });
        });
    };
    WordDocument.prototype.loadNumberingImage = function (id) {
        return __awaiter(this, void 0, void 0, function () {
            var x;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.loadResource(this.numberingPart, id, "blob")];
                    case 1:
                        x = _a.sent();
                        return [2, this.blobToURL(x)];
                }
            });
        });
    };
    WordDocument.prototype.loadFont = function (id, key) {
        return __awaiter(this, void 0, void 0, function () {
            var x;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4, this.loadResource(this.fontTablePart, id, "uint8array")];
                    case 1:
                        x = _a.sent();
                        return [2, x ? this.blobToURL(new Blob([deobfuscate(x, key)])) : x];
                }
            });
        });
    };
    WordDocument.prototype.blobToURL = function (blob) {
        if (!blob)
            return null;
        if (this._options.useBase64URL) {
            return blobToBase64(blob);
        }
        return URL.createObjectURL(blob);
    };
    WordDocument.prototype.findPartByRelId = function (id, basePart) {
        var _a;
        if (basePart === void 0) { basePart = null; }
        var rel = ((_a = basePart.rels) !== null && _a !== void 0 ? _a : this.rels).find(function (r) { return r.id == id; });
        var folder = basePart ? splitPath(basePart.path)[0] : '';
        return rel ? this.partsMap[resolvePath(rel.target, folder)] : null;
    };
    WordDocument.prototype.getPathById = function (part, id) {
        var rel = part.rels.find(function (x) { return x.id == id; });
        var folder = splitPath(part.path)[0];
        return rel ? resolvePath(rel.target, folder) : null;
    };
    WordDocument.prototype.loadResource = function (part, id, outputType) {
        var path = this.getPathById(part, id);
        return path ? this._package.load(path, outputType) : Promise.resolve(null);
    };
    return WordDocument;
}());
function deobfuscate(data, guidKey) {
    var len = 16;
    var trimmed = guidKey.replace(/{|}|-/g, "");
    var numbers = new Array(len);
    for (var i = 0; i < len; i++)
        numbers[len - i - 1] = parseInt(trimmed.substr(i * 2, 2), 16);
    for (var i = 0; i < 32; i++)
        data[i] = data[i] ^ numbers[i % len];
    return data;
}

function parseBookmarkStart(elem, xml) {
    return {
        type: DomType.BookmarkStart,
        id: xml.attr(elem, "id"),
        name: xml.attr(elem, "name"),
        colFirst: xml.intAttr(elem, "colFirst"),
        colLast: xml.intAttr(elem, "colLast")
    };
}
function parseBookmarkEnd(elem, xml) {
    return {
        type: DomType.BookmarkEnd,
        id: xml.attr(elem, "id")
    };
}

var VmlElement = (function (_super) {
    __extends(VmlElement, _super);
    function VmlElement() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.type = DomType.VmlElement;
        _this.attrs = {};
        return _this;
    }
    return VmlElement;
}(OpenXmlElementBase));
function parseVmlElement(elem, parser) {
    var _a;
    var result = new VmlElement();
    switch (elem.localName) {
        case "rect":
            result.tagName = "rect";
            Object.assign(result.attrs, { width: '100%', height: '100%' });
            break;
        case "oval":
            result.tagName = "ellipse";
            Object.assign(result.attrs, { cx: "50%", cy: "50%", rx: "50%", ry: "50%" });
            break;
        case "line":
            result.tagName = "line";
            break;
        case "shape":
            result.tagName = "g";
            break;
        case "textbox":
            result.tagName = "foreignObject";
            Object.assign(result.attrs, { width: '100%', height: '100%' });
            break;
        default:
            return null;
    }
    for (var _i = 0, _b = globalXmlParser.attrs(elem); _i < _b.length; _i++) {
        var at = _b[_i];
        switch (at.localName) {
            case "style":
                result.cssStyleText = at.value;
                break;
            case "fillcolor":
                result.attrs.fill = at.value;
                break;
            case "from":
                var _c = parsePoint(at.value), x1 = _c[0], y1 = _c[1];
                Object.assign(result.attrs, { x1: x1, y1: y1 });
                break;
            case "to":
                var _d = parsePoint(at.value), x2 = _d[0], y2 = _d[1];
                Object.assign(result.attrs, { x2: x2, y2: y2 });
                break;
        }
    }
    for (var _e = 0, _f = globalXmlParser.elements(elem); _e < _f.length; _e++) {
        var el = _f[_e];
        switch (el.localName) {
            case "stroke":
                Object.assign(result.attrs, parseStroke(el));
                break;
            case "fill":
                Object.assign(result.attrs, parseFill());
                break;
            case "imagedata":
                result.tagName = "image";
                Object.assign(result.attrs, { width: '100%', height: '100%' });
                result.imageHref = {
                    id: globalXmlParser.attr(el, "id"),
                    title: globalXmlParser.attr(el, "title"),
                };
                break;
            case "txbxContent":
                (_a = result.children).push.apply(_a, parser.parseBodyElements(el));
                break;
            default:
                var child = parseVmlElement(el, parser);
                child && result.children.push(child);
                break;
        }
    }
    return result;
}
function parseStroke(el) {
    var _a;
    return {
        'stroke': globalXmlParser.attr(el, "color"),
        'stroke-width': (_a = globalXmlParser.lengthAttr(el, "weight", LengthUsage.Emu)) !== null && _a !== void 0 ? _a : '1px'
    };
}
function parseFill(el) {
    return {};
}
function parsePoint(val) {
    return val.split(",");
}

var WmlComment = (function (_super) {
    __extends(WmlComment, _super);
    function WmlComment() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.type = DomType.Comment;
        return _this;
    }
    return WmlComment;
}(OpenXmlElementBase));
var WmlCommentReference = (function (_super) {
    __extends(WmlCommentReference, _super);
    function WmlCommentReference(id) {
        var _this = _super.call(this) || this;
        _this.id = id;
        _this.type = DomType.CommentReference;
        return _this;
    }
    return WmlCommentReference;
}(OpenXmlElementBase));
var WmlCommentRangeStart = (function (_super) {
    __extends(WmlCommentRangeStart, _super);
    function WmlCommentRangeStart(id) {
        var _this = _super.call(this) || this;
        _this.id = id;
        _this.type = DomType.CommentRangeStart;
        return _this;
    }
    return WmlCommentRangeStart;
}(OpenXmlElementBase));
var WmlCommentRangeEnd = (function (_super) {
    __extends(WmlCommentRangeEnd, _super);
    function WmlCommentRangeEnd(id) {
        var _this = _super.call(this) || this;
        _this.id = id;
        _this.type = DomType.CommentRangeEnd;
        return _this;
    }
    return WmlCommentRangeEnd;
}(OpenXmlElementBase));

var autos = {
    shd: "inherit",
    color: "black",
    borderColor: "black",
    highlight: "transparent"
};
var supportedNamespaceURIs = [];
var mmlTagMap = {
    "oMath": DomType.MmlMath,
    "oMathPara": DomType.MmlMathParagraph,
    "f": DomType.MmlFraction,
    "func": DomType.MmlFunction,
    "fName": DomType.MmlFunctionName,
    "num": DomType.MmlNumerator,
    "den": DomType.MmlDenominator,
    "rad": DomType.MmlRadical,
    "deg": DomType.MmlDegree,
    "e": DomType.MmlBase,
    "sSup": DomType.MmlSuperscript,
    "sSub": DomType.MmlSubscript,
    "sPre": DomType.MmlPreSubSuper,
    "sup": DomType.MmlSuperArgument,
    "sub": DomType.MmlSubArgument,
    "d": DomType.MmlDelimiter,
    "nary": DomType.MmlNary,
    "eqArr": DomType.MmlEquationArray,
    "lim": DomType.MmlLimit,
    "limLow": DomType.MmlLimitLower,
    "m": DomType.MmlMatrix,
    "mr": DomType.MmlMatrixRow,
    "box": DomType.MmlBox,
    "bar": DomType.MmlBar,
    "groupChr": DomType.MmlGroupChar
};
var DocumentParser = (function () {
    function DocumentParser(options) {
        this.options = __assign({ ignoreWidth: false, debug: false }, options);
    }
    DocumentParser.prototype.parseNotes = function (xmlDoc, elemName, elemClass) {
        var result = [];
        for (var _i = 0, _a = globalXmlParser.elements(xmlDoc, elemName); _i < _a.length; _i++) {
            var el = _a[_i];
            var node = new elemClass();
            node.id = globalXmlParser.attr(el, "id");
            node.noteType = globalXmlParser.attr(el, "type");
            node.children = this.parseBodyElements(el);
            result.push(node);
        }
        return result;
    };
    DocumentParser.prototype.parseComments = function (xmlDoc) {
        var result = [];
        for (var _i = 0, _a = globalXmlParser.elements(xmlDoc, "comment"); _i < _a.length; _i++) {
            var el = _a[_i];
            var item = new WmlComment();
            item.id = globalXmlParser.attr(el, "id");
            item.author = globalXmlParser.attr(el, "author");
            item.initials = globalXmlParser.attr(el, "initials");
            item.date = globalXmlParser.attr(el, "date");
            item.children = this.parseBodyElements(el);
            result.push(item);
        }
        return result;
    };
    DocumentParser.prototype.parseDocumentFile = function (xmlDoc) {
        var xbody = globalXmlParser.element(xmlDoc, "body");
        var background = globalXmlParser.element(xmlDoc, "background");
        var sectPr = globalXmlParser.element(xbody, "sectPr");
        return {
            type: DomType.Document,
            children: this.parseBodyElements(xbody),
            props: sectPr ? parseSectionProperties(sectPr, globalXmlParser) : {},
            cssStyle: background ? this.parseBackground(background) : {},
        };
    };
    DocumentParser.prototype.parseBackground = function (elem) {
        var result = {};
        var color = xmlUtil.colorAttr(elem, "color");
        if (color) {
            result["background-color"] = color;
        }
        return result;
    };
    DocumentParser.prototype.parseBodyElements = function (element) {
        var _this = this;
        var children = [];
        for (var _i = 0, _a = globalXmlParser.elements(element); _i < _a.length; _i++) {
            var elem = _a[_i];
            switch (elem.localName) {
                case "p":
                    children.push(this.parseParagraph(elem));
                    break;
                case "tbl":
                    children.push(this.parseTable(elem));
                    break;
                case "sdt":
                    children.push.apply(children, this.parseSdt(elem, function (e) { return _this.parseBodyElements(e); }));
                    break;
            }
        }
        return children;
    };
    DocumentParser.prototype.parseStylesFile = function (xstyles) {
        var _this = this;
        var result = [];
        xmlUtil.foreach(xstyles, function (n) {
            switch (n.localName) {
                case "style":
                    result.push(_this.parseStyle(n));
                    break;
                case "docDefaults":
                    result.push(_this.parseDefaultStyles(n));
                    break;
            }
        });
        return result;
    };
    DocumentParser.prototype.parseDefaultStyles = function (node) {
        var _this = this;
        var result = {
            id: null,
            name: null,
            target: null,
            basedOn: null,
            styles: []
        };
        xmlUtil.foreach(node, function (c) {
            switch (c.localName) {
                case "rPrDefault":
                    var rPr = globalXmlParser.element(c, "rPr");
                    if (rPr)
                        result.styles.push({
                            target: "span",
                            values: _this.parseDefaultProperties(rPr, {})
                        });
                    break;
                case "pPrDefault":
                    var pPr = globalXmlParser.element(c, "pPr");
                    if (pPr)
                        result.styles.push({
                            target: "p",
                            values: _this.parseDefaultProperties(pPr, {})
                        });
                    break;
            }
        });
        return result;
    };
    DocumentParser.prototype.parseStyle = function (node) {
        var _this = this;
        var result = {
            id: globalXmlParser.attr(node, "styleId"),
            isDefault: globalXmlParser.boolAttr(node, "default"),
            name: null,
            target: null,
            basedOn: null,
            styles: [],
            linked: null
        };
        switch (globalXmlParser.attr(node, "type")) {
            case "paragraph":
                result.target = "p";
                break;
            case "table":
                result.target = "table";
                break;
            case "character":
                result.target = "span";
                break;
        }
        xmlUtil.foreach(node, function (n) {
            switch (n.localName) {
                case "basedOn":
                    result.basedOn = globalXmlParser.attr(n, "val");
                    break;
                case "name":
                    result.name = globalXmlParser.attr(n, "val");
                    break;
                case "link":
                    result.linked = globalXmlParser.attr(n, "val");
                    break;
                case "next":
                    result.next = globalXmlParser.attr(n, "val");
                    break;
                case "aliases":
                    result.aliases = globalXmlParser.attr(n, "val").split(",");
                    break;
                case "pPr":
                    result.styles.push({
                        target: "p",
                        values: _this.parseDefaultProperties(n, {})
                    });
                    result.paragraphProps = parseParagraphProperties(n, globalXmlParser);
                    break;
                case "rPr":
                    result.styles.push({
                        target: "span",
                        values: _this.parseDefaultProperties(n, {})
                    });
                    result.runProps = parseRunProperties(n, globalXmlParser);
                    break;
                case "tblPr":
                case "tcPr":
                    result.styles.push({
                        target: "td",
                        values: _this.parseDefaultProperties(n, {})
                    });
                    break;
                case "tblStylePr":
                    for (var _i = 0, _a = _this.parseTableStyle(n); _i < _a.length; _i++) {
                        var s = _a[_i];
                        result.styles.push(s);
                    }
                    break;
                case "rsid":
                case "qFormat":
                case "hidden":
                case "semiHidden":
                case "unhideWhenUsed":
                case "autoRedefine":
                case "uiPriority":
                    break;
                default:
                    _this.options.debug && console.warn("DOCX: Unknown style element: ".concat(n.localName));
            }
        });
        return result;
    };
    DocumentParser.prototype.parseTableStyle = function (node) {
        var _this = this;
        var result = [];
        var type = globalXmlParser.attr(node, "type");
        var selector = "";
        var modificator = "";
        switch (type) {
            case "firstRow":
                modificator = ".first-row";
                selector = "tr.first-row td";
                break;
            case "lastRow":
                modificator = ".last-row";
                selector = "tr.last-row td";
                break;
            case "firstCol":
                modificator = ".first-col";
                selector = "td.first-col";
                break;
            case "lastCol":
                modificator = ".last-col";
                selector = "td.last-col";
                break;
            case "band1Vert":
                modificator = ":not(.no-vband)";
                selector = "td.odd-col";
                break;
            case "band2Vert":
                modificator = ":not(.no-vband)";
                selector = "td.even-col";
                break;
            case "band1Horz":
                modificator = ":not(.no-hband)";
                selector = "tr.odd-row";
                break;
            case "band2Horz":
                modificator = ":not(.no-hband)";
                selector = "tr.even-row";
                break;
            default: return [];
        }
        xmlUtil.foreach(node, function (n) {
            switch (n.localName) {
                case "pPr":
                    result.push({
                        target: "".concat(selector, " p"),
                        mod: modificator,
                        values: _this.parseDefaultProperties(n, {})
                    });
                    break;
                case "rPr":
                    result.push({
                        target: "".concat(selector, " span"),
                        mod: modificator,
                        values: _this.parseDefaultProperties(n, {})
                    });
                    break;
                case "tblPr":
                case "tcPr":
                    result.push({
                        target: selector,
                        mod: modificator,
                        values: _this.parseDefaultProperties(n, {})
                    });
                    break;
            }
        });
        return result;
    };
    DocumentParser.prototype.parseNumberingFile = function (xnums) {
        var _this = this;
        var result = [];
        var mapping = {};
        var bullets = [];
        xmlUtil.foreach(xnums, function (n) {
            switch (n.localName) {
                case "abstractNum":
                    _this.parseAbstractNumbering(n, bullets)
                        .forEach(function (x) { return result.push(x); });
                    break;
                case "numPicBullet":
                    bullets.push(_this.parseNumberingPicBullet(n));
                    break;
                case "num":
                    var numId = globalXmlParser.attr(n, "numId");
                    var abstractNumId = globalXmlParser.elementAttr(n, "abstractNumId", "val");
                    mapping[abstractNumId] = numId;
                    break;
            }
        });
        result.forEach(function (x) { return x.id = mapping[x.id]; });
        return result;
    };
    DocumentParser.prototype.parseNumberingPicBullet = function (elem) {
        var pict = globalXmlParser.element(elem, "pict");
        var shape = pict && globalXmlParser.element(pict, "shape");
        var imagedata = shape && globalXmlParser.element(shape, "imagedata");
        return imagedata ? {
            id: globalXmlParser.intAttr(elem, "numPicBulletId"),
            src: globalXmlParser.attr(imagedata, "id"),
            style: globalXmlParser.attr(shape, "style")
        } : null;
    };
    DocumentParser.prototype.parseAbstractNumbering = function (node, bullets) {
        var _this = this;
        var result = [];
        var id = globalXmlParser.attr(node, "abstractNumId");
        xmlUtil.foreach(node, function (n) {
            switch (n.localName) {
                case "lvl":
                    result.push(_this.parseNumberingLevel(id, n, bullets));
                    break;
            }
        });
        return result;
    };
    DocumentParser.prototype.parseNumberingLevel = function (id, node, bullets) {
        var _this = this;
        var result = {
            id: id,
            level: globalXmlParser.intAttr(node, "ilvl"),
            start: 1,
            pStyleName: undefined,
            pStyle: {},
            rStyle: {},
            suff: "tab"
        };
        xmlUtil.foreach(node, function (n) {
            switch (n.localName) {
                case "start":
                    result.start = globalXmlParser.intAttr(n, "val");
                    break;
                case "pPr":
                    _this.parseDefaultProperties(n, result.pStyle);
                    break;
                case "rPr":
                    _this.parseDefaultProperties(n, result.rStyle);
                    break;
                case "lvlPicBulletId":
                    var id = globalXmlParser.intAttr(n, "val");
                    result.bullet = bullets.find(function (x) { return x.id == id; });
                    break;
                case "lvlText":
                    result.levelText = globalXmlParser.attr(n, "val");
                    break;
                case "pStyle":
                    result.pStyleName = globalXmlParser.attr(n, "val");
                    break;
                case "numFmt":
                    result.format = globalXmlParser.attr(n, "val");
                    break;
                case "suff":
                    result.suff = globalXmlParser.attr(n, "val");
                    break;
            }
        });
        return result;
    };
    DocumentParser.prototype.parseSdt = function (node, parser) {
        var sdtContent = globalXmlParser.element(node, "sdtContent");
        return sdtContent ? parser(sdtContent) : [];
    };
    DocumentParser.prototype.parseInserted = function (node, parentParser) {
        var _a, _b;
        return {
            type: DomType.Inserted,
            children: (_b = (_a = parentParser(node)) === null || _a === void 0 ? void 0 : _a.children) !== null && _b !== void 0 ? _b : []
        };
    };
    DocumentParser.prototype.parseDeleted = function (node, parentParser) {
        var _a, _b;
        return {
            type: DomType.Deleted,
            children: (_b = (_a = parentParser(node)) === null || _a === void 0 ? void 0 : _a.children) !== null && _b !== void 0 ? _b : []
        };
    };
    DocumentParser.prototype.parseParagraph = function (node) {
        var _a;
        var _this = this;
        var result = { type: DomType.Paragraph, children: [] };
        for (var _i = 0, _b = globalXmlParser.elements(node); _i < _b.length; _i++) {
            var el = _b[_i];
            switch (el.localName) {
                case "pPr":
                    this.parseParagraphProperties(el, result);
                    break;
                case "r":
                    result.children.push(this.parseRun(el, result));
                    break;
                case "hyperlink":
                    result.children.push(this.parseHyperlink(el, result));
                    break;
                case "bookmarkStart":
                    result.children.push(parseBookmarkStart(el, globalXmlParser));
                    break;
                case "bookmarkEnd":
                    result.children.push(parseBookmarkEnd(el, globalXmlParser));
                    break;
                case "commentRangeStart":
                    result.children.push(new WmlCommentRangeStart(globalXmlParser.attr(el, "id")));
                    break;
                case "commentRangeEnd":
                    result.children.push(new WmlCommentRangeEnd(globalXmlParser.attr(el, "id")));
                    break;
                case "oMath":
                case "oMathPara":
                    result.children.push(this.parseMathElement(el));
                    break;
                case "sdt":
                    (_a = result.children).push.apply(_a, this.parseSdt(el, function (e) { return _this.parseParagraph(e).children; }));
                    break;
                case "ins":
                    result.children.push(this.parseInserted(el, function (e) { return _this.parseParagraph(e); }));
                    break;
                case "del":
                    result.children.push(this.parseDeleted(el, function (e) { return _this.parseParagraph(e); }));
                    break;
            }
        }
        return result;
    };
    DocumentParser.prototype.parseParagraphProperties = function (elem, paragraph) {
        var _this = this;
        this.parseDefaultProperties(elem, paragraph.cssStyle = {}, null, function (c) {
            if (parseParagraphProperty(c, paragraph, globalXmlParser))
                return true;
            switch (c.localName) {
                case "pStyle":
                    paragraph.styleName = globalXmlParser.attr(c, "val");
                    break;
                case "cnfStyle":
                    paragraph.className = values.classNameOfCnfStyle(c);
                    break;
                case "framePr":
                    _this.parseFrame(c, paragraph);
                    break;
                case "rPr":
                    break;
                default:
                    return false;
            }
            return true;
        });
    };
    DocumentParser.prototype.parseFrame = function (node, paragraph) {
        var dropCap = globalXmlParser.attr(node, "dropCap");
        if (dropCap == "drop")
            paragraph.cssStyle["float"] = "left";
    };
    DocumentParser.prototype.parseHyperlink = function (node, parent) {
        var _this = this;
        var result = { type: DomType.Hyperlink, parent: parent, children: [] };
        var anchor = globalXmlParser.attr(node, "anchor");
        var relId = globalXmlParser.attr(node, "id");
        if (anchor)
            result.href = "#" + anchor;
        if (relId)
            result.id = relId;
        xmlUtil.foreach(node, function (c) {
            switch (c.localName) {
                case "r":
                    result.children.push(_this.parseRun(c, result));
                    break;
            }
        });
        return result;
    };
    DocumentParser.prototype.parseRun = function (node, parent) {
        var _this = this;
        var result = { type: DomType.Run, parent: parent, children: [] };
        xmlUtil.foreach(node, function (c) {
            c = _this.checkAlternateContent(c);
            switch (c.localName) {
                case "t":
                    result.children.push({
                        type: DomType.Text,
                        text: c.textContent
                    });
                    break;
                case "delText":
                    result.children.push({
                        type: DomType.DeletedText,
                        text: c.textContent
                    });
                    break;
                case "commentReference":
                    result.children.push(new WmlCommentReference(globalXmlParser.attr(c, "id")));
                    break;
                case "fldSimple":
                    result.children.push({
                        type: DomType.SimpleField,
                        instruction: globalXmlParser.attr(c, "instr"),
                        lock: globalXmlParser.boolAttr(c, "lock", false),
                        dirty: globalXmlParser.boolAttr(c, "dirty", false)
                    });
                    break;
                case "instrText":
                    result.fieldRun = true;
                    result.children.push({
                        type: DomType.Instruction,
                        text: c.textContent
                    });
                    break;
                case "fldChar":
                    result.fieldRun = true;
                    result.children.push({
                        type: DomType.ComplexField,
                        charType: globalXmlParser.attr(c, "fldCharType"),
                        lock: globalXmlParser.boolAttr(c, "lock", false),
                        dirty: globalXmlParser.boolAttr(c, "dirty", false)
                    });
                    break;
                case "noBreakHyphen":
                    result.children.push({ type: DomType.NoBreakHyphen });
                    break;
                case "br":
                    result.children.push({
                        type: DomType.Break,
                        break: globalXmlParser.attr(c, "type") || "textWrapping"
                    });
                    break;
                case "lastRenderedPageBreak":
                    result.children.push({
                        type: DomType.Break,
                        break: "lastRenderedPageBreak"
                    });
                    break;
                case "sym":
                    result.children.push({
                        type: DomType.Symbol,
                        font: globalXmlParser.attr(c, "font"),
                        char: globalXmlParser.attr(c, "char")
                    });
                    break;
                case "tab":
                    result.children.push({ type: DomType.Tab });
                    break;
                case "footnoteReference":
                    result.children.push({
                        type: DomType.FootnoteReference,
                        id: globalXmlParser.attr(c, "id")
                    });
                    break;
                case "endnoteReference":
                    result.children.push({
                        type: DomType.EndnoteReference,
                        id: globalXmlParser.attr(c, "id")
                    });
                    break;
                case "drawing":
                    var d = _this.parseDrawing(c);
                    if (d)
                        result.children = [d];
                    break;
                case "pict":
                    result.children.push(_this.parseVmlPicture(c));
                    break;
                case "rPr":
                    _this.parseRunProperties(c, result);
                    break;
            }
        });
        return result;
    };
    DocumentParser.prototype.parseMathElement = function (elem) {
        var propsTag = "".concat(elem.localName, "Pr");
        var result = { type: mmlTagMap[elem.localName], children: [] };
        for (var _i = 0, _a = globalXmlParser.elements(elem); _i < _a.length; _i++) {
            var el = _a[_i];
            var childType = mmlTagMap[el.localName];
            if (childType) {
                result.children.push(this.parseMathElement(el));
            }
            else if (el.localName == "r") {
                var run = this.parseRun(el);
                run.type = DomType.MmlRun;
                result.children.push(run);
            }
            else if (el.localName == propsTag) {
                result.props = this.parseMathProperies(el);
            }
        }
        return result;
    };
    DocumentParser.prototype.parseMathProperies = function (elem) {
        var result = {};
        for (var _i = 0, _a = globalXmlParser.elements(elem); _i < _a.length; _i++) {
            var el = _a[_i];
            switch (el.localName) {
                case "chr":
                    result.char = globalXmlParser.attr(el, "val");
                    break;
                case "vertJc":
                    result.verticalJustification = globalXmlParser.attr(el, "val");
                    break;
                case "pos":
                    result.position = globalXmlParser.attr(el, "val");
                    break;
                case "degHide":
                    result.hideDegree = globalXmlParser.boolAttr(el, "val");
                    break;
                case "begChr":
                    result.beginChar = globalXmlParser.attr(el, "val");
                    break;
                case "endChr":
                    result.endChar = globalXmlParser.attr(el, "val");
                    break;
            }
        }
        return result;
    };
    DocumentParser.prototype.parseRunProperties = function (elem, run) {
        this.parseDefaultProperties(elem, run.cssStyle = {}, null, function (c) {
            switch (c.localName) {
                case "rStyle":
                    run.styleName = globalXmlParser.attr(c, "val");
                    break;
                case "vertAlign":
                    run.verticalAlign = values.valueOfVertAlign(c, true);
                    break;
                default:
                    return false;
            }
            return true;
        });
    };
    DocumentParser.prototype.parseVmlPicture = function (elem) {
        var result = { type: DomType.VmlPicture, children: [] };
        for (var _i = 0, _a = globalXmlParser.elements(elem); _i < _a.length; _i++) {
            var el = _a[_i];
            var child = parseVmlElement(el, this);
            child && result.children.push(child);
        }
        return result;
    };
    DocumentParser.prototype.checkAlternateContent = function (elem) {
        var _a;
        if (elem.localName != 'AlternateContent')
            return elem;
        var choice = globalXmlParser.element(elem, "Choice");
        if (choice) {
            var requires = globalXmlParser.attr(choice, "Requires");
            var namespaceURI = elem.lookupNamespaceURI(requires);
            if (supportedNamespaceURIs.includes(namespaceURI))
                return choice.firstElementChild;
        }
        return (_a = globalXmlParser.element(elem, "Fallback")) === null || _a === void 0 ? void 0 : _a.firstElementChild;
    };
    DocumentParser.prototype.parseDrawing = function (node) {
        for (var _i = 0, _a = globalXmlParser.elements(node); _i < _a.length; _i++) {
            var n = _a[_i];
            switch (n.localName) {
                case "inline":
                case "anchor":
                    return this.parseDrawingWrapper(n);
            }
        }
    };
    DocumentParser.prototype.parseDrawingWrapper = function (node) {
        var _a;
        var result = { type: DomType.Drawing, children: [], cssStyle: {} };
        var isAnchor = node.localName == "anchor";
        var wrapType = null;
        var simplePos = globalXmlParser.boolAttr(node, "simplePos");
        globalXmlParser.boolAttr(node, "behindDoc");
        var posX = { relative: "page", align: "left", offset: "0" };
        var posY = { relative: "page", align: "top", offset: "0" };
        for (var _i = 0, _b = globalXmlParser.elements(node); _i < _b.length; _i++) {
            var n = _b[_i];
            switch (n.localName) {
                case "simplePos":
                    if (simplePos) {
                        posX.offset = globalXmlParser.lengthAttr(n, "x", LengthUsage.Emu);
                        posY.offset = globalXmlParser.lengthAttr(n, "y", LengthUsage.Emu);
                    }
                    break;
                case "extent":
                    result.cssStyle["width"] = globalXmlParser.lengthAttr(n, "cx", LengthUsage.Emu);
                    result.cssStyle["height"] = globalXmlParser.lengthAttr(n, "cy", LengthUsage.Emu);
                    break;
                case "positionH":
                case "positionV":
                    if (!simplePos) {
                        var pos = n.localName == "positionH" ? posX : posY;
                        var alignNode = globalXmlParser.element(n, "align");
                        var offsetNode = globalXmlParser.element(n, "posOffset");
                        pos.relative = (_a = globalXmlParser.attr(n, "relativeFrom")) !== null && _a !== void 0 ? _a : pos.relative;
                        if (alignNode)
                            pos.align = alignNode.textContent;
                        if (offsetNode)
                            pos.offset = xmlUtil.sizeValue(offsetNode, LengthUsage.Emu);
                    }
                    break;
                case "wrapTopAndBottom":
                    wrapType = "wrapTopAndBottom";
                    break;
                case "wrapNone":
                    wrapType = "wrapNone";
                    break;
                case "graphic":
                    var g = this.parseGraphic(n);
                    if (g)
                        result.children.push(g);
                    break;
            }
        }
        if (wrapType == "wrapTopAndBottom") {
            result.cssStyle['display'] = 'block';
            if (posX.align) {
                result.cssStyle['text-align'] = posX.align;
                result.cssStyle['width'] = "100%";
            }
        }
        else if (wrapType == "wrapNone") {
            result.cssStyle['display'] = 'block';
            result.cssStyle['position'] = 'relative';
            result.cssStyle["width"] = "0px";
            result.cssStyle["height"] = "0px";
            if (posX.offset)
                result.cssStyle["left"] = posX.offset;
            if (posY.offset)
                result.cssStyle["top"] = posY.offset;
        }
        else if (isAnchor && (posX.align == 'left' || posX.align == 'right')) {
            result.cssStyle["float"] = posX.align;
        }
        return result;
    };
    DocumentParser.prototype.parseGraphic = function (elem) {
        var graphicData = globalXmlParser.element(elem, "graphicData");
        for (var _i = 0, _a = globalXmlParser.elements(graphicData); _i < _a.length; _i++) {
            var n = _a[_i];
            switch (n.localName) {
                case "pic":
                    return this.parsePicture(n);
            }
        }
        return null;
    };
    DocumentParser.prototype.parsePicture = function (elem) {
        var result = { type: DomType.Image, src: "", cssStyle: {} };
        var blipFill = globalXmlParser.element(elem, "blipFill");
        var blip = globalXmlParser.element(blipFill, "blip");
        result.src = globalXmlParser.attr(blip, "embed");
        var spPr = globalXmlParser.element(elem, "spPr");
        var xfrm = globalXmlParser.element(spPr, "xfrm");
        result.cssStyle["position"] = "relative";
        for (var _i = 0, _a = globalXmlParser.elements(xfrm); _i < _a.length; _i++) {
            var n = _a[_i];
            switch (n.localName) {
                case "ext":
                    result.cssStyle["width"] = globalXmlParser.lengthAttr(n, "cx", LengthUsage.Emu);
                    result.cssStyle["height"] = globalXmlParser.lengthAttr(n, "cy", LengthUsage.Emu);
                    break;
                case "off":
                    result.cssStyle["left"] = globalXmlParser.lengthAttr(n, "x", LengthUsage.Emu);
                    result.cssStyle["top"] = globalXmlParser.lengthAttr(n, "y", LengthUsage.Emu);
                    break;
            }
        }
        return result;
    };
    DocumentParser.prototype.parseTable = function (node) {
        var _this = this;
        var result = { type: DomType.Table, children: [] };
        xmlUtil.foreach(node, function (c) {
            switch (c.localName) {
                case "tr":
                    result.children.push(_this.parseTableRow(c));
                    break;
                case "tblGrid":
                    result.columns = _this.parseTableColumns(c);
                    break;
                case "tblPr":
                    _this.parseTableProperties(c, result);
                    break;
            }
        });
        return result;
    };
    DocumentParser.prototype.parseTableColumns = function (node) {
        var result = [];
        xmlUtil.foreach(node, function (n) {
            switch (n.localName) {
                case "gridCol":
                    result.push({ width: globalXmlParser.lengthAttr(n, "w") });
                    break;
            }
        });
        return result;
    };
    DocumentParser.prototype.parseTableProperties = function (elem, table) {
        var _this = this;
        table.cssStyle = {};
        table.cellStyle = {};
        this.parseDefaultProperties(elem, table.cssStyle, table.cellStyle, function (c) {
            switch (c.localName) {
                case "tblStyle":
                    table.styleName = globalXmlParser.attr(c, "val");
                    break;
                case "tblLook":
                    table.className = values.classNameOftblLook(c);
                    break;
                case "tblpPr":
                    _this.parseTablePosition(c, table);
                    break;
                case "tblStyleColBandSize":
                    table.colBandSize = globalXmlParser.intAttr(c, "val");
                    break;
                case "tblStyleRowBandSize":
                    table.rowBandSize = globalXmlParser.intAttr(c, "val");
                    break;
                default:
                    return false;
            }
            return true;
        });
        switch (table.cssStyle["text-align"]) {
            case "center":
                delete table.cssStyle["text-align"];
                table.cssStyle["margin-left"] = "auto";
                table.cssStyle["margin-right"] = "auto";
                break;
            case "right":
                delete table.cssStyle["text-align"];
                table.cssStyle["margin-left"] = "auto";
                break;
        }
    };
    DocumentParser.prototype.parseTablePosition = function (node, table) {
        var topFromText = globalXmlParser.lengthAttr(node, "topFromText");
        var bottomFromText = globalXmlParser.lengthAttr(node, "bottomFromText");
        var rightFromText = globalXmlParser.lengthAttr(node, "rightFromText");
        var leftFromText = globalXmlParser.lengthAttr(node, "leftFromText");
        table.cssStyle["float"] = 'left';
        table.cssStyle["margin-bottom"] = values.addSize(table.cssStyle["margin-bottom"], bottomFromText);
        table.cssStyle["margin-left"] = values.addSize(table.cssStyle["margin-left"], leftFromText);
        table.cssStyle["margin-right"] = values.addSize(table.cssStyle["margin-right"], rightFromText);
        table.cssStyle["margin-top"] = values.addSize(table.cssStyle["margin-top"], topFromText);
    };
    DocumentParser.prototype.parseTableRow = function (node) {
        var _this = this;
        var result = { type: DomType.Row, children: [] };
        xmlUtil.foreach(node, function (c) {
            switch (c.localName) {
                case "tc":
                    result.children.push(_this.parseTableCell(c));
                    break;
                case "trPr":
                    _this.parseTableRowProperties(c, result);
                    break;
            }
        });
        return result;
    };
    DocumentParser.prototype.parseTableRowProperties = function (elem, row) {
        row.cssStyle = this.parseDefaultProperties(elem, {}, null, function (c) {
            switch (c.localName) {
                case "cnfStyle":
                    row.className = values.classNameOfCnfStyle(c);
                    break;
                case "tblHeader":
                    row.isHeader = globalXmlParser.boolAttr(c, "val");
                    break;
                default:
                    return false;
            }
            return true;
        });
    };
    DocumentParser.prototype.parseTableCell = function (node) {
        var _this = this;
        var result = { type: DomType.Cell, children: [] };
        xmlUtil.foreach(node, function (c) {
            switch (c.localName) {
                case "tbl":
                    result.children.push(_this.parseTable(c));
                    break;
                case "p":
                    result.children.push(_this.parseParagraph(c));
                    break;
                case "tcPr":
                    _this.parseTableCellProperties(c, result);
                    break;
            }
        });
        return result;
    };
    DocumentParser.prototype.parseTableCellProperties = function (elem, cell) {
        cell.cssStyle = this.parseDefaultProperties(elem, {}, null, function (c) {
            var _a;
            switch (c.localName) {
                case "gridSpan":
                    cell.span = globalXmlParser.intAttr(c, "val", null);
                    break;
                case "vMerge":
                    cell.verticalMerge = (_a = globalXmlParser.attr(c, "val")) !== null && _a !== void 0 ? _a : "continue";
                    break;
                case "cnfStyle":
                    cell.className = values.classNameOfCnfStyle(c);
                    break;
                default:
                    return false;
            }
            return true;
        });
    };
    DocumentParser.prototype.parseDefaultProperties = function (elem, style, childStyle, handler) {
        var _this = this;
        if (style === void 0) { style = null; }
        if (childStyle === void 0) { childStyle = null; }
        if (handler === void 0) { handler = null; }
        style = style || {};
        xmlUtil.foreach(elem, function (c) {
            if (handler === null || handler === void 0 ? void 0 : handler(c))
                return;
            switch (c.localName) {
                case "jc":
                    style["text-align"] = values.valueOfJc(c);
                    break;
                case "textAlignment":
                    style["vertical-align"] = values.valueOfTextAlignment(c);
                    break;
                case "color":
                    style["color"] = xmlUtil.colorAttr(c, "val", null, autos.color);
                    break;
                case "sz":
                    style["font-size"] = style["min-height"] = globalXmlParser.lengthAttr(c, "val", LengthUsage.FontSize);
                    break;
                case "shd":
                    style["background-color"] = xmlUtil.colorAttr(c, "fill", null, autos.shd);
                    break;
                case "highlight":
                    style["background-color"] = xmlUtil.colorAttr(c, "val", null, autos.highlight);
                    break;
                case "vertAlign":
                    break;
                case "position":
                    style.verticalAlign = globalXmlParser.lengthAttr(c, "val", LengthUsage.FontSize);
                    break;
                case "tcW":
                    if (_this.options.ignoreWidth)
                        break;
                case "tblW":
                    style["width"] = values.valueOfSize(c, "w");
                    break;
                case "trHeight":
                    _this.parseTrHeight(c, style);
                    break;
                case "strike":
                    style["text-decoration"] = globalXmlParser.boolAttr(c, "val", true) ? "line-through" : "none";
                    break;
                case "b":
                    style["font-weight"] = globalXmlParser.boolAttr(c, "val", true) ? "bold" : "normal";
                    break;
                case "i":
                    style["font-style"] = globalXmlParser.boolAttr(c, "val", true) ? "italic" : "normal";
                    break;
                case "caps":
                    style["text-transform"] = globalXmlParser.boolAttr(c, "val", true) ? "uppercase" : "none";
                    break;
                case "smallCaps":
                    style["font-variant"] = globalXmlParser.boolAttr(c, "val", true) ? "small-caps" : "none";
                    break;
                case "u":
                    _this.parseUnderline(c, style);
                    break;
                case "ind":
                case "tblInd":
                    _this.parseIndentation(c, style);
                    break;
                case "rFonts":
                    _this.parseFont(c, style);
                    break;
                case "tblBorders":
                    _this.parseBorderProperties(c, childStyle || style);
                    break;
                case "tblCellSpacing":
                    style["border-spacing"] = values.valueOfMargin(c);
                    style["border-collapse"] = "separate";
                    break;
                case "pBdr":
                    _this.parseBorderProperties(c, style);
                    break;
                case "bdr":
                    style["border"] = values.valueOfBorder(c);
                    break;
                case "tcBorders":
                    _this.parseBorderProperties(c, style);
                    break;
                case "vanish":
                    if (globalXmlParser.boolAttr(c, "val", true))
                        style["display"] = "none";
                    break;
                case "kern":
                    break;
                case "noWrap":
                    break;
                case "tblCellMar":
                case "tcMar":
                    _this.parseMarginProperties(c, childStyle || style);
                    break;
                case "tblLayout":
                    style["table-layout"] = values.valueOfTblLayout(c);
                    break;
                case "vAlign":
                    style["vertical-align"] = values.valueOfTextAlignment(c);
                    break;
                case "spacing":
                    if (elem.localName == "pPr")
                        _this.parseSpacing(c, style);
                    break;
                case "wordWrap":
                    if (globalXmlParser.boolAttr(c, "val"))
                        style["overflow-wrap"] = "break-word";
                    break;
                case "suppressAutoHyphens":
                    style["hyphens"] = globalXmlParser.boolAttr(c, "val", true) ? "none" : "auto";
                    break;
                case "lang":
                    style["$lang"] = globalXmlParser.attr(c, "val");
                    break;
                case "bCs":
                case "iCs":
                case "szCs":
                case "tabs":
                case "outlineLvl":
                case "contextualSpacing":
                case "tblStyleColBandSize":
                case "tblStyleRowBandSize":
                case "webHidden":
                case "pageBreakBefore":
                case "suppressLineNumbers":
                case "keepLines":
                case "keepNext":
                case "widowControl":
                case "bidi":
                case "rtl":
                case "noProof":
                    break;
                default:
                    if (_this.options.debug)
                        console.warn("DOCX: Unknown document element: ".concat(elem.localName, ".").concat(c.localName));
                    break;
            }
        });
        return style;
    };
    DocumentParser.prototype.parseUnderline = function (node, style) {
        var val = globalXmlParser.attr(node, "val");
        if (val == null)
            return;
        switch (val) {
            case "dash":
            case "dashDotDotHeavy":
            case "dashDotHeavy":
            case "dashedHeavy":
            case "dashLong":
            case "dashLongHeavy":
            case "dotDash":
            case "dotDotDash":
                style["text-decoration"] = "underline dashed";
                break;
            case "dotted":
            case "dottedHeavy":
                style["text-decoration"] = "underline dotted";
                break;
            case "double":
                style["text-decoration"] = "underline double";
                break;
            case "single":
            case "thick":
                style["text-decoration"] = "underline";
                break;
            case "wave":
            case "wavyDouble":
            case "wavyHeavy":
                style["text-decoration"] = "underline wavy";
                break;
            case "words":
                style["text-decoration"] = "underline";
                break;
            case "none":
                style["text-decoration"] = "none";
                break;
        }
        var col = xmlUtil.colorAttr(node, "color");
        if (col)
            style["text-decoration-color"] = col;
    };
    DocumentParser.prototype.parseFont = function (node, style) {
        var ascii = globalXmlParser.attr(node, "ascii");
        var asciiTheme = values.themeValue(node, "asciiTheme");
        var fonts = [ascii, asciiTheme].filter(function (x) { return x; }).join(', ');
        if (fonts.length > 0)
            style["font-family"] = fonts;
    };
    DocumentParser.prototype.parseIndentation = function (node, style) {
        var firstLine = globalXmlParser.lengthAttr(node, "firstLine");
        var hanging = globalXmlParser.lengthAttr(node, "hanging");
        var left = globalXmlParser.lengthAttr(node, "left");
        var start = globalXmlParser.lengthAttr(node, "start");
        var right = globalXmlParser.lengthAttr(node, "right");
        var end = globalXmlParser.lengthAttr(node, "end");
        if (firstLine)
            style["text-indent"] = firstLine;
        if (hanging)
            style["text-indent"] = "-".concat(hanging);
        if (left || start)
            style["margin-left"] = left || start;
        if (right || end)
            style["margin-right"] = right || end;
    };
    DocumentParser.prototype.parseSpacing = function (node, style) {
        var before = globalXmlParser.lengthAttr(node, "before");
        var after = globalXmlParser.lengthAttr(node, "after");
        var line = globalXmlParser.intAttr(node, "line", null);
        var lineRule = globalXmlParser.attr(node, "lineRule");
        if (before)
            style["margin-top"] = before;
        if (after)
            style["margin-bottom"] = after;
        if (line !== null) {
            switch (lineRule) {
                case "auto":
                    style["line-height"] = "".concat((line / 240).toFixed(2));
                    break;
                case "atLeast":
                    style["line-height"] = "calc(100% + ".concat(line / 20, "pt)");
                    break;
                default:
                    style["line-height"] = style["min-height"] = "".concat(line / 20, "pt");
                    break;
            }
        }
    };
    DocumentParser.prototype.parseMarginProperties = function (node, output) {
        xmlUtil.foreach(node, function (c) {
            switch (c.localName) {
                case "left":
                    output["padding-left"] = values.valueOfMargin(c);
                    break;
                case "right":
                    output["padding-right"] = values.valueOfMargin(c);
                    break;
                case "top":
                    output["padding-top"] = values.valueOfMargin(c);
                    break;
                case "bottom":
                    output["padding-bottom"] = values.valueOfMargin(c);
                    break;
            }
        });
    };
    DocumentParser.prototype.parseTrHeight = function (node, output) {
        switch (globalXmlParser.attr(node, "hRule")) {
            case "exact":
                output["height"] = globalXmlParser.lengthAttr(node, "val");
                break;
            case "atLeast":
            default:
                output["height"] = globalXmlParser.lengthAttr(node, "val");
                break;
        }
    };
    DocumentParser.prototype.parseBorderProperties = function (node, output) {
        xmlUtil.foreach(node, function (c) {
            switch (c.localName) {
                case "start":
                case "left":
                    output["border-left"] = values.valueOfBorder(c);
                    break;
                case "end":
                case "right":
                    output["border-right"] = values.valueOfBorder(c);
                    break;
                case "top":
                    output["border-top"] = values.valueOfBorder(c);
                    break;
                case "bottom":
                    output["border-bottom"] = values.valueOfBorder(c);
                    break;
            }
        });
    };
    return DocumentParser;
}());
var knownColors = ['black', 'blue', 'cyan', 'darkBlue', 'darkCyan', 'darkGray', 'darkGreen', 'darkMagenta', 'darkRed', 'darkYellow', 'green', 'lightGray', 'magenta', 'none', 'red', 'white', 'yellow'];
var xmlUtil = (function () {
    function xmlUtil() {
    }
    xmlUtil.foreach = function (node, cb) {
        for (var i = 0; i < node.childNodes.length; i++) {
            var n = node.childNodes[i];
            if (n.nodeType == Node.ELEMENT_NODE)
                cb(n);
        }
    };
    xmlUtil.colorAttr = function (node, attrName, defValue, autoColor) {
        if (defValue === void 0) { defValue = null; }
        if (autoColor === void 0) { autoColor = 'black'; }
        var v = globalXmlParser.attr(node, attrName);
        if (v) {
            if (v == "auto") {
                return autoColor;
            }
            else if (knownColors.includes(v)) {
                return v;
            }
            return "#".concat(v);
        }
        var themeColor = globalXmlParser.attr(node, "themeColor");
        return themeColor ? "var(--docx-".concat(themeColor, "-color)") : defValue;
    };
    xmlUtil.sizeValue = function (node, type) {
        if (type === void 0) { type = LengthUsage.Dxa; }
        return convertLength(node.textContent, type);
    };
    return xmlUtil;
}());
var values = (function () {
    function values() {
    }
    values.themeValue = function (c, attr) {
        var val = globalXmlParser.attr(c, attr);
        return val ? "var(--docx-".concat(val, "-font)") : null;
    };
    values.valueOfSize = function (c, attr) {
        var type = LengthUsage.Dxa;
        switch (globalXmlParser.attr(c, "type")) {
            case "dxa": break;
            case "pct":
                type = LengthUsage.Percent;
                break;
            case "auto": return "auto";
        }
        return globalXmlParser.lengthAttr(c, attr, type);
    };
    values.valueOfMargin = function (c) {
        return globalXmlParser.lengthAttr(c, "w");
    };
    values.valueOfBorder = function (c) {
        var type = globalXmlParser.attr(c, "val");
        if (type == "nil")
            return "none";
        var color = xmlUtil.colorAttr(c, "color");
        var size = globalXmlParser.lengthAttr(c, "sz", LengthUsage.Border);
        return "".concat(size, " solid ").concat(color == "auto" ? autos.borderColor : color);
    };
    values.valueOfTblLayout = function (c) {
        var type = globalXmlParser.attr(c, "val");
        return type == "fixed" ? "fixed" : "auto";
    };
    values.classNameOfCnfStyle = function (c) {
        var val = globalXmlParser.attr(c, "val");
        var classes = [
            'first-row', 'last-row', 'first-col', 'last-col',
            'odd-col', 'even-col', 'odd-row', 'even-row',
            'ne-cell', 'nw-cell', 'se-cell', 'sw-cell'
        ];
        return classes.filter(function (_, i) { return val[i] == '1'; }).join(' ');
    };
    values.valueOfJc = function (c) {
        var type = globalXmlParser.attr(c, "val");
        switch (type) {
            case "start":
            case "left": return "left";
            case "center": return "center";
            case "end":
            case "right": return "right";
            case "both": return "justify";
        }
        return type;
    };
    values.valueOfVertAlign = function (c, asTagName) {
        if (asTagName === void 0) { asTagName = false; }
        var type = globalXmlParser.attr(c, "val");
        switch (type) {
            case "subscript": return "sub";
            case "superscript": return asTagName ? "sup" : "super";
        }
        return asTagName ? null : type;
    };
    values.valueOfTextAlignment = function (c) {
        var type = globalXmlParser.attr(c, "val");
        switch (type) {
            case "auto":
            case "baseline": return "baseline";
            case "top": return "top";
            case "center": return "middle";
            case "bottom": return "bottom";
        }
        return type;
    };
    values.addSize = function (a, b) {
        if (a == null)
            return b;
        if (b == null)
            return a;
        return "calc(".concat(a, " + ").concat(b, ")");
    };
    values.classNameOftblLook = function (c) {
        var val = globalXmlParser.hexAttr(c, "val", 0);
        var className = "";
        if (globalXmlParser.boolAttr(c, "firstRow") || (val & 0x0020))
            className += " first-row";
        if (globalXmlParser.boolAttr(c, "lastRow") || (val & 0x0040))
            className += " last-row";
        if (globalXmlParser.boolAttr(c, "firstColumn") || (val & 0x0080))
            className += " first-col";
        if (globalXmlParser.boolAttr(c, "lastColumn") || (val & 0x0100))
            className += " last-col";
        if (globalXmlParser.boolAttr(c, "noHBand") || (val & 0x0200))
            className += " no-hband";
        if (globalXmlParser.boolAttr(c, "noVBand") || (val & 0x0400))
            className += " no-vband";
        return className.trim();
    };
    return values;
}());

var defaultTab = { pos: 0, leader: "none", style: "left" };
var maxTabs = 50;
function computePixelToPoint(container) {
    if (container === void 0) { container = document.body; }
    var temp = document.createElement("div");
    temp.style.width = '100pt';
    container.appendChild(temp);
    var result = 100 / temp.offsetWidth;
    container.removeChild(temp);
    return result;
}
function updateTabStop(elem, tabs, defaultTabSize, pixelToPoint) {
    if (pixelToPoint === void 0) { pixelToPoint = 72 / 96; }
    var p = elem.closest("p");
    var ebb = elem.getBoundingClientRect();
    var pbb = p.getBoundingClientRect();
    var pcs = getComputedStyle(p);
    var tabStops = (tabs === null || tabs === void 0 ? void 0 : tabs.length) > 0 ? tabs.map(function (t) { return ({
        pos: lengthToPoint(t.position),
        leader: t.leader,
        style: t.style
    }); }).sort(function (a, b) { return a.pos - b.pos; }) : [defaultTab];
    var lastTab = tabStops[tabStops.length - 1];
    var pWidthPt = pbb.width * pixelToPoint;
    var size = lengthToPoint(defaultTabSize);
    var pos = lastTab.pos + size;
    if (pos < pWidthPt) {
        for (; pos < pWidthPt && tabStops.length < maxTabs; pos += size) {
            tabStops.push(__assign(__assign({}, defaultTab), { pos: pos }));
        }
    }
    var marginLeft = parseFloat(pcs.marginLeft);
    var pOffset = pbb.left + marginLeft;
    var left = (ebb.left - pOffset) * pixelToPoint;
    var tab = tabStops.find(function (t) { return t.style != "clear" && t.pos > left; });
    if (tab == null)
        return;
    var width = 1;
    if (tab.style == "right" || tab.style == "center") {
        var tabStops_1 = Array.from(p.querySelectorAll(".".concat(elem.className)));
        var nextIdx = tabStops_1.indexOf(elem) + 1;
        var range = document.createRange();
        range.setStart(elem, 1);
        if (nextIdx < tabStops_1.length) {
            range.setEndBefore(tabStops_1[nextIdx]);
        }
        else {
            range.setEndAfter(p);
        }
        var mul = tab.style == "center" ? 0.5 : 1;
        var nextBB = range.getBoundingClientRect();
        var offset = nextBB.left + mul * nextBB.width - (pbb.left - marginLeft);
        width = tab.pos - offset * pixelToPoint;
    }
    else {
        width = tab.pos - left;
    }
    elem.innerHTML = "&nbsp;";
    elem.style.textDecoration = "inherit";
    elem.style.wordSpacing = "".concat(width.toFixed(0), "pt");
    switch (tab.leader) {
        case "dot":
        case "middleDot":
            elem.style.textDecoration = "underline";
            elem.style.textDecorationStyle = "dotted";
            break;
        case "hyphen":
        case "heavy":
        case "underscore":
            elem.style.textDecoration = "underline";
            break;
    }
}
function lengthToPoint(length) {
    return parseFloat(length);
}

var ns = {
    svg: "http://www.w3.org/2000/svg",
    mathML: "http://www.w3.org/1998/Math/MathML"
};
var HtmlRenderer = (function () {
    function HtmlRenderer(htmlDocument) {
        this.htmlDocument = htmlDocument;
        this.className = "docx";
        this.styleMap = {};
        this.currentPart = null;
        this.tableVerticalMerges = [];
        this.currentVerticalMerge = null;
        this.tableCellPositions = [];
        this.currentCellPosition = null;
        this.footnoteMap = {};
        this.endnoteMap = {};
        this.currentEndnoteIds = [];
        this.usedHederFooterParts = [];
        this.currentTabs = [];
        this.tabsTimeout = 0;
        this.commentMap = {};
        this.tasks = [];
        this.postRenderTasks = [];
        this.createElement = createElement;
    }
    HtmlRenderer.prototype.render = function (document, bodyContainer, styleContainer, options) {
        var _a;
        if (styleContainer === void 0) { styleContainer = null; }
        this.document = document;
        this.options = options;
        this.className = options.className;
        this.rootSelector = options.inWrapper ? ".".concat(this.className, "-wrapper") : ':root';
        this.styleMap = null;
        this.tasks = [];
        if (this.options.renderComments && Highlight) {
            this.commentHighlight = new Highlight();
        }
        styleContainer = styleContainer || bodyContainer;
        removeAllElements(styleContainer);
        removeAllElements(bodyContainer);
        appendComment(styleContainer, "docxjs library predefined styles");
        styleContainer.appendChild(this.renderDefaultStyle());
        if (document.themePart) {
            appendComment(styleContainer, "docxjs document theme values");
            this.renderTheme(document.themePart, styleContainer);
        }
        if (document.stylesPart != null) {
            this.styleMap = this.processStyles(document.stylesPart.styles);
            appendComment(styleContainer, "docxjs document styles");
            styleContainer.appendChild(this.renderStyles(document.stylesPart.styles));
        }
        if (document.numberingPart) {
            this.prodessNumberings(document.numberingPart.domNumberings);
            appendComment(styleContainer, "docxjs document numbering styles");
            styleContainer.appendChild(this.renderNumbering(document.numberingPart.domNumberings, styleContainer));
        }
        if (document.footnotesPart) {
            this.footnoteMap = keyBy(document.footnotesPart.notes, function (x) { return x.id; });
        }
        if (document.endnotesPart) {
            this.endnoteMap = keyBy(document.endnotesPart.notes, function (x) { return x.id; });
        }
        if (document.settingsPart) {
            this.defaultTabSize = (_a = document.settingsPart.settings) === null || _a === void 0 ? void 0 : _a.defaultTabStop;
        }
        if (!options.ignoreFonts && document.fontTablePart)
            this.renderFontTable(document.fontTablePart, styleContainer);
        var sectionElements = this.renderSections(document.documentPart.body);
        if (this.options.inWrapper) {
            bodyContainer.appendChild(this.renderWrapper(sectionElements));
        }
        else {
            appendChildren(bodyContainer, sectionElements);
        }
        if (this.commentHighlight && options.renderComments) {
            CSS.highlights.set("".concat(this.className, "-comments"), this.commentHighlight);
        }
        this.refreshTabStops();
        this.postRenderTasks.forEach(function (t) { return t(); });
    };
    HtmlRenderer.prototype.renderTheme = function (themePart, styleContainer) {
        var _a, _b;
        var variables = {};
        var fontScheme = (_a = themePart.theme) === null || _a === void 0 ? void 0 : _a.fontScheme;
        if (fontScheme) {
            if (fontScheme.majorFont) {
                variables['--docx-majorHAnsi-font'] = fontScheme.majorFont.latinTypeface;
            }
            if (fontScheme.minorFont) {
                variables['--docx-minorHAnsi-font'] = fontScheme.minorFont.latinTypeface;
            }
        }
        var colorScheme = (_b = themePart.theme) === null || _b === void 0 ? void 0 : _b.colorScheme;
        if (colorScheme) {
            for (var _i = 0, _c = Object.entries(colorScheme.colors); _i < _c.length; _i++) {
                var _d = _c[_i], k = _d[0], v = _d[1];
                variables["--docx-".concat(k, "-color")] = "#".concat(v);
            }
        }
        var cssText = this.styleToString(".".concat(this.className), variables);
        styleContainer.appendChild(createStyleElement(cssText));
    };
    HtmlRenderer.prototype.renderFontTable = function (fontsPart, styleContainer) {
        var _this = this;
        var _loop_1 = function (f) {
            var _loop_2 = function (ref) {
                this_1.tasks.push(this_1.document.loadFont(ref.id, ref.key).then(function (fontData) {
                    var cssValues = {
                        'font-family': f.name,
                        'src': "url(".concat(fontData, ")")
                    };
                    if (ref.type == "bold" || ref.type == "boldItalic") {
                        cssValues['font-weight'] = 'bold';
                    }
                    if (ref.type == "italic" || ref.type == "boldItalic") {
                        cssValues['font-style'] = 'italic';
                    }
                    appendComment(styleContainer, "docxjs ".concat(f.name, " font"));
                    var cssText = _this.styleToString("@font-face", cssValues);
                    styleContainer.appendChild(createStyleElement(cssText));
                    _this.refreshTabStops();
                }));
            };
            for (var _b = 0, _c = f.embedFontRefs; _b < _c.length; _b++) {
                var ref = _c[_b];
                _loop_2(ref);
            }
        };
        var this_1 = this;
        for (var _i = 0, _a = fontsPart.fonts; _i < _a.length; _i++) {
            var f = _a[_i];
            _loop_1(f);
        }
    };
    HtmlRenderer.prototype.processStyleName = function (className) {
        return className ? "".concat(this.className, "_").concat(escapeClassName(className)) : this.className;
    };
    HtmlRenderer.prototype.processStyles = function (styles) {
        var stylesMap = keyBy(styles.filter(function (x) { return x.id != null; }), function (x) { return x.id; });
        for (var _i = 0, _a = styles.filter(function (x) { return x.basedOn; }); _i < _a.length; _i++) {
            var style = _a[_i];
            var baseStyle = stylesMap[style.basedOn];
            if (baseStyle) {
                style.paragraphProps = mergeDeep(style.paragraphProps, baseStyle.paragraphProps);
                style.runProps = mergeDeep(style.runProps, baseStyle.runProps);
                var _loop_3 = function (baseValues) {
                    var styleValues = style.styles.find(function (x) { return x.target == baseValues.target; });
                    if (styleValues) {
                        this_2.copyStyleProperties(baseValues.values, styleValues.values);
                    }
                    else {
                        style.styles.push(__assign(__assign({}, baseValues), { values: __assign({}, baseValues.values) }));
                    }
                };
                var this_2 = this;
                for (var _b = 0, _c = baseStyle.styles; _b < _c.length; _b++) {
                    var baseValues = _c[_b];
                    _loop_3(baseValues);
                }
            }
            else if (this.options.debug)
                console.warn("Can't find base style ".concat(style.basedOn));
        }
        for (var _d = 0, styles_1 = styles; _d < styles_1.length; _d++) {
            var style = styles_1[_d];
            style.cssName = this.processStyleName(style.id);
        }
        return stylesMap;
    };
    HtmlRenderer.prototype.prodessNumberings = function (numberings) {
        var _a;
        for (var _i = 0, _b = numberings.filter(function (n) { return n.pStyleName; }); _i < _b.length; _i++) {
            var num = _b[_i];
            var style = this.findStyle(num.pStyleName);
            if ((_a = style === null || style === void 0 ? void 0 : style.paragraphProps) === null || _a === void 0 ? void 0 : _a.numbering) {
                style.paragraphProps.numbering.level = num.level;
            }
        }
    };
    HtmlRenderer.prototype.processElement = function (element) {
        if (element.children) {
            for (var _i = 0, _a = element.children; _i < _a.length; _i++) {
                var e = _a[_i];
                e.parent = element;
                if (e.type == DomType.Table) {
                    this.processTable(e);
                }
                else {
                    this.processElement(e);
                }
            }
        }
    };
    HtmlRenderer.prototype.processTable = function (table) {
        for (var _i = 0, _a = table.children; _i < _a.length; _i++) {
            var r = _a[_i];
            for (var _b = 0, _c = r.children; _b < _c.length; _b++) {
                var c = _c[_b];
                c.cssStyle = this.copyStyleProperties(table.cellStyle, c.cssStyle, [
                    "border-left", "border-right", "border-top", "border-bottom",
                    "padding-left", "padding-right", "padding-top", "padding-bottom"
                ]);
                this.processElement(c);
            }
        }
    };
    HtmlRenderer.prototype.copyStyleProperties = function (input, output, attrs) {
        if (attrs === void 0) { attrs = null; }
        if (!input)
            return output;
        if (output == null)
            output = {};
        if (attrs == null)
            attrs = Object.getOwnPropertyNames(input);
        for (var _i = 0, attrs_1 = attrs; _i < attrs_1.length; _i++) {
            var key = attrs_1[_i];
            if (input.hasOwnProperty(key) && !output.hasOwnProperty(key))
                output[key] = input[key];
        }
        return output;
    };
    HtmlRenderer.prototype.createPageElement = function (className, props) {
        var elem = this.createElement("section", { className: className });
        if (props) {
            if (props.pageMargins) {
                elem.style.paddingLeft = props.pageMargins.left;
                elem.style.paddingRight = props.pageMargins.right;
                elem.style.paddingTop = props.pageMargins.top;
                elem.style.paddingBottom = props.pageMargins.bottom;
            }
            if (props.pageSize) {
                if (!this.options.ignoreWidth)
                    elem.style.width = props.pageSize.width;
                if (!this.options.ignoreHeight)
                    elem.style.minHeight = props.pageSize.height;
            }
        }
        return elem;
    };
    HtmlRenderer.prototype.createSectionContent = function (props) {
        var elem = this.createElement("article");
        if (props.columns && props.columns.numberOfColumns) {
            elem.style.columnCount = "".concat(props.columns.numberOfColumns);
            elem.style.columnGap = props.columns.space;
            if (props.columns.separator) {
                elem.style.columnRule = "1px solid black";
            }
        }
        return elem;
    };
    HtmlRenderer.prototype.renderSections = function (document) {
        var result = [];
        this.processElement(document);
        var sections = this.splitBySection(document.children, document.props);
        var pages = this.groupByPageBreaks(sections);
        var prevProps = null;
        for (var i = 0, l = pages.length; i < l; i++) {
            this.currentFootnoteIds = [];
            var section = pages[i][0];
            var props = section.sectProps;
            var pageElement = this.createPageElement(this.className, props);
            this.renderStyleValues(document.cssStyle, pageElement);
            this.options.renderHeaders && this.renderHeaderFooter(props.headerRefs, props, result.length, prevProps != props, pageElement);
            for (var _i = 0, _a = pages[i]; _i < _a.length; _i++) {
                var sect = _a[_i];
                var contentElement = this.createSectionContent(sect.sectProps);
                this.renderElements(sect.elements, contentElement);
                pageElement.appendChild(contentElement);
                props = sect.sectProps;
            }
            if (this.options.renderFootnotes) {
                this.renderNotes(this.currentFootnoteIds, this.footnoteMap, pageElement);
            }
            if (this.options.renderEndnotes && i == l - 1) {
                this.renderNotes(this.currentEndnoteIds, this.endnoteMap, pageElement);
            }
            this.options.renderFooters && this.renderHeaderFooter(props.footerRefs, props, result.length, prevProps != props, pageElement);
            result.push(pageElement);
            prevProps = props;
        }
        return result;
    };
    HtmlRenderer.prototype.renderHeaderFooter = function (refs, props, page, firstOfSection, into) {
        var _a, _b;
        if (!refs)
            return;
        var ref = (_b = (_a = (props.titlePage && firstOfSection ? refs.find(function (x) { return x.type == "first"; }) : null)) !== null && _a !== void 0 ? _a : (page % 2 == 1 ? refs.find(function (x) { return x.type == "even"; }) : null)) !== null && _b !== void 0 ? _b : refs.find(function (x) { return x.type == "default"; });
        var part = ref && this.document.findPartByRelId(ref.id, this.document.documentPart);
        if (part) {
            this.currentPart = part;
            if (!this.usedHederFooterParts.includes(part.path)) {
                this.processElement(part.rootElement);
                this.usedHederFooterParts.push(part.path);
            }
            var el = this.renderElements([part.rootElement], into)[0];
            if (props === null || props === void 0 ? void 0 : props.pageMargins) {
                if (part.rootElement.type === DomType.Header) {
                    el.style.marginTop = "calc(".concat(props.pageMargins.header, " - ").concat(props.pageMargins.top, ")");
                    el.style.minHeight = "calc(".concat(props.pageMargins.top, " - ").concat(props.pageMargins.header, ")");
                }
                else if (part.rootElement.type === DomType.Footer) {
                    el.style.marginBottom = "calc(".concat(props.pageMargins.footer, " - ").concat(props.pageMargins.bottom, ")");
                    el.style.minHeight = "calc(".concat(props.pageMargins.bottom, " - ").concat(props.pageMargins.footer, ")");
                }
            }
            this.currentPart = null;
        }
    };
    HtmlRenderer.prototype.isPageBreakElement = function (elem) {
        if (elem.type != DomType.Break)
            return false;
        if (elem.break == "lastRenderedPageBreak")
            return !this.options.ignoreLastRenderedPageBreak;
        return elem.break == "page";
    };
    HtmlRenderer.prototype.isPageBreakSection = function (prev, next) {
        var _a, _b, _c, _d, _e, _f;
        if (!prev)
            return false;
        if (!next)
            return false;
        return ((_a = prev.pageSize) === null || _a === void 0 ? void 0 : _a.orientation) != ((_b = next.pageSize) === null || _b === void 0 ? void 0 : _b.orientation)
            || ((_c = prev.pageSize) === null || _c === void 0 ? void 0 : _c.width) != ((_d = next.pageSize) === null || _d === void 0 ? void 0 : _d.width)
            || ((_e = prev.pageSize) === null || _e === void 0 ? void 0 : _e.height) != ((_f = next.pageSize) === null || _f === void 0 ? void 0 : _f.height);
    };
    HtmlRenderer.prototype.splitBySection = function (elements, defaultProps) {
        var _this = this;
        var _a;
        var current = { sectProps: null, elements: [], pageBreak: false };
        var result = [current];
        for (var _i = 0, elements_1 = elements; _i < elements_1.length; _i++) {
            var elem = elements_1[_i];
            if (elem.type == DomType.Paragraph) {
                var s = this.findStyle(elem.styleName);
                if ((_a = s === null || s === void 0 ? void 0 : s.paragraphProps) === null || _a === void 0 ? void 0 : _a.pageBreakBefore) {
                    current.sectProps = sectProps;
                    current.pageBreak = true;
                    current = { sectProps: null, elements: [], pageBreak: false };
                    result.push(current);
                }
            }
            current.elements.push(elem);
            if (elem.type == DomType.Paragraph) {
                var p = elem;
                var sectProps = p.sectionProps;
                var pBreakIndex = -1;
                var rBreakIndex = -1;
                if (this.options.breakPages && p.children) {
                    pBreakIndex = p.children.findIndex(function (r) {
                        var _a, _b;
                        rBreakIndex = (_b = (_a = r.children) === null || _a === void 0 ? void 0 : _a.findIndex(_this.isPageBreakElement.bind(_this))) !== null && _b !== void 0 ? _b : -1;
                        return rBreakIndex != -1;
                    });
                }
                if (sectProps || pBreakIndex != -1) {
                    current.sectProps = sectProps;
                    current.pageBreak = pBreakIndex != -1;
                    current = { sectProps: null, elements: [], pageBreak: false };
                    result.push(current);
                }
                if (pBreakIndex != -1) {
                    var breakRun = p.children[pBreakIndex];
                    var splitRun = rBreakIndex < breakRun.children.length - 1;
                    if (pBreakIndex < p.children.length - 1 || splitRun) {
                        var children = elem.children;
                        var newParagraph = __assign(__assign({}, elem), { children: children.slice(pBreakIndex) });
                        elem.children = children.slice(0, pBreakIndex);
                        current.elements.push(newParagraph);
                        if (splitRun) {
                            var runChildren = breakRun.children;
                            var newRun = __assign(__assign({}, breakRun), { children: runChildren.slice(0, rBreakIndex) });
                            elem.children.push(newRun);
                            breakRun.children = runChildren.slice(rBreakIndex);
                        }
                    }
                }
            }
        }
        var currentSectProps = null;
        for (var i = result.length - 1; i >= 0; i--) {
            if (result[i].sectProps == null) {
                result[i].sectProps = currentSectProps !== null && currentSectProps !== void 0 ? currentSectProps : defaultProps;
            }
            else {
                currentSectProps = result[i].sectProps;
            }
        }
        return result;
    };
    HtmlRenderer.prototype.groupByPageBreaks = function (sections) {
        var current = [];
        var prev;
        var result = [current];
        for (var _i = 0, sections_1 = sections; _i < sections_1.length; _i++) {
            var s = sections_1[_i];
            current.push(s);
            if (this.options.ignoreLastRenderedPageBreak || s.pageBreak || this.isPageBreakSection(prev, s.sectProps))
                result.push(current = []);
            prev = s.sectProps;
        }
        return result.filter(function (x) { return x.length > 0; });
    };
    HtmlRenderer.prototype.renderWrapper = function (children) {
        return this.createElement("div", { className: "".concat(this.className, "-wrapper") }, children);
    };
    HtmlRenderer.prototype.renderDefaultStyle = function () {
        var c = this.className;
        var styleText = "\n.".concat(c, "-wrapper { background: gray; padding: 30px; padding-bottom: 0px; display: flex; flex-flow: column; align-items: center; } \n.").concat(c, "-wrapper>section.").concat(c, " { background: white; box-shadow: 0 0 10px rgba(0, 0, 0, 0.5); margin-bottom: 30px; }\n.").concat(c, " { color: black; hyphens: auto; text-underline-position: from-font; }\nsection.").concat(c, " { box-sizing: border-box; display: flex; flex-flow: column nowrap; position: relative; overflow: hidden; }\nsection.").concat(c, ">article { margin-bottom: auto; z-index: 1; }\nsection.").concat(c, ">footer { z-index: 1; }\n.").concat(c, " table { border-collapse: collapse; }\n.").concat(c, " table td, .").concat(c, " table th { vertical-align: top; }\n.").concat(c, " p { margin: 0pt; min-height: 1em; }\n.").concat(c, " span { white-space: pre-wrap; overflow-wrap: break-word; }\n.").concat(c, " a { color: inherit; text-decoration: inherit; }\n.").concat(c, " svg { fill: transparent; }\n");
        if (this.options.renderComments) {
            styleText += "\n.".concat(c, "-comment-ref { cursor: default; }\n.").concat(c, "-comment-popover { display: none; z-index: 1000; padding: 0.5rem; background: white; position: absolute; box-shadow: 0 0 0.25rem rgba(0, 0, 0, 0.25); width: 30ch; }\n.").concat(c, "-comment-ref:hover~.").concat(c, "-comment-popover { display: block; }\n.").concat(c, "-comment-author,.").concat(c, "-comment-date { font-size: 0.875rem; color: #888; }\n");
        }
        return createStyleElement(styleText);
    };
    HtmlRenderer.prototype.renderNumbering = function (numberings, styleContainer) {
        var _this = this;
        var styleText = "";
        var resetCounters = [];
        var _loop_4 = function () {
            selector = "p.".concat(this_3.numberingClass(num.id, num.level));
            listStyleType = "none";
            if (num.bullet) {
                var valiable_1 = "--".concat(this_3.className, "-").concat(num.bullet.src).toLowerCase();
                styleText += this_3.styleToString("".concat(selector, ":before"), {
                    "content": "' '",
                    "display": "inline-block",
                    "background": "var(".concat(valiable_1, ")")
                }, num.bullet.style);
                this_3.tasks.push(this_3.document.loadNumberingImage(num.bullet.src).then(function (data) {
                    var text = "".concat(_this.rootSelector, " { ").concat(valiable_1, ": url(").concat(data, ") }");
                    styleContainer.appendChild(createStyleElement(text));
                }));
            }
            else if (num.levelText) {
                var counter = this_3.numberingCounter(num.id, num.level);
                var counterReset = counter + " " + (num.start - 1);
                if (num.level > 0) {
                    styleText += this_3.styleToString("p.".concat(this_3.numberingClass(num.id, num.level - 1)), {
                        "counter-reset": counterReset
                    });
                }
                resetCounters.push(counterReset);
                styleText += this_3.styleToString("".concat(selector, ":before"), __assign({ "content": this_3.levelTextToContent(num.levelText, num.suff, num.id, this_3.numFormatToCssValue(num.format)), "counter-increment": counter }, num.rStyle));
            }
            else {
                listStyleType = this_3.numFormatToCssValue(num.format);
            }
            styleText += this_3.styleToString(selector, __assign({ "display": "list-item", "list-style-position": "inside", "list-style-type": listStyleType }, num.pStyle));
        };
        var this_3 = this, selector, listStyleType;
        for (var _i = 0, numberings_1 = numberings; _i < numberings_1.length; _i++) {
            var num = numberings_1[_i];
            _loop_4();
        }
        if (resetCounters.length > 0) {
            styleText += this.styleToString(this.rootSelector, {
                "counter-reset": resetCounters.join(" ")
            });
        }
        return createStyleElement(styleText);
    };
    HtmlRenderer.prototype.renderStyles = function (styles) {
        var _a;
        var styleText = "";
        var stylesMap = this.styleMap;
        var defautStyles = keyBy(styles.filter(function (s) { return s.isDefault; }), function (s) { return s.target; });
        for (var _i = 0, styles_2 = styles; _i < styles_2.length; _i++) {
            var style = styles_2[_i];
            var subStyles = style.styles;
            if (style.linked) {
                var linkedStyle = style.linked && stylesMap[style.linked];
                if (linkedStyle)
                    subStyles = subStyles.concat(linkedStyle.styles);
                else if (this.options.debug)
                    console.warn("Can't find linked style ".concat(style.linked));
            }
            for (var _b = 0, subStyles_1 = subStyles; _b < subStyles_1.length; _b++) {
                var subStyle = subStyles_1[_b];
                var selector = "".concat((_a = style.target) !== null && _a !== void 0 ? _a : '', ".").concat(style.cssName);
                if (style.target != subStyle.target)
                    selector += " ".concat(subStyle.target);
                if (defautStyles[style.target] == style)
                    selector = ".".concat(this.className, " ").concat(style.target, ", ") + selector;
                styleText += this.styleToString(selector, subStyle.values);
            }
        }
        return createStyleElement(styleText);
    };
    HtmlRenderer.prototype.renderNotes = function (noteIds, notesMap, into) {
        var notes = noteIds.map(function (id) { return notesMap[id]; }).filter(function (x) { return x; });
        if (notes.length > 0) {
            var result = this.createElement("ol", null, this.renderElements(notes));
            into.appendChild(result);
        }
    };
    HtmlRenderer.prototype.renderElement = function (elem) {
        switch (elem.type) {
            case DomType.Paragraph:
                return this.renderParagraph(elem);
            case DomType.BookmarkStart:
                return this.renderBookmarkStart(elem);
            case DomType.BookmarkEnd:
                return null;
            case DomType.Run:
                return this.renderRun(elem);
            case DomType.Table:
                return this.renderTable(elem);
            case DomType.Row:
                return this.renderTableRow(elem);
            case DomType.Cell:
                return this.renderTableCell(elem);
            case DomType.Hyperlink:
                return this.renderHyperlink(elem);
            case DomType.Drawing:
                return this.renderDrawing(elem);
            case DomType.Image:
                return this.renderImage(elem);
            case DomType.Text:
                return this.renderText(elem);
            case DomType.Text:
                return this.renderText(elem);
            case DomType.DeletedText:
                return this.renderDeletedText(elem);
            case DomType.Tab:
                return this.renderTab(elem);
            case DomType.Symbol:
                return this.renderSymbol(elem);
            case DomType.Break:
                return this.renderBreak(elem);
            case DomType.Footer:
                return this.renderContainer(elem, "footer");
            case DomType.Header:
                return this.renderContainer(elem, "header");
            case DomType.Footnote:
            case DomType.Endnote:
                return this.renderContainer(elem, "li");
            case DomType.FootnoteReference:
                return this.renderFootnoteReference(elem);
            case DomType.EndnoteReference:
                return this.renderEndnoteReference(elem);
            case DomType.NoBreakHyphen:
                return this.createElement("wbr");
            case DomType.VmlPicture:
                return this.renderVmlPicture(elem);
            case DomType.VmlElement:
                return this.renderVmlElement(elem);
            case DomType.MmlMath:
                return this.renderContainerNS(elem, ns.mathML, "math", { xmlns: ns.mathML });
            case DomType.MmlMathParagraph:
                return this.renderContainer(elem, "span");
            case DomType.MmlFraction:
                return this.renderContainerNS(elem, ns.mathML, "mfrac");
            case DomType.MmlBase:
                return this.renderContainerNS(elem, ns.mathML, elem.parent.type == DomType.MmlMatrixRow ? "mtd" : "mrow");
            case DomType.MmlNumerator:
            case DomType.MmlDenominator:
            case DomType.MmlFunction:
            case DomType.MmlLimit:
            case DomType.MmlBox:
                return this.renderContainerNS(elem, ns.mathML, "mrow");
            case DomType.MmlGroupChar:
                return this.renderMmlGroupChar(elem);
            case DomType.MmlLimitLower:
                return this.renderContainerNS(elem, ns.mathML, "munder");
            case DomType.MmlMatrix:
                return this.renderContainerNS(elem, ns.mathML, "mtable");
            case DomType.MmlMatrixRow:
                return this.renderContainerNS(elem, ns.mathML, "mtr");
            case DomType.MmlRadical:
                return this.renderMmlRadical(elem);
            case DomType.MmlSuperscript:
                return this.renderContainerNS(elem, ns.mathML, "msup");
            case DomType.MmlSubscript:
                return this.renderContainerNS(elem, ns.mathML, "msub");
            case DomType.MmlDegree:
            case DomType.MmlSuperArgument:
            case DomType.MmlSubArgument:
                return this.renderContainerNS(elem, ns.mathML, "mn");
            case DomType.MmlFunctionName:
                return this.renderContainerNS(elem, ns.mathML, "ms");
            case DomType.MmlDelimiter:
                return this.renderMmlDelimiter(elem);
            case DomType.MmlRun:
                return this.renderMmlRun(elem);
            case DomType.MmlNary:
                return this.renderMmlNary(elem);
            case DomType.MmlPreSubSuper:
                return this.renderMmlPreSubSuper(elem);
            case DomType.MmlBar:
                return this.renderMmlBar(elem);
            case DomType.MmlEquationArray:
                return this.renderMllList(elem);
            case DomType.Inserted:
                return this.renderInserted(elem);
            case DomType.Deleted:
                return this.renderDeleted(elem);
            case DomType.CommentRangeStart:
                return this.renderCommentRangeStart(elem);
            case DomType.CommentRangeEnd:
                return this.renderCommentRangeEnd(elem);
            case DomType.CommentReference:
                return this.renderCommentReference(elem);
        }
        return null;
    };
    HtmlRenderer.prototype.renderChildren = function (elem, into) {
        return this.renderElements(elem.children, into);
    };
    HtmlRenderer.prototype.renderElements = function (elems, into) {
        var _this = this;
        if (elems == null)
            return null;
        var result = elems.flatMap(function (e) { return _this.renderElement(e); }).filter(function (e) { return e != null; });
        if (into)
            appendChildren(into, result);
        return result;
    };
    HtmlRenderer.prototype.renderContainer = function (elem, tagName, props) {
        return this.createElement(tagName, props, this.renderChildren(elem));
    };
    HtmlRenderer.prototype.renderContainerNS = function (elem, ns, tagName, props) {
        return createElementNS(ns, tagName, props, this.renderChildren(elem));
    };
    HtmlRenderer.prototype.renderParagraph = function (elem) {
        var _a, _b, _c, _d;
        var result = this.createElement("p");
        var style = this.findStyle(elem.styleName);
        (_a = elem.tabs) !== null && _a !== void 0 ? _a : (elem.tabs = (_b = style === null || style === void 0 ? void 0 : style.paragraphProps) === null || _b === void 0 ? void 0 : _b.tabs);
        this.renderClass(elem, result);
        this.renderChildren(elem, result);
        this.renderStyleValues(elem.cssStyle, result);
        this.renderCommonProperties(result.style, elem);
        var numbering = (_c = elem.numbering) !== null && _c !== void 0 ? _c : (_d = style === null || style === void 0 ? void 0 : style.paragraphProps) === null || _d === void 0 ? void 0 : _d.numbering;
        if (numbering) {
            result.classList.add(this.numberingClass(numbering.id, numbering.level));
        }
        return result;
    };
    HtmlRenderer.prototype.renderRunProperties = function (style, props) {
        this.renderCommonProperties(style, props);
    };
    HtmlRenderer.prototype.renderCommonProperties = function (style, props) {
        if (props == null)
            return;
        if (props.color) {
            style["color"] = props.color;
        }
        if (props.fontSize) {
            style["font-size"] = props.fontSize;
        }
    };
    HtmlRenderer.prototype.renderHyperlink = function (elem) {
        var result = this.createElement("a");
        this.renderChildren(elem, result);
        this.renderStyleValues(elem.cssStyle, result);
        if (elem.href) {
            result.href = elem.href;
        }
        else if (elem.id) {
            var rel = this.document.documentPart.rels
                .find(function (it) { return it.id == elem.id && it.targetMode === "External"; });
            result.href = rel === null || rel === void 0 ? void 0 : rel.target;
        }
        return result;
    };
    HtmlRenderer.prototype.renderCommentRangeStart = function (commentStart) {
        var _a;
        if (!this.options.renderComments)
            return null;
        var rng = new Range();
        (_a = this.commentHighlight) === null || _a === void 0 ? void 0 : _a.add(rng);
        var result = this.htmlDocument.createComment("start of comment #".concat(commentStart.id));
        this.later(function () { return rng.setStart(result, 0); });
        this.commentMap[commentStart.id] = rng;
        return result;
    };
    HtmlRenderer.prototype.renderCommentRangeEnd = function (commentEnd) {
        if (!this.options.renderComments)
            return null;
        var rng = this.commentMap[commentEnd.id];
        var result = this.htmlDocument.createComment("end of comment #".concat(commentEnd.id));
        this.later(function () { return rng === null || rng === void 0 ? void 0 : rng.setEnd(result, 0); });
        return result;
    };
    HtmlRenderer.prototype.renderCommentReference = function (commentRef) {
        var _a;
        if (!this.options.renderComments)
            return null;
        var comment = (_a = this.document.commentsPart) === null || _a === void 0 ? void 0 : _a.commentMap[commentRef.id];
        if (!comment)
            return null;
        var frg = new DocumentFragment();
        var commentRefEl = createElement("span", { className: "".concat(this.className, "-comment-ref") }, ['💬']);
        var commentsContainerEl = createElement("div", { className: "".concat(this.className, "-comment-popover") });
        this.renderCommentContent(comment, commentsContainerEl);
        frg.appendChild(this.htmlDocument.createComment("comment #".concat(comment.id, " by ").concat(comment.author, " on ").concat(comment.date)));
        frg.appendChild(commentRefEl);
        frg.appendChild(commentsContainerEl);
        return frg;
    };
    HtmlRenderer.prototype.renderCommentContent = function (comment, container) {
        container.appendChild(createElement('div', { className: "".concat(this.className, "-comment-author") }, [comment.author]));
        container.appendChild(createElement('div', { className: "".concat(this.className, "-comment-date") }, [new Date(comment.date).toLocaleString()]));
        this.renderChildren(comment, container);
    };
    HtmlRenderer.prototype.renderDrawing = function (elem) {
        var result = this.createElement("div");
        result.style.display = "inline-block";
        result.style.position = "relative";
        result.style.textIndent = "0px";
        this.renderChildren(elem, result);
        this.renderStyleValues(elem.cssStyle, result);
        return result;
    };
    HtmlRenderer.prototype.renderImage = function (elem) {
        var result = this.createElement("img");
        this.renderStyleValues(elem.cssStyle, result);
        if (this.document) {
            this.tasks.push(this.document.loadDocumentImage(elem.src, this.currentPart).then(function (x) {
                result.src = x;
            }));
        }
        return result;
    };
    HtmlRenderer.prototype.renderText = function (elem) {
        return this.htmlDocument.createTextNode(elem.text);
    };
    HtmlRenderer.prototype.renderDeletedText = function (elem) {
        return this.options.renderEndnotes ? this.htmlDocument.createTextNode(elem.text) : null;
    };
    HtmlRenderer.prototype.renderBreak = function (elem) {
        if (elem.break == "textWrapping") {
            return this.createElement("br");
        }
        return null;
    };
    HtmlRenderer.prototype.renderInserted = function (elem) {
        if (this.options.renderChanges)
            return this.renderContainer(elem, "ins");
        return this.renderChildren(elem);
    };
    HtmlRenderer.prototype.renderDeleted = function (elem) {
        if (this.options.renderChanges)
            return this.renderContainer(elem, "del");
        return null;
    };
    HtmlRenderer.prototype.renderSymbol = function (elem) {
        var span = this.createElement("span");
        span.style.fontFamily = elem.font;
        span.innerHTML = "&#x".concat(elem.char, ";");
        return span;
    };
    HtmlRenderer.prototype.renderFootnoteReference = function (elem) {
        var result = this.createElement("sup");
        this.currentFootnoteIds.push(elem.id);
        result.textContent = "".concat(this.currentFootnoteIds.length);
        return result;
    };
    HtmlRenderer.prototype.renderEndnoteReference = function (elem) {
        var result = this.createElement("sup");
        this.currentEndnoteIds.push(elem.id);
        result.textContent = "".concat(this.currentEndnoteIds.length);
        return result;
    };
    HtmlRenderer.prototype.renderTab = function (elem) {
        var _a;
        var tabSpan = this.createElement("span");
        tabSpan.innerHTML = "&emsp;";
        if (this.options.experimental) {
            tabSpan.className = this.tabStopClass();
            var stops = (_a = findParent(elem, DomType.Paragraph)) === null || _a === void 0 ? void 0 : _a.tabs;
            this.currentTabs.push({ stops: stops, span: tabSpan });
        }
        return tabSpan;
    };
    HtmlRenderer.prototype.renderBookmarkStart = function (elem) {
        var result = this.createElement("span");
        result.id = elem.name;
        return result;
    };
    HtmlRenderer.prototype.renderRun = function (elem) {
        if (elem.fieldRun)
            return null;
        var result = this.createElement("span");
        if (elem.id)
            result.id = elem.id;
        this.renderClass(elem, result);
        this.renderStyleValues(elem.cssStyle, result);
        if (elem.verticalAlign) {
            var wrapper = this.createElement(elem.verticalAlign);
            this.renderChildren(elem, wrapper);
            result.appendChild(wrapper);
        }
        else {
            this.renderChildren(elem, result);
        }
        return result;
    };
    HtmlRenderer.prototype.renderTable = function (elem) {
        var result = this.createElement("table");
        this.tableCellPositions.push(this.currentCellPosition);
        this.tableVerticalMerges.push(this.currentVerticalMerge);
        this.currentVerticalMerge = {};
        this.currentCellPosition = { col: 0, row: 0 };
        if (elem.columns)
            result.appendChild(this.renderTableColumns(elem.columns));
        this.renderClass(elem, result);
        this.renderChildren(elem, result);
        this.renderStyleValues(elem.cssStyle, result);
        this.currentVerticalMerge = this.tableVerticalMerges.pop();
        this.currentCellPosition = this.tableCellPositions.pop();
        return result;
    };
    HtmlRenderer.prototype.renderTableColumns = function (columns) {
        var result = this.createElement("colgroup");
        for (var _i = 0, columns_1 = columns; _i < columns_1.length; _i++) {
            var col = columns_1[_i];
            var colElem = this.createElement("col");
            if (col.width)
                colElem.style.width = col.width;
            result.appendChild(colElem);
        }
        return result;
    };
    HtmlRenderer.prototype.renderTableRow = function (elem) {
        var result = this.createElement("tr");
        this.currentCellPosition.col = 0;
        this.renderClass(elem, result);
        this.renderChildren(elem, result);
        this.renderStyleValues(elem.cssStyle, result);
        this.currentCellPosition.row++;
        return result;
    };
    HtmlRenderer.prototype.renderTableCell = function (elem) {
        var result = this.createElement("td");
        var key = this.currentCellPosition.col;
        if (elem.verticalMerge) {
            if (elem.verticalMerge == "restart") {
                this.currentVerticalMerge[key] = result;
                result.rowSpan = 1;
            }
            else if (this.currentVerticalMerge[key]) {
                this.currentVerticalMerge[key].rowSpan += 1;
                result.style.display = "none";
            }
        }
        else {
            this.currentVerticalMerge[key] = null;
        }
        this.renderClass(elem, result);
        this.renderChildren(elem, result);
        this.renderStyleValues(elem.cssStyle, result);
        if (elem.span)
            result.colSpan = elem.span;
        this.currentCellPosition.col += result.colSpan;
        return result;
    };
    HtmlRenderer.prototype.renderVmlPicture = function (elem) {
        var result = createElement("div");
        this.renderChildren(elem, result);
        return result;
    };
    HtmlRenderer.prototype.renderVmlElement = function (elem) {
        var _a, _b;
        var container = createSvgElement("svg");
        container.setAttribute("style", elem.cssStyleText);
        var result = this.renderVmlChildElement(elem);
        if ((_a = elem.imageHref) === null || _a === void 0 ? void 0 : _a.id) {
            this.tasks.push((_b = this.document) === null || _b === void 0 ? void 0 : _b.loadDocumentImage(elem.imageHref.id, this.currentPart).then(function (x) { return result.setAttribute("href", x); }));
        }
        container.appendChild(result);
        requestAnimationFrame(function () {
            var bb = container.firstElementChild.getBBox();
            container.setAttribute("width", "".concat(Math.ceil(bb.x + bb.width)));
            container.setAttribute("height", "".concat(Math.ceil(bb.y + bb.height)));
        });
        return container;
    };
    HtmlRenderer.prototype.renderVmlChildElement = function (elem) {
        var result = createSvgElement(elem.tagName);
        Object.entries(elem.attrs).forEach(function (_a) {
            var k = _a[0], v = _a[1];
            return result.setAttribute(k, v);
        });
        for (var _i = 0, _a = elem.children; _i < _a.length; _i++) {
            var child = _a[_i];
            if (child.type == DomType.VmlElement) {
                result.appendChild(this.renderVmlChildElement(child));
            }
            else {
                result.appendChild.apply(result, asArray(this.renderElement(child)));
            }
        }
        return result;
    };
    HtmlRenderer.prototype.renderMmlRadical = function (elem) {
        var _a;
        var base = elem.children.find(function (el) { return el.type == DomType.MmlBase; });
        if ((_a = elem.props) === null || _a === void 0 ? void 0 : _a.hideDegree) {
            return createElementNS(ns.mathML, "msqrt", null, this.renderElements([base]));
        }
        var degree = elem.children.find(function (el) { return el.type == DomType.MmlDegree; });
        return createElementNS(ns.mathML, "mroot", null, this.renderElements([base, degree]));
    };
    HtmlRenderer.prototype.renderMmlDelimiter = function (elem) {
        var _a, _b;
        var children = [];
        children.push(createElementNS(ns.mathML, "mo", null, [(_a = elem.props.beginChar) !== null && _a !== void 0 ? _a : '(']));
        children.push.apply(children, this.renderElements(elem.children));
        children.push(createElementNS(ns.mathML, "mo", null, [(_b = elem.props.endChar) !== null && _b !== void 0 ? _b : ')']));
        return createElementNS(ns.mathML, "mrow", null, children);
    };
    HtmlRenderer.prototype.renderMmlNary = function (elem) {
        var _a, _b;
        var children = [];
        var grouped = keyBy(elem.children, function (x) { return x.type; });
        var sup = grouped[DomType.MmlSuperArgument];
        var sub = grouped[DomType.MmlSubArgument];
        var supElem = sup ? createElementNS(ns.mathML, "mo", null, asArray(this.renderElement(sup))) : null;
        var subElem = sub ? createElementNS(ns.mathML, "mo", null, asArray(this.renderElement(sub))) : null;
        var charElem = createElementNS(ns.mathML, "mo", null, [(_b = (_a = elem.props) === null || _a === void 0 ? void 0 : _a.char) !== null && _b !== void 0 ? _b : '\u222B']);
        if (supElem || subElem) {
            children.push(createElementNS(ns.mathML, "munderover", null, [charElem, subElem, supElem]));
        }
        else if (supElem) {
            children.push(createElementNS(ns.mathML, "mover", null, [charElem, supElem]));
        }
        else if (subElem) {
            children.push(createElementNS(ns.mathML, "munder", null, [charElem, subElem]));
        }
        else {
            children.push(charElem);
        }
        children.push.apply(children, this.renderElements(grouped[DomType.MmlBase].children));
        return createElementNS(ns.mathML, "mrow", null, children);
    };
    HtmlRenderer.prototype.renderMmlPreSubSuper = function (elem) {
        var children = [];
        var grouped = keyBy(elem.children, function (x) { return x.type; });
        var sup = grouped[DomType.MmlSuperArgument];
        var sub = grouped[DomType.MmlSubArgument];
        var supElem = sup ? createElementNS(ns.mathML, "mo", null, asArray(this.renderElement(sup))) : null;
        var subElem = sub ? createElementNS(ns.mathML, "mo", null, asArray(this.renderElement(sub))) : null;
        var stubElem = createElementNS(ns.mathML, "mo", null);
        children.push(createElementNS(ns.mathML, "msubsup", null, [stubElem, subElem, supElem]));
        children.push.apply(children, this.renderElements(grouped[DomType.MmlBase].children));
        return createElementNS(ns.mathML, "mrow", null, children);
    };
    HtmlRenderer.prototype.renderMmlGroupChar = function (elem) {
        var tagName = elem.props.verticalJustification === "bot" ? "mover" : "munder";
        var result = this.renderContainerNS(elem, ns.mathML, tagName);
        if (elem.props.char) {
            result.appendChild(createElementNS(ns.mathML, "mo", null, [elem.props.char]));
        }
        return result;
    };
    HtmlRenderer.prototype.renderMmlBar = function (elem) {
        var result = this.renderContainerNS(elem, ns.mathML, "mrow");
        switch (elem.props.position) {
            case "top":
                result.style.textDecoration = "overline";
                break;
            case "bottom":
                result.style.textDecoration = "underline";
                break;
        }
        return result;
    };
    HtmlRenderer.prototype.renderMmlRun = function (elem) {
        var result = createElementNS(ns.mathML, "ms");
        this.renderClass(elem, result);
        this.renderStyleValues(elem.cssStyle, result);
        this.renderChildren(elem, result);
        return result;
    };
    HtmlRenderer.prototype.renderMllList = function (elem) {
        var result = createElementNS(ns.mathML, "mtable");
        this.renderClass(elem, result);
        this.renderStyleValues(elem.cssStyle, result);
        this.renderChildren(elem);
        for (var _i = 0, _a = this.renderChildren(elem); _i < _a.length; _i++) {
            var child = _a[_i];
            result.appendChild(createElementNS(ns.mathML, "mtr", null, [
                createElementNS(ns.mathML, "mtd", null, [child])
            ]));
        }
        return result;
    };
    HtmlRenderer.prototype.renderStyleValues = function (style, ouput) {
        for (var k in style) {
            if (k.startsWith("$")) {
                ouput.setAttribute(k.slice(1), style[k]);
            }
            else {
                ouput.style[k] = style[k];
            }
        }
    };
    HtmlRenderer.prototype.renderClass = function (input, ouput) {
        if (input.className)
            ouput.className = input.className;
        if (input.styleName)
            ouput.classList.add(this.processStyleName(input.styleName));
    };
    HtmlRenderer.prototype.findStyle = function (styleName) {
        var _a;
        return styleName && ((_a = this.styleMap) === null || _a === void 0 ? void 0 : _a[styleName]);
    };
    HtmlRenderer.prototype.numberingClass = function (id, lvl) {
        return "".concat(this.className, "-num-").concat(id, "-").concat(lvl);
    };
    HtmlRenderer.prototype.tabStopClass = function () {
        return "".concat(this.className, "-tab-stop");
    };
    HtmlRenderer.prototype.styleToString = function (selectors, values, cssText) {
        if (cssText === void 0) { cssText = null; }
        var result = "".concat(selectors, " {\r\n");
        for (var key in values) {
            if (key.startsWith('$'))
                continue;
            result += "  ".concat(key, ": ").concat(values[key], ";\r\n");
        }
        if (cssText)
            result += cssText;
        return result + "}\r\n";
    };
    HtmlRenderer.prototype.numberingCounter = function (id, lvl) {
        return "".concat(this.className, "-num-").concat(id, "-").concat(lvl);
    };
    HtmlRenderer.prototype.levelTextToContent = function (text, suff, id, numformat) {
        var _this = this;
        var _a;
        var suffMap = {
            "tab": "\\9",
            "space": "\\a0",
        };
        var result = text.replace(/%\d*/g, function (s) {
            var lvl = parseInt(s.substring(1), 10) - 1;
            return "\"counter(".concat(_this.numberingCounter(id, lvl), ", ").concat(numformat, ")\"");
        });
        return "\"".concat(result).concat((_a = suffMap[suff]) !== null && _a !== void 0 ? _a : "", "\"");
    };
    HtmlRenderer.prototype.numFormatToCssValue = function (format) {
        var _a;
        var mapping = {
            none: "none",
            bullet: "disc",
            decimal: "decimal",
            lowerLetter: "lower-alpha",
            upperLetter: "upper-alpha",
            lowerRoman: "lower-roman",
            upperRoman: "upper-roman",
            decimalZero: "decimal-leading-zero",
            aiueo: "katakana",
            aiueoFullWidth: "katakana",
            chineseCounting: "simp-chinese-informal",
            chineseCountingThousand: "simp-chinese-informal",
            chineseLegalSimplified: "simp-chinese-formal",
            chosung: "hangul-consonant",
            ideographDigital: "cjk-ideographic",
            ideographTraditional: "cjk-heavenly-stem",
            ideographLegalTraditional: "trad-chinese-formal",
            ideographZodiac: "cjk-earthly-branch",
            iroha: "katakana-iroha",
            irohaFullWidth: "katakana-iroha",
            japaneseCounting: "japanese-informal",
            japaneseDigitalTenThousand: "cjk-decimal",
            japaneseLegal: "japanese-formal",
            thaiNumbers: "thai",
            koreanCounting: "korean-hangul-formal",
            koreanDigital: "korean-hangul-formal",
            koreanDigital2: "korean-hanja-informal",
            hebrew1: "hebrew",
            hebrew2: "hebrew",
            hindiNumbers: "devanagari",
            ganada: "hangul",
            taiwaneseCounting: "cjk-ideographic",
            taiwaneseCountingThousand: "cjk-ideographic",
            taiwaneseDigital: "cjk-decimal",
        };
        return (_a = mapping[format]) !== null && _a !== void 0 ? _a : format;
    };
    HtmlRenderer.prototype.refreshTabStops = function () {
        var _this = this;
        if (!this.options.experimental)
            return;
        clearTimeout(this.tabsTimeout);
        this.tabsTimeout = setTimeout(function () {
            var pixelToPoint = computePixelToPoint();
            for (var _i = 0, _a = _this.currentTabs; _i < _a.length; _i++) {
                var tab = _a[_i];
                updateTabStop(tab.span, tab.stops, _this.defaultTabSize, pixelToPoint);
            }
        }, 500);
    };
    HtmlRenderer.prototype.later = function (func) {
        this.postRenderTasks.push(func);
    };
    return HtmlRenderer;
}());
function createElement(tagName, props, children) {
    return createElementNS(undefined, tagName, props, children);
}
function createSvgElement(tagName, props, children) {
    return createElementNS(ns.svg, tagName, props, children);
}
function createElementNS(ns, tagName, props, children) {
    var result = ns ? document.createElementNS(ns, tagName) : document.createElement(tagName);
    Object.assign(result, props);
    children && appendChildren(result, children);
    return result;
}
function removeAllElements(elem) {
    elem.innerHTML = '';
}
function appendChildren(elem, children) {
    children.forEach(function (c) { return elem.appendChild(isString(c) ? document.createTextNode(c) : c); });
}
function createStyleElement(cssText) {
    return createElement("style", { innerHTML: cssText });
}
function appendComment(elem, comment) {
    elem.appendChild(document.createComment(comment));
}
function findParent(elem, type) {
    var parent = elem.parent;
    while (parent != null && parent.type != type)
        parent = parent.parent;
    return parent;
}

var defaultOptions = {
    ignoreHeight: false,
    ignoreWidth: false,
    ignoreFonts: false,
    breakPages: true,
    debug: false,
    experimental: false,
    className: "docx",
    inWrapper: true,
    trimXmlDeclaration: true,
    ignoreLastRenderedPageBreak: true,
    renderHeaders: true,
    renderFooters: true,
    renderFootnotes: true,
    renderEndnotes: true,
    useBase64URL: false,
    renderChanges: false,
    renderComments: false
};
function parseAsync(data, userOptions) {
    var ops = __assign(__assign({}, defaultOptions), userOptions);
    return WordDocument.load(data, new DocumentParser(ops), ops);
}
function renderDocument(document, bodyContainer, styleContainer, userOptions) {
    return __awaiter(this, void 0, void 0, function () {
        var ops, renderer;
        return __generator(this, function (_a) {
            ops = __assign(__assign({}, defaultOptions), userOptions);
            renderer = new HtmlRenderer(window.document);
            renderer.render(document, bodyContainer, styleContainer, ops);
            return [2, Promise.allSettled(renderer.tasks)];
        });
    });
}
function renderAsync(data, bodyContainer, styleContainer, userOptions) {
    return __awaiter(this, void 0, void 0, function () {
        var doc;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4, parseAsync(data, userOptions)];
                case 1:
                    doc = _a.sent();
                    return [4, renderDocument(doc, bodyContainer, styleContainer, userOptions)];
                case 2:
                    _a.sent();
                    return [2, doc];
            }
        });
    });
}

export { defaultOptions, parseAsync, renderAsync, renderDocument };
