// hfy-tools@2.1.6/src/utils/date.ts
export class SimpleDate extends Date {
  constructor(d?: Date | number | string) {
    if (d) {
      super(d)
      return this
    }
    super()
    return this
  }

  /**
   * 获取当前年月日
   * @param type: y年 m月 d日
   * @param splitor -分隔符
   * @returns
   */
  getCurrentDate = (type: 'y' | 'm' | 'd', splitor?: string): string => {
    const dMap = {
      y: this.getFullYear().toString(),
      m: String(this.getMonth() + 1).padStart(2, '0'),
      d: String(this.getDate()).padStart(2, '0')
    }
    return dMap[type] || Object.values(dMap).join(splitor || '-')
  }

  /**
   * 获取当前时分秒
   * @param type: h时 m分 s秒
   * @param splitor -分隔符
   * @returns
   */
  getCurrentTime = (type: 'h' | 'm' | 's', splitor?: string): string => {
    const dMap = {
      h: this.getHours().toString().padStart(2, '0'),
      m: String(this.getMinutes()).padStart(2, '0'),
      s: String(this.getSeconds()).padStart(2, '0')
    }
    return dMap[type] || Object.values(dMap).join(splitor || ':')
  }

  getCurrentWeek() {
    const year = this.getFullYear()
    const month = this.getMonth()
    const day = this.getDate()
    const week = this.getDay()
    const off = day - week
    return [
      new Date(year, month, off + 1).toLocaleDateString(),
      new Date(year, month, off + 7).toLocaleDateString()
    ]
  }

  addYear(value: number) {
    this.setFullYear(this.getFullYear() + value)
    return this
  }
  addMonth(value: number) {
    this.setMonth(this.getMonth() + value)
    return this
  }
  addDay(value: number) {
    this.setDate(this.getDate() + value)
    return this
  }
  addHours(value: number) {
    this.setHours(this.getHours() + value)
    return this
  }
  addMinutes(value: number) {
    this.setMinutes(this.getMinutes() + value)
    return this
  }
  addSeconds(value: number) {
    this.setSeconds(this.getSeconds() + value)
    return this
  }
  subYear(value: number) {
    this.setFullYear(this.getFullYear() - value)
    return this
  }
  subMonth(value: number) {
    this.setMonth(this.getMonth() - value)
    return this
  }
  subDay(value: number) {
    this.setDate(this.getDate() - value)
    return this
  }
  subHours(value: number) {
    this.setHours(this.getHours() - value)
    return this
  }
  subMinutes(value: number) {
    this.setMinutes(this.getMinutes() - value)
    return this
  }
  subSeconds(value: number) {
    this.setSeconds(this.getSeconds() - value)
    return this
  }

  diff(type: 'y' | 'm' | 'd', diffDate: Date) {
    // 计算两个日期相差的天数
    const diffTime = Math.abs(this.getTime() - diffDate.getTime()) // 得到毫秒差值
    const diffValue = {
      y: diffTime / (1000 * 3600 * 24 * 30 * 12),
      m: diffTime / (1000 * 3600 * 24 * 30),
      d: diffTime / (1000 * 3600 * 24)
    }[type]

    return Math.round(diffValue)
  }

  format(formatString = 'yyyy-MM-dd HH:mm:ss') {
    const map = {
      yyyy: this.getFullYear(),
      MM: (this.getMonth() + 1).toString().padStart(2, '0'),
      dd: this.getDate().toString().padStart(2, '0'),
      HH: this.getHours().toString().padStart(2, '0'),
      mm: this.getMinutes().toString().padStart(2, '0'),
      ss: this.getSeconds().toString().padStart(2, '0')
    }

    return formatString.replace(/yyyy|MM|dd|HH|mm|ss/g, (match) => map[match])
  }

  toString() {
    return this.format('yyyy年MM月dd日 HH时mm分ss秒')
  }
}

export function asSecondsTime(seconds: number, format = 'hh:mm:ss') {
  const map = {
    hh: parseInt(`${(seconds / 60 / 60) % 24}`.padStart(2, '0')),
    mm: parseInt(`${(seconds / 60) % 60}`.padStart(2, '0')),
    ss: parseInt(`${seconds % 60}`.padStart(2, '0'))
  }

  return format.replace(/hh|mm|ss/g, (match) => map[match])
}

/**
 * 将时间戳转换为人性化时间
 * @param timestamp
 * @returns
 */
export const asDateText = (timestamp): string => {
  function zeroize(num) {
    return (String(num).length == 1 ? '0' : '') + num
  }
  // 当前时间戳
  let curTimestamp = parseInt(`${Date.now() / 1000}`)
  // 参数时间戳与当前时间戳相差秒数
  let timestampDiff = curTimestamp - timestamp
  // 当前时间日期对象
  let curDate = new Date(curTimestamp * 1000)
  // 参数时间戳转换成的日期对象
  let tmDate = new Date(timestamp * 1000)
  let Y = tmDate.getFullYear(),
    m = tmDate.getMonth() + 1,
    d = tmDate.getDate()
  let H = tmDate.getHours(),
    i = tmDate.getMinutes(),
    s = tmDate.getSeconds()
  if (timestampDiff < 60) {
    // 一分钟以内
    return '刚刚'
  } else if (timestampDiff < 3600) {
    // 一小时前之内
    return Math.floor(timestampDiff / 60) + '分钟前'
  } else if (curDate.getFullYear() == Y && curDate.getMonth() + 1 == m && curDate.getDate() == d) {
    return '今天' + zeroize(H) + ':' + zeroize(i)
  } else {
    let newDate = new Date((curTimestamp - 86400) * 1000)
    // 参数中的时间戳加一天转换成的日期对象
    if (newDate.getFullYear() == Y && newDate.getMonth() + 1 == m && newDate.getDate() == d) {
      return '昨天' + zeroize(H) + ':' + zeroize(i)
    } else if (curDate.getFullYear() == Y) {
      return zeroize(m) + '月' + zeroize(d) + '日 ' + zeroize(H) + ':' + zeroize(i)
    } else {
      return Y + '年' + zeroize(m) + '月' + zeroize(d) + '日 ' + zeroize(H) + ':' + zeroize(i)
    }
  }
}
