// hfy-tools@2.1.6/src/utils/file.ts
/**
 * 打开文件选择对话框
 * 若浏览器不支持实验性方法：window.showOpenFilePicker，则采用input[type=file]元素进行兼容
 * @param {string | string[]} accept 文件类型限制 ，默认全部
 * @param {boolean} multiple 文件多选
 * @param {boolean} forceInput 强制input
 * @param {boolean} webkitdirectory 只选择目录限制
 * @param {number} compatible 兼容模式，默认开启
 * @param {number} cancel 兼容取消控制，为0时候则取消文件时不抛出reject，❗在使用async/await时会造成阻塞
 * @param {string} description 文件或者文件夹的描述，可选
 * @return {Promise<FileList>}
 */
export interface IChooseFileOptions {
  forceInput?: boolean
  accept?: string | []
  cancel?: number
  compatible?: boolean
  multiple?: boolean
  webkitdirectory?: boolean
  description?: string
}

export async function overwrite(content: string, options: IChooseFileOptions) {
  if (!options.compatible && window.hasOwnProperty('showOpenFilePicker')) {
    let { accept = '*', multiple, description } = options

    const acceptMap: { [accept: string]: string[] } = {}
    for (let a of (accept as string).split(',')) {
      acceptMap[a] = []
    }
    //@ts-ignore
    const fileHandleList = await window.showOpenFilePicker?.({
      multiple,
      excludeAcceptAllOption: false,
      types: [
        {
          description,
          accept: acceptMap,
        },
      ],
    })

    for (const fileHandle of fileHandleList) {
      // 创建一个可写流
      const writable = await fileHandle.createWritable()

      // 写入数据
      await writable.write(content)

      // 关闭流
      await writable.close()
    }
  }

  throw new Error('浏览器不支持!')
}

export async function chooseFile(options: IChooseFileOptions): Promise<File[]> {
  let {
    accept = '*',
    compatible = true,
    cancel = 300,
    multiple,
    webkitdirectory,
    description,
  } = options

  if (options.forceInput !== true) {
    accept.constructor === Array && (accept = accept.join(','))
    // 实验性功能
    if (!compatible && window.hasOwnProperty('showOpenFilePicker')) {
      const files = []
      const acceptMap: { [accept: string]: string[] } = {}
      for (let a of (accept as string).split(',')) {
        acceptMap[a] = []
      }
      //@ts-ignore
      const fileHandleList = await window.showOpenFilePicker?.({
        multiple,
        excludeAcceptAllOption: false,
        types: [
          {
            description,
            accept: acceptMap,
          },
        ],
      })
      for (const f of fileHandleList) {
        files.push(await f.getFile())
      }
      return files
    }
  }

  const inpEle = document.createElement('input')
  inpEle.id = `__file_${Math.trunc(Math.random() * 100000)}`
  inpEle.type = 'file'
  inpEle.style.display = 'none'
  // 文件类型限制
  inpEle.accept = accept as string
  // 多选限制
  multiple && (inpEle.multiple = multiple)
  // 选择目录限制
  if (webkitdirectory) {
    console.warn(
      '该特性是非标准的，请尽量不要在生产环境中使用它！\n' +
        'This feature is non-standard, so try not to use it in a production environment!',
    )
    inpEle.webkitdirectory = webkitdirectory
  }
  inpEle.click()

  return await new Promise((resolve, reject) => {
    let _isSelected = false
    const changeEvent = () => {
      const files = inpEle.files
      if (files) {
        _isSelected = true
        resolve(Array.from(files))
      }
    }
    const focusEvent = (event: Event) => {
      if (event.target?.constructor === Window) {
        setTimeout(() => {
          !_isSelected && reject('未选定文件\nUnselected file')
        }, cancel)
      }
    }
    inpEle.addEventListener('change', changeEvent, { once: true })
    cancel && window.addEventListener('focus', focusEvent, { once: true })
  })
}

export function downloadFile(blob: File, downloadName: string) {
  const a = document.createElement('a')
  a.href = URL.createObjectURL(blob)
  // window.open(URL.createObjectURL(res.data));
  a.download = downloadName
  a.click()
  URL.revokeObjectURL(a.href)
  a.remove()
}

// 枚举类型定义文件类型
export const FILE_TYPES = Object.freeze({
  AUDIO: 'audio',
  VIDEO: 'video',
  OFFICE: 'office',
  IMAGE: 'image',
  MODEL: 'model',
  CAD: 'cad', // 图纸文件
  ARCHIVE: 'archive',
  UNKNOWN: 'unknown',
})

// 判断文件类型函数
export function getFileType(fileName) {
  const audioExtensions = ['.mp3', '.wav', '.flac', '.aac', '.ogg']
  const videoExtensions = ['.mp4', '.avi', '.mov', '.flv', '.mkv', '.webm']
  const officeExtensions = [
    '.doc',
    '.docx',
    '.xls',
    '.xlsx',
    '.ppt',
    '.pptx',
    '.pdf',
  ]
  const imageExtensions = ['.jpg', '.jpeg', '.png', '.gif', '.bmp', '.webp']
  const modelExtensions = ['.obj', '.fbx', '.glb', '.stl', '.ply'] // 常见模型文件格式
  const cadExtensions = ['.dwg', '.dxf', '.dwt', '.stp', '.step'] // 图纸文件格式
  const archiveExtensions = ['.zip', '.tar', '.rar', '.gz', '.7z']

  // 获取文件扩展名（忽略大小写）
  const fileExtension = fileName
    .slice(((fileName.lastIndexOf('.') - 1) >>> 0) + 2)
    .toLowerCase()

  // 根据扩展名判断文件类型
  if (audioExtensions.includes(`.${fileExtension}`)) {
    return FILE_TYPES.AUDIO
  } else if (videoExtensions.includes(`.${fileExtension}`)) {
    return FILE_TYPES.VIDEO
  } else if (officeExtensions.includes(`.${fileExtension}`)) {
    return FILE_TYPES.OFFICE
  } else if (imageExtensions.includes(`.${fileExtension}`)) {
    return FILE_TYPES.IMAGE
  } else if (modelExtensions.includes(`.${fileExtension}`)) {
    return FILE_TYPES.MODEL
  } else if (cadExtensions.includes(`.${fileExtension}`)) {
    return FILE_TYPES.CAD
  } else if (archiveExtensions.includes(`.${fileExtension}`)) {
    return FILE_TYPES.ARCHIVE
  } else {
    return FILE_TYPES.UNKNOWN
  }
}

export async function urlToBlob(url: string) {
  try {
    // 使用 fetch 获取文件内容
    const response = await fetch(url)

    // 检查响应是否成功
    if (!response.ok) {
      throw new Error('网络错误，无法获取文件')
    }

    // 将响应体转换为 Blob
    const blob = await response.blob()

    // 返回 Blob 对象
    return blob
  } catch (error) {
    console.error('获取 URL 失败:', error)
  }
}
