// hfy-tools@2.1.6/src/utils/oss.ts
import COS, {
  CosError,
  DeleteMultipleObjectResult,
  GetBucketResult,
  GetObjectResult,
  GetObjectUrlResult,
  PutObjectCopyResult,
  SliceUploadFileResult
} from 'cos-js-sdk-v5'
import * as core from './core'
import * as base64 from './base64'
import * as fetch from './fetch'
import { SimpleDate } from './date'
export interface IFileJSON {
  key: string
  dir: string
  name: string
  isDir?: string
}
class OSS {
  Bucket: string
  Region: string
  SecretKey: string
  SecretId: string
  prefix: string
  isTemp: boolean
  cos: any
  params: { Bucket: string; Region: string }
  constructor({
    Bucket,
    Region,
    SecretKey,
    SecretId,
    prefix,
    isTemp
  }: {
    Bucket: string
    Region: string
    SecretKey: string
    SecretId: string
    prefix?: string
    isTemp?: boolean
  }) {
    this.Bucket = Bucket
    this.Region = Region
    this.SecretKey = SecretKey
    this.SecretId = SecretId
    this.prefix = prefix
    this.isTemp = isTemp
    this.prefix = prefix || base64.encode('GOSS')
    this.cos = new COS({
      SecretId: SecretId,
      SecretKey: SecretKey
    })

    this.params = { Bucket, Region }
  }

  /**
   *
   * @param file 文件
   * @param key 上传文件key
   * @param progress 上传进度函数回调
   * @param dir 目录
   * @returns
   */
  upload({
    file,
    dir,
    progress,
    key,
    expirationDate
  }: {
    file: File
    dir?: string
    progress?: Function
    key?: string
    expirationDate?: Date
  }): Promise<SliceUploadFileResult> {
    return new Promise((resolve, reject) => {
      let uploadKey = base64.encode(
        JSON.stringify({ name: file.name, id: key || core.guid(), dir })
      )
      // console.log('文件上传', uploadKey, this.isTemp, file)
      if (this.isTemp) {
        const d = new SimpleDate()
        d.addSeconds(3)
        uploadKey = `temp/${expirationDate?.getTime() || d.getTime()}/` + uploadKey
      }
      // console.log('上传的key', `${pre}/${key}`)
      this.cos.sliceUploadFile(
        {
          ...this.params,
          Key: `${this.prefix}/${uploadKey}`,
          StorageClass: 'STANDARD', // [Standard, Standard_IA，Archive]
          Body: file, // 上传文件对象
          onProgress: progress
        },
        (err: CosError, data: SliceUploadFileResult) => {
          if (err) {
            reject(err)
            return
          }
          resolve(data)
        }
      )
    })
  }

  /**
   * 批量删除文件
   * @param keys 文件key
   * @returns
   */
  multiDelete(...keys: string[]): Promise<DeleteMultipleObjectResult> {
    return new Promise((resolve, rej) => {
      this.cos.deleteMultipleObject(
        {
          ...this.params,
          Objects: keys.map((o) => ({ Key: o }))
        },
        (err: CosError, data: DeleteMultipleObjectResult): void => {
          resolve(data)
        }
      )
    })
  }

  /**
   * 获取文件
   * @param key
   * @param progress
   * @param isGetUrl
   * @returns
   */
  getFile(
    key: string,
    progress: Function,
    isGetUrl: boolean = true
  ): Promise<GetObjectUrlResult | GetObjectResult> {
    return new Promise((resolve, reject) => {
      // console.log(key, progress, isGetUrl);
      if (key.startsWith(this.prefix + '/temp')) {
        const expireDate = new Date(key.split('/')[2])
        if (expireDate <= new Date()) {
          this.multiDelete(key)
          reject('文件已过期')
          return
        }
      }
      if (isGetUrl) {
        this.cos.getObjectUrl(
          {
            ...this.params,
            Key: `${key}`,
            Sign: true
          },
          (err: CosError, data: GetObjectUrlResult): void => {
            if (err) {
              reject(new Error(err.message))
              return
            }
            resolve(data)
          }
        )
        return
      }

      this.cos.getObject(
        {
          ...this.params,
          DataType: 'blob',
          Key: `${key}`
        },
        (err: CosError, data: GetObjectResult) => {
          if (err) {
            reject(new Error(err.message))
            return
          }
          resolve(data)
        }
      )
    })
  }

  /**
   * 文件拷贝
   * @param key
   * @param sourceKey
   * @param pms
   * @returns
   */
  copy(key: string, sourceKey: string, pms = {}): Promise<PutObjectCopyResult> {
    return new Promise((resolve, reject) => {
      key = base64.encode(JSON.stringify({ key, ...pms }))
      this.cos.putObjectCopy(
        {
          ...this.params,
          Key: `${this.prefix}/${key}`,
          CopySource: `${this.params.Bucket}.cos.ap-guangzhou.myqcloud.com/${sourceKey}`
        },
        (err: CosError, data: PutObjectCopyResult) => {
          if (err) {
            reject(new Error(err.message))
            return
          }
          resolve(data)
        }
      )
    })
  }

  /**
   * 根据前缀获取文件
   * @param pre
   * @returns
   */
  getFiles<T extends IFileJSON = IFileJSON>(): Promise<T[]> {
    return new Promise((resolve, reject) => {
      this.cos.getBucket(
        {
          ...this.params,
          Prefix: `${this.prefix}`
        },
        (err: CosError, data: GetBucketResult) => {
          if (err) {
            reject(new Error(err.message))
            resolve([])
            return
          }

          const result: any[] = data.Contents.reduce<any[]>((res, o) => {
            // console.log('contents', o, o.Key.split('/')[2])
            if (o.Key.startsWith(`${this.prefix}/temp`)) {
              const expireDate = new Date(Number(o.Key.split('/')[2]))
              if (expireDate <= new Date()) {
                this.multiDelete(o.Key)
                reject('文件已过期')
                return res
              }
            }
            let valueStr = o.Key.split('/').pop()
            valueStr = base64.decode(`${valueStr}`)
            try {
              res.push({ ...JSON.parse(valueStr), key: o.Key })
            } catch (error) {
              console.warn(error)
              res.push({ name: '解析失败', key: o.Key })
            }
            return res
          }, [])

          resolve(result)
        }
      )
    })
  }

  async getDirectory() {
    const result: any[] = await this.getFiles()
    return this.buildDirectory(result)
  }

  buildDirectory(files: any[]) {
    return files.reduce<any>((res, cur) => {
      let dirName = cur.id
      if (cur.dir) {
        dirName = cur.dir
      }
      res[dirName] = [cur].concat(res[dirName] ?? [])

      return res
    }, {})
  }
}

class DB {
  OSS: OSS
  constructor(ossParams: {
    Bucket: string
    Region: string
    SecretKey: string
    SecretId: string
    prefix?: string
    isTemp?: boolean
  }) {
    this.OSS = new OSS({
      ...ossParams,
      prefix: ossParams.prefix ?? base64.encode('GOSS') + '/table'
    })
  }

  add<T>(name: string, data: any) {
    return new Promise<T>(async (resolve, reject) => {
      const list = await this.get<any[]>(name)
      list.push({ ...data, guid: core.guid() })
      const blob = new File([JSON.stringify(list)], name + '.json', {
        type: 'application/json'
      })

      this.OSS.upload({ file: blob, dir: '', key: '' })
    })
  }

  update(name: string, id: string, data: any) {
    return new Promise(async (resolve, reject) => {
      const list = await this.get<any[]>(name)
      const idx = list.findIndex((o) => Object.is(o.guid, id))
      list.splice(idx, 1, { ...data, guid: id })
      this.add(name, list)
    })
  }

  delete(name: string, id: string) {
    return new Promise(async (resolve, reject) => {
      const list = await this.get<any[]>(name)
      this.add(
        name,
        list.filter((o) => o.guid !== id)
      )
    })
  }

  get<T>(name: string): Promise<T> {
    let uploadKey = base64.encode(JSON.stringify({ name: name + '.json', id: '', dir: '' }))
    let prefix = base64.encode('GOSS')

    return fetch.get<T>(
      `https://jf-temp-1301446188.cos.ap-guangzhou.myqcloud.com/${prefix}/table/${uploadKey}`
    ) as Promise<T>
  }

  list() {
    return new Promise((resolve) => {
      this.OSS.getFiles().then(resolve)
    })
  }
}

export default {
  COS: OSS,
  DB
}
