// hfy-tools@2.1.6/src/utils/tree.ts
/**
 * 根据 id 查找节点
 * @param {Array} treeData 树数据
 * @param {string} id 要查找的节点 id
 * @param {string} key 要查找的节点 key
 * @returns {Object|null} 返回节点对象或 null
 */

type INode<T extends string = string, V extends any = any> = Record<T, V> & {
  children?: INode<T, V>[]
}

export function findNodeById<
  K extends string,
  I = string | number | symbol,
  T extends INode<K> = INode<K>,
>(treeData: T[], val: I, key: K = 'id' as K, childName = 'children'): T | null {
  for (let node of treeData) {
    // @ts-ignore
    if (Object.is(node[key], val)) {
      return node as T
    }
    if (node[childName]) {
      const foundNode = findNodeById(node[childName], val, key)
      if (foundNode) {
        return foundNode as T
      }
    }
  }
  return null
}

/**
 * 根据当前节点 id 返回所有父节点 id 路径
 * @param {Array} treeData 树数据
 * @param {string} targetId 目标节点的 id
 * @param {string} idField 节点id 字段
 * @param {string} parentField 父节点 字段
 * @returns {Array} 返回从目标节点到根节点的 id 路径
 */
export function findParentPathById<
  I extends string,
  P extends string,
  X extends string | number | symbol,
  T extends INode<I | P> = INode<I | P>,
>(
  treeData: T[],
  targetId: any,
  idField: I = 'id' as I,
  parentField: P = 'pid' as P,
): X[] | null {
  for (let node of treeData) {
    if (Object.is(node[idField], targetId)) {
      const path = [node[idField]] // 初始化路径数组，包含当前节点的 id
      let parentId = node[parentField]

      // 逐级回溯父节点
      while (parentId !== null) {
        const parentNode = findNodeById(treeData, parentId, idField) // 查找父节点
        if (parentNode) {
          path.unshift(parentNode[idField]!) // 将父节点 id 加到路径前面
          parentId = parentNode[parentField]! // 更新父节点 id
        } else {
          break // 如果找不到父节点，结束回溯
        }
      }

      return path // 返回从目标节点到根节点的路径
    }

    // 如果有子节点，递归查找

    if (node.children) {
      const result = findParentPathById(
        node.children,
        targetId,
        idField,
        parentField,
      )
      if (result) {
        return [node[idField], ...(result as X[])] // 拼接当前节点和子节点路径
      }
    }
  }

  // 如果没有找到目标节点，返回 null
  return null
}

/**
 * 根据当前节点 id 返回所有父节点 对象
 * @param {Array} treeData 树数据
 * @param {string} targetId 目标节点的 id
 * @param {string} idField 节点id 字段
 * @param {string} parentField 父节点 字段
 * @returns {Array} 返回从目标节点到根节点的 id 路径
 */
export function findParentObjectById<
  I extends string,
  P extends string,
  X extends string | number | symbol,
  T extends INode<I | P> = INode<I | P>,
>(
  treeData: T[],
  targetId: any,
  idField: I = 'id' as I,
  parentField: P = 'pid' as P,
): T[] {
  for (let node of treeData) {
    if (Object.is(node[idField], targetId)) {
      const parents: T[] = [node] // 初始化路径数组，包含当前节点的 id
      let parentId: any = node[parentField]

      // 逐级回溯父节点
      while (parentId !== null) {
        const parentNode = findNodeById(treeData, parentId, idField) // 查找父节点
        if (parentNode) {
          parents.unshift(parentNode) // 将父节点 id 加到路径前面
          parentId = parentNode[parentField] // 更新父节点 id
        } else {
          break // 如果找不到父节点，结束回溯
        }
      }

      return parents // 返回从目标节点到根节点的路径
    }

    // 如果有子节点，递归查找

    if (node.children) {
      const result = findParentObjectById(
        node.children,
        targetId,
        idField,
        parentField,
      ) as T[]
      return [node, ...result] // 拼接当前节点和子节点路径
    }
  }

  // 如果没有找到目标节点
  return []
}

export function mapTree<T extends INode = INode>(
  treeData: T[],
  callback: (c: T) => T,
): T[] {
  return treeData?.map((o) => {
    const node = callback(o)
    if (node) {
      return { ...node, children: mapTree(o.children as T[], callback) }
    }
    return node
  })
}

export function filterTree<T extends INode = INode>(
  treeData: T[],
  callback: (c: T) => boolean,
): T[] {
  return treeData.filter((o) => {
    if (o.children) {
      o.children = filterTree(o.children as T[], callback)
    }
    return callback(o)
  })
}

export function containsValueTree<F extends string, V extends any = string>(
  tree: INode<F, V>[],
  field: F,
  value: V = '' as V,
): boolean {
  return tree.some((o) => {
    if (Object.is(o[field], value)) {
      return true
    }
    if (o.children) {
      return containsValueTree(o.children, field, value)
    }

    return false
  })
}

// // 数组转树
export function arrayToTree<
  I extends string,
  P extends string,
  T extends any[] = Record<I | P, never>[],
>(
  array: T,
  keyField: I = 'id' as I,
  parentField: P = 'pid' as P,
  rootValue = 0,
): (Record<I | P, never> & { children: Record<I | P, never>[] })[] {
  return array
    .filter((o) => Object.is(o[parentField], rootValue))
    .map((o) => {
      return {
        ...o,
        children: arrayToTree(array, keyField, parentField, o[keyField]),
      }
    })
}

export interface TreeNode {
  id: string | number
  parentId: string | number | null
  children?: TreeNode[]
  [key: string]: any
}

export interface TreeOptions {
  idKey?: string
  parentKey?: string
  childrenKey?: string
  rootValue?: string | number | null
}

export function strongArrayToTree<T extends TreeNode>(
  items: T[],
  options: TreeOptions = {},
): T[] {
  const {
    idKey = 'id',
    parentKey = 'parentId',
    childrenKey = 'children',
    rootValue = null,
  } = options

  const nodeMap = new Map<string | number, T>()
  const tree: T[] = []

  items.forEach((item) => {
    nodeMap.set(item[idKey], {
      ...item,
      [childrenKey]: [],
    } as T)
  })

  items.forEach((item) => {
    const node = nodeMap.get(item[idKey])!
    const parentId = item[parentKey]

    if (parentId === rootValue || !nodeMap.has(parentId)) {
      tree.push(node)
    } else {
      const parent = nodeMap.get(parentId)
      if (parent) {
        parent[childrenKey]!.push(node)
      }
    }
  })

  return tree
}
