// hfy-tools@2.1.6/dist/utils/core.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.optionsToNumber = exports.numberToOptionsText = exports.numberToOptions = exports.EventBus = exports.LRUCache = exports.throttle = exports.camelToKebab = exports.kebabToCamel = exports.isFunction = void 0;
exports.guid = guid;
exports.deepClone = deepClone;
exports.setPathValue = setPathValue;
exports.getPathValue = getPathValue;
exports.getLocationByIP = getLocationByIP;
exports.getCurrentLocation = getCurrentLocation;
exports.chineseToHashCode = chineseToHashCode;
exports.jsonToUrlParams = jsonToUrlParams;
exports.safeParseJson = safeParseJson;
exports.safeStringifyJson = safeStringifyJson;
/**
 * 获取uuid
 * @returns
 */
function guid() {
    function S4() {
        return (((1 + Math.random()) * 0x10000) | 0).toString(16).substring(1);
    }
    return (S4() +
        S4() +
        '-' +
        S4() +
        '-' +
        S4() +
        '-' +
        S4() +
        '-' +
        S4() +
        S4() +
        S4());
}
/**
 * 是一个函数
 * @param v
 * @returns
 */
const isFunction = (v) => [
    '[object Function]',
    '[object GeneratorFunction]',
    '[object AsyncFunction]',
    '[object Promise]',
].includes(Object.prototype.toString.call(v));
exports.isFunction = isFunction;
/**
 * 深拷贝
 * @param obj
 * @returns
 */
function deepClone(obj) {
    if (typeof obj !== 'object' || obj == null) {
        return obj;
    }
    let result;
    if (obj instanceof Array) {
        result = [];
    }
    else {
        result = {};
    }
    for (let key in obj) {
        // 保证 key 不是原型的属性
        if (obj.hasOwnProperty(key)) {
            // 递归调用！！！
            result[key] = deepClone(obj[key]);
        }
    }
    // 返回结果
    return result;
}
/**
 * 大驼峰
 * @param str
 * @returns
 */
const kebabToCamel = (str) => str.replace(/-./g, (m) => m.toUpperCase()[1]);
exports.kebabToCamel = kebabToCamel;
/**
 * 小驼峰
 * @param str
 * @returns
 */
const camelToKebab = (str) => str.replace(/([a-z0-9])([A-Z])/g, '$1-$2').toLowerCase();
exports.camelToKebab = camelToKebab;
/**
 * 设置值 避免属性名不存在的情况 a.b.c
 * @param target
 * @param property
 * @param value
 * @returns
 */
function setPathValue(target, property, value) {
    const propertyArr = property.split('.');
    return propertyArr.reduce((pre, cur, index) => {
        if (index === propertyArr.length - 1) {
            pre[cur] = value;
            return target;
        }
        pre[cur] = {};
        return pre[cur];
    }, target);
}
function getPathValue(target, property, defaultValue) {
    const propertyArr = property.split('.');
    return propertyArr.reduce((result, cur, index) => {
        if (index === propertyArr.length - 1) {
            return result[cur] ?? defaultValue;
        }
        return result[cur];
    }, target);
}
/**
 * 节流
 * @param fn
 * @param delay
 * @returns
 */
const throttle = (fn, delay) => {
    let timer;
    let prevTime;
    return function (...args) {
        let currTime = Date.now();
        let context = this;
        if (!prevTime)
            prevTime = currTime;
        clearTimeout(timer);
        if (currTime - prevTime > delay) {
            prevTime = currTime;
            fn.apply(context, args);
            clearTimeout(timer);
            return;
        }
        timer = setTimeout(function () {
            prevTime = Date.now();
            timer = null;
            fn.apply(context, args);
        }, delay);
    };
};
exports.throttle = throttle;
/**
 * LRU缓存
 */
class LRUCache {
    cache;
    size;
    constructor(size) {
        this.size = size;
        this.cache = new Map();
    }
    get(key) {
        const hasKey = this.cache.has(key);
        if (hasKey) {
            const val = this.cache.get(key);
            this.cache.delete(key);
            this.cache.set(key, val);
            return val;
        }
        else {
            return -1;
        }
    }
    set(key, value) {
        const hasKey = this.cache.has(key);
        if (hasKey) {
            this.cache.delete(key);
        }
        this.cache.set(key, value);
        if (this.cache.size > this.size) {
            // this.cache.keys().next().value用于获取最前面（即使用次数最少的key）,然后删除
            this.cache.delete(this.cache.keys().next().value);
        }
    }
}
exports.LRUCache = LRUCache;
/**
 * 事件分发
 */
class EventBus {
    eventObject;
    onceKeys;
    constructor() {
        // 初始化事件列表
        this.eventObject = {};
        this.onceKeys = [];
    }
    //发布事件
    publish(eventName, ...args) {
        Object.values(this.eventObject[eventName] || {}).forEach((o) => o instanceof Function && o(...args));
        let event = this.onceKeys.find((o) => o.em === eventName);
        event && delete this.eventObject[event.em]?.[event.fm];
    }
    //订阅事件
    subscribe(eventName, fn = () => { }, once = false) {
        once && this.onceKeys.push({ em: eventName, fm: fn });
        this.eventObject[eventName] ??= {};
        this.eventObject[eventName][fn] = fn;
        return () => {
            delete this.eventObject[eventName][fn];
        };
    }
    //清除事件
    clear(eventName) {
        if (eventName) {
            delete this.eventObject[eventName];
        }
        else {
            this.eventObject = {};
        }
    }
}
exports.EventBus = EventBus;
// 位运算 转换函数
const numberToOptions = (num, options) => {
    if (!num || isNaN(num))
        return [];
    return options
        .filter((item) => (num & item.value) === item.value)
        .map((o) => o.value);
};
exports.numberToOptions = numberToOptions;
// 位运算 转换函数
const numberToOptionsText = (num, options) => {
    if (!num || isNaN(num))
        return [];
    return options
        .filter((item) => (num & item.value) === item.value)
        .map((o) => o.label)
        .join(',');
};
exports.numberToOptionsText = numberToOptionsText;
// 位运算 合并
const optionsToNumber = (options) => {
    if (!options || !Array.isArray(options))
        return 0;
    return options.reduce((acc, cur) => acc | cur, 0);
};
exports.optionsToNumber = optionsToNumber;
// 使用IP地址获取大致位置（精度较低）
async function getLocationByIP() {
    try {
        const response = await fetch('https://ipapi.co/json/');
        const data = await response.json();
        return {
            latitude: data.latitude,
            longitude: data.longitude,
            accuracy: 5000, // IP定位精度通常为5km左右
            source: 'ip',
        };
    }
    catch (error) {
        console.error('IP定位失败:', error);
        return null;
    }
}
// 使用前面定义的getCurrentLocation函数
function getCurrentLocation(options = {}) {
    return new Promise((resolve, reject) => {
        if (!('geolocation' in navigator)) {
            reject(new Error('您的浏览器不支持地理定位'));
            return;
        }
        console.log('获取', navigator);
        const rs = navigator.geolocation.getCurrentPosition((position) => {
            console.warn(position, 'position');
            resolve(position.coords);
        }, async (error) => {
            console.warn(error, 'position watch');
            resolve(await getLocationByIP());
        }, 
        // options
        {
            enableHighAccuracy: true,
            timeout: 10000,
            maximumAge: 0,
        });
    });
}
function chineseToHashCode(str) {
    let hash = 0;
    for (let i = 0; i < str.length; i++) {
        hash = (hash << 5) - hash + str.charCodeAt(i);
        hash |= 0; // 转换为32位整数
    }
    return Math.abs(hash);
}
function jsonToUrlParams(json) {
    const params = new URLSearchParams();
    for (const [key, value] of Object.entries(json)) {
        if (Array.isArray(value)) {
            // 如果值是数组，循环处理每个元素
            value.forEach((val) => params.append(key, val));
        }
        else {
            // @ts-ignore
            params.append(key, value);
        }
    }
    return params.toString();
}
/**
 * 安全地将字符串解析为 JSON
 * @param jsonString 要解析的 JSON 字符串
 * @param defaultValue 解析失败时返回的默认值
 * @returns 解析后的对象或默认值
 */
function safeParseJson(jsonString, defaultValue = null) {
    try {
        return jsonString ? JSON.parse(jsonString) : defaultValue;
    }
    catch (error) {
        console.error('JSON 解析失败:', error);
        return defaultValue;
    }
}
/**
 * 安全地将对象序列化为 JSON 字符串
 * @param value 要序列化的值
 * @param defaultValue 序列化失败时返回的默认字符串
 * @returns JSON 字符串或默认值
 */
function safeStringifyJson(value, defaultValue = 'null') {
    try {
        return JSON.stringify(value);
    }
    catch (error) {
        console.error('JSON 序列化失败:', error);
        return defaultValue;
    }
}
