// hfy-tools@2.1.6/dist/utils/echart.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.getMapOption = void 0;
const core = __importStar(require("./core"));
const lineTopList = [];
// 全国各省份数据示例
const districtData = [
    { name: '北京', id: core.guid(), value: 267, point: [116.405285, 39.904989] },
    { name: '上海', id: core.guid(), value: 200, point: [121.472644, 31.231706] },
    { name: '广东', id: core.guid(), value: 180, point: [113.280637, 23.125178] },
    { name: '浙江', id: core.guid(), value: 150, point: [120.153576, 30.287459] },
    { name: '江苏', id: core.guid(), value: 130, point: [118.767413, 32.041546] },
    { name: '山东', id: core.guid(), value: 120, point: [117.000923, 36.675807] },
    { name: '四川', id: core.guid(), value: 110, point: [104.065735, 30.659462] },
    { name: '河南', id: core.guid(), value: 100, point: [113.665412, 34.757975] },
    { name: '湖北', id: core.guid(), value: 90, point: [114.298572, 30.584355] },
    { name: '福建', id: core.guid(), value: 80, point: [119.306239, 26.075302] },
];
// 渐变层颜色
const colorList = [
    '#8b5e70',
    '#81596d',
    '#78556a',
    '#6e5068',
    '#644c65',
    '#5b4762',
    '#51435f',
    '#483e5c',
    '#3e3a59',
    '#343557',
    '#2b3154',
    '#212c51',
];
// 生成渐变图层
const geoList = [];
for (let i = 1; i <= colorList.length; i++) {
    const mapOption = {
        map: 'china',
        aspectScale: 0.85,
        emphasis: {
            disabled: true,
        },
        z: 12 - i,
        layoutCenter: ['50%', `${i * 0.3 + 50}%`],
        layoutSize: '100%',
        itemStyle: {
            normal: {
                areaColor: colorList[i - 1],
                borderWidth: 0,
            },
        },
    };
    if (i === colorList.length) {
        mapOption.itemStyle.normal.shadowColor = 'rgba(0, 0, 0, 0.71)';
        mapOption.itemStyle.normal.shadowBlur = 100;
    }
    geoList.push(mapOption);
}
// 生成地图数据柱数据
const getLineData = () => {
    const lineSeriesData = [];
    const maxValue = Math.max(...districtData.map((item) => item.value));
    districtData.forEach((item, index) => {
        // 柱子
        const lineData = {
            type: 'lines',
            zlevel: 5,
            effect: {
                show: false,
                symbolSize: 5,
            },
            lineStyle: {
                width: 10,
                color: {
                    type: 'linear',
                    x: 1,
                    y: 0,
                    x2: 0,
                    y2: 0,
                    colorStops: [
                        {
                            offset: 0,
                            color: 'rgba(232, 204, 149, 1)',
                        },
                        {
                            offset: 1,
                            color: 'rgba(170, 144, 91, 1)',
                        },
                    ],
                    global: false,
                },
                opacity: 1,
                curveness: 0,
            },
            label: {
                show: 0,
                position: 'end',
                formatter: '245',
            },
            silent: true,
            data: [
                {
                    ...item,
                    coords: [
                        item.point,
                        [item.point[0], item.point[1] + (item.value / maxValue) * 0.5],
                    ], // 调整高度比例
                },
            ],
        };
        // 柱子顶部
        const lineTop = {
            type: 'scatter',
            coordinateSystem: 'geo',
            geoIndex: 0,
            zlevel: 5,
            label: {
                show: false,
            },
            symbol: 'circle',
            symbolSize: [10, 5],
            itemStyle: {
                color: 'rgba(255, 255, 179, 1)',
            },
            silent: true,
            data: [
                [item.point[0], item.point[1] + (item.value / maxValue) * 0.5],
                [item.point[0], item.point[1] + (item.value / maxValue) * 0.5],
            ],
        };
        // 柱子底部
        const lineBottom = {
            type: 'scatter',
            coordinateSystem: 'geo',
            geoIndex: 0,
            zlevel: 4,
            label: {
                show: false,
            },
            symbol: 'circle',
            symbolSize: [10, 5],
            itemStyle: {
                color: {
                    type: 'linear',
                    x: 1,
                    y: 0,
                    x2: 0,
                    y2: 0,
                    colorStops: [
                        {
                            offset: 0,
                            color: 'rgba(232, 204, 149, 1)',
                        },
                        {
                            offset: 1,
                            color: 'rgba(170, 144, 91, 1)',
                        },
                    ],
                    global: false,
                },
            },
            silent: true,
            data: [item.point],
        };
        // 底部光圈
        const lineBottomCircle = {
            name: 'Top ' + (index + 1),
            type: 'effectScatter',
            coordinateSystem: 'geo',
            data: [
                {
                    name: item.name,
                    value: [item.point[0], item.point[1], item.value],
                },
            ],
            showEffectOn: 'render',
            rippleEffect: {
                scale: 5,
                brushType: 'stroke',
            },
            label: {
                normal: {
                    formatter: '{b}',
                    position: 'bottom',
                    show: false,
                    color: '#fff',
                    distance: 10,
                },
            },
            symbol: 'circle',
            symbolSize: [20, 10],
            itemStyle: {
                normal: {
                    color: 'rgba(232, 204, 149, 1)',
                    shadowBlur: 10,
                    shadowColor: 'rgba(232, 204, 149, 1)',
                },
                opacity: 1,
            },
            zlevel: 4,
        };
        // 顶部图标（只显示前5个省份的顶部图标）
        let lineTopIcon = null;
        if (index < 5) {
            lineTopIcon = {
                type: 'scatter',
                coordinateSystem: 'geo',
                geoIndex: 0,
                zlevel: 5,
                label: {
                    normal: {
                        show: true,
                        formatter: function (params) {
                            return `{cityName|${params.name}}\n {value|${params.data.data}} {unit|个}`;
                        },
                        rich: {
                            cityName: {
                                color: 'rgba(201, 211, 234, 1)',
                                fontSize: 14,
                                padding: [6, 0, 4, 48],
                            },
                            value: {
                                color: 'rgba(255, 187, 94, 1)',
                                fontSize: 18,
                                fontWeight: 800,
                                padding: [0, 0, 0, 44],
                            },
                            unit: {
                                color: 'rgba(255, 187, 94, 1)',
                                fontSize: 14,
                            },
                        },
                    },
                    emphasis: {
                        show: true,
                    },
                },
                symbol: `image://${lineTopList[index]}`,
                symbolSize: [143, 48],
                symbolOffset: [0, 0],
                z: 999,
                data: [
                    {
                        name: item.name,
                        data: item.value,
                        value: [
                            item.point[0],
                            item.point[1] + (item.value / maxValue) * 0.5 + 0.2,
                        ],
                    },
                ],
            };
        }
        lineSeriesData.push(lineData);
        lineSeriesData.push(lineTop);
        lineSeriesData.push(lineBottom);
        lineSeriesData.push(lineBottomCircle);
        if (lineTopIcon) {
            lineSeriesData.push(lineTopIcon);
        }
    });
    console.log(lineSeriesData, 'lineSeriesData');
    return lineSeriesData;
};
// 获取地图配置
const getMapOption = (name) => {
    const lineSeriesData = getLineData();
    const option = {
        animation: false,
        progressive: 100,
        useGPU: true,
        silent: false, // 改为false以启用交互
        geo: {
            roam: true,
            map: name || 'china',
            aspectScale: 0.85,
            layoutCenter: ['50%', '50%'],
            layoutSize: '100%',
            z: 12,
            itemStyle: {
                normal: {
                    areaColor: {
                        // TODO
                        image: 'mapBg',
                    },
                    borderColor: '#8aa5db',
                    borderWidth: 1,
                },
            },
            emphasis: {
                disabled: true,
            },
            label: {
                show: true,
                color: '#fff',
                fontSize: 12,
                fontWeight: 'normal',
                position: 'inside',
                distance: 5,
                align: 'center',
                verticalAlign: 'middle',
                backgroundColor: 'rgba(0,0,0,0.3)',
                padding: [3, 5],
                borderColor: 'rgba(255,255,255,0.5)',
                borderWidth: 1,
                borderRadius: 4,
                shadowColor: 'rgba(0,0,0,0.5)',
                shadowBlur: 5,
                shadowOffsetX: 1,
                shadowOffsetY: 1,
            },
        },
        series: [
            // 地图数据柱子
            ...lineSeriesData,
        ],
    };
    console.log(option, 'option geo');
    return option;
};
exports.getMapOption = getMapOption;
