// hfy-tools@2.1.6/dist/utils/fetch.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.get = get;
exports.post = post;
exports.postForm = postForm;
exports.del = del;
exports.put = put;
/**
 * fetch get
 * @param url
 * @returns
 */
async function get(url, data, headers) {
    if (data) {
        if (url.indexOf('?') === -1) {
            url += '?';
        }
        else {
            url += '&';
        }
        url += new URLSearchParams(data).toString();
    }
    return await (await fetch(url, {
        headers,
        method: 'get',
    })).json();
}
/**
 * fetch post
 * @param url
 * @returns
 */
async function post(url, data, headers) {
    return await (await fetch(url, {
        method: 'post',
        body: data ? JSON.stringify(data) : undefined,
        headers: {
            'Content-Type': 'application/json',
            ...(headers || {}),
        },
    })).json();
}
/**
 * fetch blob
 * @param url
 * @returns
 */
async function postForm(url, data, headers) {
    let body = new FormData();
    Object.keys(data).forEach((o) => {
        body.append(o, data[o]);
    });
    return await (await fetch(url, {
        method: 'post',
        body,
        headers: {
            ...headers,
            'Content-Type': 'multipart/form-data',
        },
    })).json();
}
async function del(url, headers) {
    return await (await fetch(url, {
        method: 'delete',
        headers,
    })).json();
}
async function put(url, data, headers) {
    return await (await fetch(url, {
        method: 'put',
        body: data ? JSON.stringify(data) : undefined,
        headers: {
            'Content-Type': 'application/json',
            ...(headers || {}),
        },
    })).json();
}
