// hfy-tools@2.1.6/dist/utils/indexDB.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DB = exports.DBCode = void 0;
var DBCode;
(function (DBCode) {
    DBCode[DBCode["success"] = 200] = "success";
    DBCode[DBCode["error"] = 401] = "error";
    DBCode[DBCode["open"] = 91001] = "open";
    DBCode[DBCode["save"] = 91002] = "save";
    DBCode[DBCode["update"] = 91003] = "update";
    DBCode[DBCode["get"] = 91004] = "get";
    DBCode[DBCode["delete"] = 91005] = "delete";
})(DBCode || (exports.DBCode = DBCode = {}));
class DB {
    dbName;
    dataBase;
    sqlVersion;
    dbVersion;
    recrive;
    on(name) { }
    getDataFromObjectStore(db, storeName) {
        const transaction = db.transaction(storeName, 'readonly');
        const objectStore = transaction.objectStore(storeName);
        // 获取所有数据
        const request = objectStore.getAll();
        request.onsuccess = function (event) {
            console.log(`数据表 ${storeName} 的所有数据:`, event.target.result);
            this.on('timer');
        };
        request.onerror = function (event) {
            console.error(`获取 ${storeName} 数据失败:`, event.target.error);
        };
    }
    async run() {
        const objectStoreNames = this.dataBase.objectStoreNames;
        // 遍历所有对象存储
        for (let i = 0; i < objectStoreNames.length; i++) {
            const storeName = objectStoreNames[i];
            this.getDataFromObjectStore(this.dataBase, storeName);
        }
    }
    getIndexedDB = () => {
        const globalScope = window;
        return (globalScope.indexedDB ||
            globalScope.msIndexedDB ||
            globalScope.mozIndexedDB ||
            globalScope.webkitIndexedDB);
    };
    constructor(dbName) {
        this.dbName = dbName;
        this.sqlVersion = '1'; // webSQL 的版本号，由于ios的问题，版本号的写法不一样
        this.dbVersion = 1; // indexDB的版本号
        // 数据库连接池，避免同一个页面重复创建和销毁
        // @ts-ignore
        this.indexedDB = getIndexedDB();
        if (!indexedDB) {
            throw new Error('浏览器不支持indexDB');
        }
        this.run();
    }
    async connect() {
        const self = this;
        return new Promise((resolve, reject) => {
            if (this.dataBase) {
                resolve(this.dataBase);
                return;
            }
            const request = indexedDB.open(this.dbName, this.dbVersion);
            request.onerror = () => {
                reject(new Error('数据库连接失败'));
            };
            request.onsuccess = () => {
                // @ts-ignore
                self.dataBase = request.result;
                resolve(request.result);
            };
            request.onupgradeneeded = (event) => {
                // @ts-ignore
                self.dataBase = event.target?.result;
                // @ts-ignore
                resolve(event.target?.result);
            };
        });
    }
    // 创建“表”
    async createTable(tableName) {
        const db = await this.connect();
        if (!db.objectStoreNames.contains(tableName)) {
            return db.createObjectStore(tableName, {
                keyPath: 'id',
                autoIncrement: true,
            });
        }
        return db.transaction(tableName, 'readwrite').objectStore(tableName);
    }
    // 加数据
    save(tableName, params) {
        return new Promise(async (resolve) => {
            await this.connect();
            const request = this.dataBase
                .transaction(tableName, 'readwrite')
                .objectStore(tableName)
                .add(params);
            request.onsuccess = () => {
                resolve({ code: DBCode.success, success: true }); //保存成功有success 字段
            };
            request.onerror = (e) => {
                resolve({ code: DBCode.save, error: e });
            };
        });
    }
    update(tableName, params) {
        return new Promise(async (resolve) => {
            await this.connect();
            const request = this.dataBase
                .transaction([tableName], 'readwrite')
                .objectStore(tableName)
                .put(params);
            request.onsuccess = () => {
                console.log('数据更新成功');
                resolve({ code: DBCode.success, msg: '数据更新成功' });
            };
            request.onerror = (event) => {
                resolve({ code: DBCode.update, msg: '数据更新失败', error: event });
            };
        });
    }
    // 查单个
    get(tableName, params) {
        return new Promise(async (resolve) => {
            await this.connect();
            const objectStore = this.dataBase
                .transaction([tableName])
                .objectStore(tableName);
            const request = objectStore.get(params);
            request.onerror = (e) => {
                resolve({ code: DBCode.get, error: e });
            };
            request.onsuccess = (event) => {
                resolve({ code: DBCode.success, data: request.result });
            };
        });
    }
    // 获取表格所有数据
    list(tableName, params) {
        return new Promise(async (resolve) => {
            await this.connect();
            const objectStore = this.dataBase
                .transaction([tableName])
                .objectStore(tableName);
            let result = [];
            objectStore.openCursor(params).onsuccess = (event) => {
                const cursor = event.target.result;
                if (cursor) {
                    console.log('Id: ' + cursor.key);
                    result.push(cursor.value);
                    cursor.continue();
                }
                else {
                    resolve({ code: DBCode.success, data: result });
                }
            };
        });
    }
    // 删
    remove(tableName, params) {
        return new Promise(async (resolve) => {
            await this.connect();
            const request = this.dataBase
                .transaction([tableName], 'readwrite')
                .objectStore(tableName)
                .delete(params);
            request.onsuccess = () => {
                resolve({ code: DBCode.success, msg: '数据删除成功' });
            };
            request.onerror = (e) => {
                resolve({ code: DBCode.delete, error: e });
            };
        });
    }
    // 清表
    clear(tableName) {
        this.dataBase
            .transaction([tableName], 'readwrite')
            .objectStore(tableName)
            .clear();
    }
    close() {
        window.indexedDB.deleteDatabase(this.dbName);
    }
}
exports.DB = DB;
