// hfy-tools@2.1.6/dist/utils/math.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.formatDecimal = void 0;
exports.add = add;
exports.subtract = subtract;
exports.multiply = multiply;
exports.divide = divide;
exports.round = round;
exports.compare = compare;
const decimal_js_1 = __importDefault(require("decimal.js"));
/**
 * 保留几位小数不是四舍五入
 * @param value 数值
 * @param decimal  =保留几位小数
 * @returns 处理后的值
 */
const formatDecimal = (value, decimal = 7) => {
    value = value.toString();
    const index = value.indexOf('.');
    if (index !== -1) {
        value = value.substring(0, decimal + index + 1);
    }
    else {
        value = value.substring(0);
    }
    return +parseFloat(value).toFixed(decimal);
};
exports.formatDecimal = formatDecimal;
// 加法 (0.1 + 0.2 = 0.3)
function add(a, b) {
    return new decimal_js_1.default(a).add(new decimal_js_1.default(b)).toNumber();
}
// 减法 (1.0 - 0.7 = 0.3)
function subtract(a, b) {
    return new decimal_js_1.default(a).sub(new decimal_js_1.default(b)).toNumber();
}
// 乘法 (0.6 * 3 = 1.8)
function multiply(a, b) {
    return new decimal_js_1.default(a).mul(new decimal_js_1.default(b)).toNumber();
}
// 除法 (1.2 / 3 = 0.4)
function divide(a, b) {
    return new decimal_js_1.default(a).div(new decimal_js_1.default(b)).toNumber();
}
// 四舍五入
function round(num, precision = 2) {
    return new decimal_js_1.default(num).toDecimalPlaces(precision).toNumber();
}
// 比较大小
function compare(a, b) {
    return new decimal_js_1.default(a).comparedTo(new decimal_js_1.default(b));
}
