// hfy-tools@2.1.6/dist/utils/oss.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const cos_js_sdk_v5_1 = __importDefault(require("cos-js-sdk-v5"));
const core = __importStar(require("./core"));
const base64 = __importStar(require("./base64"));
const fetch = __importStar(require("./fetch"));
const date_1 = require("./date");
class OSS {
    Bucket;
    Region;
    SecretKey;
    SecretId;
    prefix;
    isTemp;
    cos;
    params;
    constructor({ Bucket, Region, SecretKey, SecretId, prefix, isTemp }) {
        this.Bucket = Bucket;
        this.Region = Region;
        this.SecretKey = SecretKey;
        this.SecretId = SecretId;
        this.prefix = prefix;
        this.isTemp = isTemp;
        this.prefix = prefix || base64.encode('GOSS');
        this.cos = new cos_js_sdk_v5_1.default({
            SecretId: SecretId,
            SecretKey: SecretKey
        });
        this.params = { Bucket, Region };
    }
    /**
     *
     * @param file 文件
     * @param key 上传文件key
     * @param progress 上传进度函数回调
     * @param dir 目录
     * @returns
     */
    upload({ file, dir, progress, key, expirationDate }) {
        return new Promise((resolve, reject) => {
            let uploadKey = base64.encode(JSON.stringify({ name: file.name, id: key || core.guid(), dir }));
            // console.log('文件上传', uploadKey, this.isTemp, file)
            if (this.isTemp) {
                const d = new date_1.SimpleDate();
                d.addSeconds(3);
                uploadKey = `temp/${expirationDate?.getTime() || d.getTime()}/` + uploadKey;
            }
            // console.log('上传的key', `${pre}/${key}`)
            this.cos.sliceUploadFile({
                ...this.params,
                Key: `${this.prefix}/${uploadKey}`,
                StorageClass: 'STANDARD', // [Standard, Standard_IA，Archive]
                Body: file, // 上传文件对象
                onProgress: progress
            }, (err, data) => {
                if (err) {
                    reject(err);
                    return;
                }
                resolve(data);
            });
        });
    }
    /**
     * 批量删除文件
     * @param keys 文件key
     * @returns
     */
    multiDelete(...keys) {
        return new Promise((resolve, rej) => {
            this.cos.deleteMultipleObject({
                ...this.params,
                Objects: keys.map((o) => ({ Key: o }))
            }, (err, data) => {
                resolve(data);
            });
        });
    }
    /**
     * 获取文件
     * @param key
     * @param progress
     * @param isGetUrl
     * @returns
     */
    getFile(key, progress, isGetUrl = true) {
        return new Promise((resolve, reject) => {
            // console.log(key, progress, isGetUrl);
            if (key.startsWith(this.prefix + '/temp')) {
                const expireDate = new Date(key.split('/')[2]);
                if (expireDate <= new Date()) {
                    this.multiDelete(key);
                    reject('文件已过期');
                    return;
                }
            }
            if (isGetUrl) {
                this.cos.getObjectUrl({
                    ...this.params,
                    Key: `${key}`,
                    Sign: true
                }, (err, data) => {
                    if (err) {
                        reject(new Error(err.message));
                        return;
                    }
                    resolve(data);
                });
                return;
            }
            this.cos.getObject({
                ...this.params,
                DataType: 'blob',
                Key: `${key}`
            }, (err, data) => {
                if (err) {
                    reject(new Error(err.message));
                    return;
                }
                resolve(data);
            });
        });
    }
    /**
     * 文件拷贝
     * @param key
     * @param sourceKey
     * @param pms
     * @returns
     */
    copy(key, sourceKey, pms = {}) {
        return new Promise((resolve, reject) => {
            key = base64.encode(JSON.stringify({ key, ...pms }));
            this.cos.putObjectCopy({
                ...this.params,
                Key: `${this.prefix}/${key}`,
                CopySource: `${this.params.Bucket}.cos.ap-guangzhou.myqcloud.com/${sourceKey}`
            }, (err, data) => {
                if (err) {
                    reject(new Error(err.message));
                    return;
                }
                resolve(data);
            });
        });
    }
    /**
     * 根据前缀获取文件
     * @param pre
     * @returns
     */
    getFiles() {
        return new Promise((resolve, reject) => {
            this.cos.getBucket({
                ...this.params,
                Prefix: `${this.prefix}`
            }, (err, data) => {
                if (err) {
                    reject(new Error(err.message));
                    resolve([]);
                    return;
                }
                const result = data.Contents.reduce((res, o) => {
                    // console.log('contents', o, o.Key.split('/')[2])
                    if (o.Key.startsWith(`${this.prefix}/temp`)) {
                        const expireDate = new Date(Number(o.Key.split('/')[2]));
                        if (expireDate <= new Date()) {
                            this.multiDelete(o.Key);
                            reject('文件已过期');
                            return res;
                        }
                    }
                    let valueStr = o.Key.split('/').pop();
                    valueStr = base64.decode(`${valueStr}`);
                    try {
                        res.push({ ...JSON.parse(valueStr), key: o.Key });
                    }
                    catch (error) {
                        console.warn(error);
                        res.push({ name: '解析失败', key: o.Key });
                    }
                    return res;
                }, []);
                resolve(result);
            });
        });
    }
    async getDirectory() {
        const result = await this.getFiles();
        return this.buildDirectory(result);
    }
    buildDirectory(files) {
        return files.reduce((res, cur) => {
            let dirName = cur.id;
            if (cur.dir) {
                dirName = cur.dir;
            }
            res[dirName] = [cur].concat(res[dirName] ?? []);
            return res;
        }, {});
    }
}
class DB {
    OSS;
    constructor(ossParams) {
        this.OSS = new OSS({
            ...ossParams,
            prefix: ossParams.prefix ?? base64.encode('GOSS') + '/table'
        });
    }
    add(name, data) {
        return new Promise(async (resolve, reject) => {
            const list = await this.get(name);
            list.push({ ...data, guid: core.guid() });
            const blob = new File([JSON.stringify(list)], name + '.json', {
                type: 'application/json'
            });
            this.OSS.upload({ file: blob, dir: '', key: '' });
        });
    }
    update(name, id, data) {
        return new Promise(async (resolve, reject) => {
            const list = await this.get(name);
            const idx = list.findIndex((o) => Object.is(o.guid, id));
            list.splice(idx, 1, { ...data, guid: id });
            this.add(name, list);
        });
    }
    delete(name, id) {
        return new Promise(async (resolve, reject) => {
            const list = await this.get(name);
            this.add(name, list.filter((o) => o.guid !== id));
        });
    }
    get(name) {
        let uploadKey = base64.encode(JSON.stringify({ name: name + '.json', id: '', dir: '' }));
        let prefix = base64.encode('GOSS');
        return fetch.get(`https://jf-temp-1301446188.cos.ap-guangzhou.myqcloud.com/${prefix}/table/${uploadKey}`);
    }
    list() {
        return new Promise((resolve) => {
            this.OSS.getFiles().then(resolve);
        });
    }
}
exports.default = {
    COS: OSS,
    DB
};
