// hfy-tools@2.1.6/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.web = exports.constants = exports.tree = exports.subscribe = exports.result = exports.regular = exports.postMessage = exports.math = exports.localStorage = exports.is = exports.IndexDB = exports.file = exports.fetch = exports.OSS = exports.dom = exports.document = exports.date = exports.geo = exports.office = exports.core = exports.base64 = void 0;
exports.base64 = __importStar(require("./utils/base64"));
exports.core = __importStar(require("./utils/core"));
exports.office = __importStar(require("./utils/office"));
exports.geo = __importStar(require("./utils/geo"));
exports.date = __importStar(require("./utils/date"));
exports.document = __importStar(require("./utils/document"));
exports.dom = __importStar(require("./utils/dom"));
exports.OSS = __importStar(require("./utils/oss"));
exports.fetch = __importStar(require("./utils/fetch"));
exports.file = __importStar(require("./utils/file"));
exports.IndexDB = __importStar(require("./utils/indexDB"));
exports.is = __importStar(require("./utils/is"));
exports.localStorage = __importStar(require("./utils/localStorage"));
exports.math = __importStar(require("./utils/math"));
exports.postMessage = __importStar(require("./utils/post-message"));
exports.regular = __importStar(require("./utils/regular"));
exports.result = __importStar(require("./utils/result"));
exports.subscribe = __importStar(require("./utils/subscribe"));
exports.tree = __importStar(require("./utils/tree"));
exports.constants = __importStar(require("./utils/constants"));
exports.web = __importStar(require("./component/web"));
